{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "26a4fd69",
   "metadata": {},
   "source": [
    "# Requests, Exceptions, Generators\n",
    "## Advanced Python for Life Sciences @ Physalia courses (Summer 2025)\n",
    "### Marco Chierici, Fondazione Bruno Kessler"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93778d10-5686-49fc-ac25-19a4f9603c50",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# HTTP Requests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45a471be-c78a-416a-9e4c-79a3585e1fe7",
   "metadata": {},
   "source": [
    "## Basic requests\n",
    "\n",
    "HTTP requests are commonly used to retrieve data from a specified resource, such as a website. They retrieve data from an API (Application Programmable Interface): the API defines how computer programs communicate with each other, and how you should form your query in order to get something from the website."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b8468686-b198-4573-a06e-a3942f198176",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'{\"fact\":\"A cat has more bones than a human being; humans have 206 and the cat has 230 bones.\",\"length\":83}'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "\n",
    "url = \"https://catfact.ninja/fact\"\n",
    "\n",
    "# Make a request\n",
    "response = requests.get(url)\n",
    "response.content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e42a3d6e-3d97-4986-9f33-71bc5906a0f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'fact': 'A cat has more bones than a human being; humans have 206 and the cat has 230 bones.',\n",
       " 'length': 83}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.json()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "968ae88a-9b4f-4119-a5de-b12679694c77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "200"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.status_code"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3d875df-ac57-4a21-b9ae-1062f00124bb",
   "metadata": {},
   "source": [
    "## API with parameters\n",
    "\n",
    "The \"Cat facts\" API above is very simple: you cannot pass additional parameters to it. Most of the times, however, APIs will be more complex and accept further parameters. \n",
    "\n",
    "To deal with this, you can use `requests.get(url, params)`, where\n",
    "\n",
    "- `url` is the base URL of the API, and\n",
    "- `params` is a Python dictionary containing API's parameters and corresponding values as key-value pairs.\n",
    "\n",
    "### Exercise\n",
    "\n",
    "You will retrieve a list of universities and their domain names using a web API.\n",
    "\n",
    "1. Try a simple query on the URL `http://universities.hipolabs.com/search?country=Finland` using the above method. Note that here we have:\n",
    "  - the base URL, `http://universities.hipolabs.com/search`\n",
    "  - followed by `?`\n",
    "  - followed by the parameter `country` set to the value `Finland`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25f6cbab-0e91-4517-bd5e-ce53ee9f1c5c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0e4ddb01-9ff4-47f3-a5c4-d13ed9c73da8",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = \"http://universities.hipolabs.com/search?country=Finland\"\n",
    "\n",
    "# Make a request\n",
    "response = requests.get(url)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3ca9762-92e8-4082-a6e6-5d1450149be8",
   "metadata": {},
   "source": [
    "2. Display the response in JSON format, like above, but this time limit the output to the first two elements (use Python's list slicing syntax)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a91bcbd1-b941-4b9f-ae6d-a5613688311e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "48c323ce-254c-4896-8479-b4766e8f854a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'country': 'Finland',\n",
       "  'name': 'Abo Akademi University',\n",
       "  'alpha_two_code': 'FI',\n",
       "  'state-province': None,\n",
       "  'web_pages': ['http://www.abo.fi/'],\n",
       "  'domains': ['abo.fi']},\n",
       " {'country': 'Finland',\n",
       "  'name': 'Central Ostrobothnia University of Applied Sciences',\n",
       "  'alpha_two_code': 'FI',\n",
       "  'state-province': None,\n",
       "  'web_pages': ['http://www.cou.fi/'],\n",
       "  'domains': ['cou.fi']}]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "response.json()[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52a0e0ee-1664-47cc-97e2-2f4c0cb52e50",
   "metadata": {},
   "source": [
    "3. Adapt your code and repeat the query using the `requests.get(url, params)` syntax."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f3ec268-cb3d-45a5-9cd7-0e427faa98c2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c02652a5-ec30-4938-b451-4fe901333b6d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'country': 'Finland',\n",
       "  'name': 'Abo Akademi University',\n",
       "  'alpha_two_code': 'FI',\n",
       "  'state-province': None,\n",
       "  'web_pages': ['http://www.abo.fi/'],\n",
       "  'domains': ['abo.fi']},\n",
       " {'country': 'Finland',\n",
       "  'name': 'Central Ostrobothnia University of Applied Sciences',\n",
       "  'alpha_two_code': 'FI',\n",
       "  'state-province': None,\n",
       "  'web_pages': ['http://www.cou.fi/'],\n",
       "  'domains': ['cou.fi']}]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "url = \"http://universities.hipolabs.com/search\"\n",
    "params = {\"country\": \"Finland\"}\n",
    "\n",
    "# Make a request\n",
    "response = requests.get(url, params=params)\n",
    "response.json()[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46e7c4bc-aac6-4170-86be-c39f0592ae96",
   "metadata": {},
   "source": [
    "## Querying UniProt\n",
    "\n",
    "This is a case study for querying an API with parameters. Suppose you want to download the FASTA sequence of the DNA replication licensing factor MCM7 from UniProt.\n",
    "\n",
    "By reading the [UniProt API documentation](https://www.uniprot.org/help/api_retrieve_entries), you know that the endpoint for getting a FASTA output from a UniProt ID is in the form\n",
    "\n",
    "```https://rest.uniprot.org/uniprotkb/<UNIPROT ID>.fasta```\n",
    "\n",
    "Given that the UniProt ID for MCM7 is P33993, let's give it a try with Python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c91a5556-4211-4807-8f83-f4083174a91b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">sp|P33993|MCM7_HUMAN DNA replication licensing factor MCM7 OS=Homo sapiens OX=9606 GN=MCM7 PE=1 SV=4\n",
      "MALKDYALEKEKVKKFLQEFYQDDELGKKQFKYGNQLVRLAHREQVALYVDLDDVAEDDP\n",
      "ELVDSICENARRYAKLFADAVQELLPQYKEREVVNKDVLDVYIEHRLMMEQRSRDPGMVR\n",
      "SPQNQYPAELMRRFELYFQGPSSNKPRVIREVRADSVGKLVTVRGIVTRVSEVKPKMVVA\n",
      "TYTCDQCGAETYQPIQSPTFMPLIMCPSQECQTNRSGGRLYLQTRGSRFIKFQEMKMQEH\n",
      "SDQVPVGNIPRSITVLVEGENTRIAQPGDHVSVTGIFLPILRTGFRQVVQGLLSETYLEA\n",
      "HRIVKMNKSEDDESGAGELTREELRQIAEEDFYEKLAASIAPEIYGHEDVKKALLLLLVG\n",
      "GVDQSPRGMKIRGNINICLMGDPGVAKSQLLSYIDRLAPRSQYTTGRGSSGVGLTAAVLR\n",
      "DSVSGELTLEGGALVLADQGVCCIDEFDKMAEADRTAIHEVMEQQTISIAKAGILTTLNA\n",
      "RCSILAAANPAYGRYNPRRSLEQNIQLPAALLSRFDLLWLIQDRPDRDNDLRLAQHITYV\n",
      "HQHSRQPPSQFEPLDMKLMRRYIAMCREKQPMVPESLADYITAAYVEMRREAWASKDATY\n",
      "TSARTLLAILRLSTALARLRMVDVVEKEDVNEAIRLMEMSKDSLLGDKGQTARTQRPADV\n",
      "IFATVRELVSGGRSVRFSEAEQRCVSRGFTPAQFQAALDEYEELNVWQVNASRTRITFV\n",
      "\n"
     ]
    }
   ],
   "source": [
    "uniprot_id = \"P33993\"\n",
    "url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "response = requests.get(url)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb10e305-be92-4fc1-ae01-0673fa3938c1",
   "metadata": {},
   "source": [
    "## NCBI Eutils\n",
    "\n",
    "Let's now query the [Entrez](https://www.ncbi.nlm.nih.gov/books/NBK25500/) nucleotide db for the gene id 1676324825 (MYCN proto-oncogene transcript variant 3), returning the output in Genbank format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d8df7d03-40be-4785-95cd-e09b2f23c59b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NM_001293231            1706 bp    mRNA    linear   PRI 28-APR-2025\n",
      "DEFINITION  Homo sapiens MYCN proto-oncogene, bHLH transcription factor (MYCN),\n",
      "            transcript variant 3, mRNA.\n",
      "ACCESSION   NM_001293231\n",
      "VERSION     NM_001293231.2\n",
      "KEYWORDS    RefSeq.\n",
      "SOURCE      Homo sapiens (human)\n",
      "  ORGANISM  Homo sapiens\n",
      "            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;\n",
      "            Mammalia; Eutheria; Euarchontoglires; Primates; Haplorrhini;\n",
      "            Catarrhini; Hominidae; Homo.\n",
      "REFERENCE   1  (bases 1 to 1706)\n",
      "  AUTHORS   Vempuluru,V.S., Maniar,A., Bakal,K. and Kaliki,S.\n",
      "  TITLE     Role of MYCN in retinoblastoma: A review of current literature\n",
      "  JOURNAL   Surv Ophthalmol 69 (5), 697-706 (2024)\n",
      "   PUBMED   38796108\n",
      "  REMARK    GeneRIF: Role of MYCN in retinoblastoma: A review of current\n",
      "            literature.\n",
      "            Review article\n",
      "REFERENCE   2  (bases 1 to 1706)\n",
      "  AUTHORS   Sundaramoorthy,S., Colombo,D.F., Sanalkumar,R., Broye,L., Balmas\n",
      "            Bourloud,K., Boulay,G., Cironi,L., Stamenkovic,I., Renella,R.,\n",
      "            Kuttler,F., Turcatti,G., Rivera,M.N., Muhlethaler-Mottet,A.,\n",
      "            Bardet,A.F. and Riggi,N.\n",
      "  TITLE     Preclinical spheroid models identify BMX as a therapeutic target\n",
      "            for metastatic MYCN nonamplified neuroblastoma\n",
      "  JOURNAL   JCI Insight 9 (14), e169647 (2024)\n",
      "   PUBMED   39133652\n",
      "  REMARK    GeneRIF: Preclinical spheroid models identify BMX as a therapeutic\n",
      "            target for metastatic MYCN nonamplified neuroblastoma.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   3  (bases 1 to 1706)\n",
      "  AUTHORS   Murray,J.E., Valli,E., Milazzo,G., Mayoh,C., Gifford,A.J.,\n",
      "            Fletcher,J.I., Xue,C., Jayatilleke,N., Salehzadeh,F., Gamble,L.D.,\n",
      "            Rouaen,J.R.C., Carter,D.R., Forgham,H., Sekyere,E.O., Keating,J.,\n",
      "            Eden,G., Allan,S., Alfred,S., Kusuma,F.K., Clark,A., Webber,H.,\n",
      "            Russell,A.J., de Weck,A., Kile,B.T., Santulli,M., De Rosa,P.,\n",
      "            Fleuren,E.D.G., Gao,W., Wilkinson-White,L., Low,J.K.K.,\n",
      "            Mackay,J.P., Marshall,G.M., Hilton,D.J., Giorgi,F.M., Koster,J.,\n",
      "            Perini,G., Haber,M. and Norris,M.D.\n",
      "  TITLE     The transcriptional co-repressor Runx1t1 is essential for\n",
      "            MYCN-driven neuroblastoma tumorigenesis\n",
      "  JOURNAL   Nat Commun 15 (1), 5585 (2024)\n",
      "   PUBMED   38992040\n",
      "  REMARK    GeneRIF: The transcriptional co-repressor Runx1t1 is essential for\n",
      "            MYCN-driven neuroblastoma tumorigenesis.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   4  (bases 1 to 1706)\n",
      "  AUTHORS   Besancon,R., Valsesia-Wittmann,S., Locher,C., Delloye-Bourgeois,C.,\n",
      "            Furhman,L., Tutrone,G., Bertrand,C., Jallas,A.C., Garin,E. and\n",
      "            Puisieux,A.\n",
      "  TITLE     Upstream ORF affects MYCN translation depending on exon 1b\n",
      "            alternative splicing\n",
      "  JOURNAL   BMC Cancer 9, 445 (2009)\n",
      "   PUBMED   20017904\n",
      "  REMARK    GeneRIF: Existence of upstream ORF in MYCN transcripts leads to a\n",
      "            new level of MYCN regulation.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   5  (bases 1 to 1706)\n",
      "  AUTHORS   Marcelis,C.L.M. and de Brouwer,A.P.M.\n",
      "  TITLE     Feingold Syndrome 1\n",
      "  JOURNAL   (in) Adam MP, Feldman J, Mirzaa GM, Pagon RA, Wallace SE and\n",
      "            Amemiya A (Eds.);\n",
      "            GENEREVIEWS(R);\n",
      "            (1993)\n",
      "   PUBMED   20301770\n",
      "REFERENCE   6  (bases 1 to 1706)\n",
      "  AUTHORS   Scott,D.A.\n",
      "  TITLE     Esophageal Atresia / Tracheoesophageal Fistula Overview - RETIRED\n",
      "            CHAPTER, FOR HISTORICAL REFERENCE ONLY\n",
      "  JOURNAL   (in) Adam MP, Feldman J, Mirzaa GM, Pagon RA, Wallace SE and\n",
      "            Amemiya A (Eds.);\n",
      "            GENEREVIEWS(R);\n",
      "            (1993)\n",
      "   PUBMED   20301753\n",
      "REFERENCE   7  (bases 1 to 1706)\n",
      "  AUTHORS   Hagiwara,T., Nakaya,K., Nakamura,Y., Nakajima,H., Nishimura,S. and\n",
      "            Taya,Y.\n",
      "  TITLE     Specific phosphorylation of the acidic central region of the N-myc\n",
      "            protein by casein kinase II\n",
      "  JOURNAL   Eur J Biochem 209 (3), 945-950 (1992)\n",
      "   PUBMED   1425701\n",
      "REFERENCE   8  (bases 1 to 1706)\n",
      "  AUTHORS   Armstrong,B.C. and Krystal,G.W.\n",
      "  TITLE     Isolation and characterization of complementary DNA for N-cym, a\n",
      "            gene encoded by the DNA strand opposite to N-myc\n",
      "  JOURNAL   Cell Growth Differ 3 (6), 385-390 (1992)\n",
      "   PUBMED   1419902\n",
      "REFERENCE   9  (bases 1 to 1706)\n",
      "  AUTHORS   Fougerousse,F., Meloni,R., Roudaut,C. and Beckmann,J.S.\n",
      "  TITLE     Dinucleotide repeat polymorphism at the human hemoglobin alpha-1\n",
      "            pseudo-gene (HBAP1)\n",
      "  JOURNAL   Nucleic Acids Res 20 (5), 1165 (1992)\n",
      "   PUBMED   1549498\n",
      "REFERENCE   10 (bases 1 to 1706)\n",
      "  AUTHORS   Stanton,L.W. and Bishop,J.M.\n",
      "  TITLE     Alternative processing of RNA transcribed from NMYC\n",
      "  JOURNAL   Mol Cell Biol 7 (12), 4266-4272 (1987)\n",
      "   PUBMED   3437890\n",
      "COMMENT     REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "            reference sequence was derived from AC010145.10 and BC002712.2.\n",
      "            \n",
      "            On May 31, 2019 this sequence version replaced NM_001293231.1.\n",
      "            \n",
      "            Summary: This gene is a member of the MYC family and encodes a\n",
      "            protein with a basic helix-loop-helix (bHLH) domain. This protein\n",
      "            is located in the nucleus and must dimerize with another bHLH\n",
      "            protein in order to bind DNA. Amplification of this gene is\n",
      "            associated with a variety of tumors, most notably neuroblastomas.\n",
      "            Multiple alternatively spliced transcript variants encoding\n",
      "            different isoforms have been found for this gene. [provided by\n",
      "            RefSeq, Jun 2014].\n",
      "            \n",
      "            Transcript Variant: This variant (3) lacks segment 1b and exon 2,\n",
      "            which results in an upstream AUG start codon, compared to variant\n",
      "            1. The resulting isoform (2) has a shorter and distinct N-terminus,\n",
      "            compared to isoform 1.\n",
      "            \n",
      "            Sequence Note: This RefSeq record was created from transcript and\n",
      "            genomic sequence data to make the sequence consistent with the\n",
      "            reference genome assembly. The genomic coordinates used for the\n",
      "            transcript record were based on transcript alignments.\n",
      "            \n",
      "            Publication Note:  This RefSeq record includes a subset of the\n",
      "            publications that are available for this gene. Please see the Gene\n",
      "            record to access additional publications.\n",
      "            \n",
      "            ##Evidence-Data-START##\n",
      "            Transcript exon combination :: BE382888.1, BI194921.1 [ECO:0000332]\n",
      "            ##Evidence-Data-END##\n",
      "PRIMARY     REFSEQ_SPAN         PRIMARY_IDENTIFIER PRIMARY_SPAN        COMP\n",
      "            1-194               AC010145.10        65895-66088\n",
      "            195-1706            BC002712.2         932-2443\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..1706\n",
      "                     /organism=\"Homo sapiens\"\n",
      "                     /mol_type=\"mRNA\"\n",
      "                     /db_xref=\"taxon:9606\"\n",
      "                     /chromosome=\"2\"\n",
      "                     /map=\"2p24.3\"\n",
      "     gene            1..1706\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /note=\"MYCN proto-oncogene, bHLH transcription factor\"\n",
      "                     /db_xref=\"GeneID:4613\"\n",
      "                     /db_xref=\"HGNC:HGNC:7559\"\n",
      "                     /db_xref=\"MIM:164840\"\n",
      "     exon            1..194\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /inference=\"alignment:Splign:2.1.0\"\n",
      "     CDS             38..799\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /note=\"isoform 2 is encoded by transcript variant 3;\n",
      "                     neuroblastoma MYC oncogene; N-myc proto-oncogene protein;\n",
      "                     oncogene NMYC; neuroblastoma-derived v-myc avian\n",
      "                     myelocytomatosis viral related oncogene; class E basic\n",
      "                     helix-loop-helix protein 37; v-myc avian myelocytomatosis\n",
      "                     viral oncogene neuroblastoma derived homolog\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"N-myc proto-oncogene protein isoform 2\"\n",
      "                     /protein_id=\"NP_001280160.1\"\n",
      "                     /db_xref=\"CCDS:CCDS86823.1\"\n",
      "                     /db_xref=\"GeneID:4613\"\n",
      "                     /db_xref=\"HGNC:HGNC:7559\"\n",
      "                     /db_xref=\"MIM:164840\"\n",
      "                     /translation=\"MRGAPGNCVGAEQALARRKRAQTVAIRGHPRPPGPPGDTRAESP\n",
      "                     PDPLQSAGDDEDDEEEDEEEEIDVVTVEKRRSSSNTKAVTTFTITVRPKNAALGPGRA\n",
      "                     QSSELILKRCLPIHQQHNYAAPSPYVESEDAPPQKKIKSEASPRPLKSVIPPKAKSLS\n",
      "                     PRNSDSEDSERRRNHNILERQRRNDLRSSFLTLRDHVPELVKNEKAAKVVILKKATEY\n",
      "                     VHSLQAEEHQLLLEKEKLQARQQQLLKKIEHARTC\"\n",
      "     exon            195..1706\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /inference=\"alignment:Splign:2.1.0\"\n",
      "     regulatory      1686..1691\n",
      "                     /regulatory_class=\"polyA_signal_sequence\"\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /note=\"hexamer: AATAAA\"\n",
      "     polyA_site      1706\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /note=\"major polyA site\"\n",
      "ORIGIN      \n",
      "        1 aggctgtgac agtcatctgt ctggacgcgc tgggtggatg cggggggctc ctgggaactg\n",
      "       61 tgttggagcc gagcaagcgc tagccaggcg caagcgcgca cagactgtag ccatccgagg\n",
      "      121 acacccccgc ccccccggcc cacccggaga cacccgcgca gaatcgcctc cggatcccct\n",
      "      181 gcagtcggcg ggagatgatg aagatgatga agaggaagat gaagaggaag aaatcgacgt\n",
      "      241 ggtcactgtg gagaagcggc gttcctcctc caacaccaag gctgtcacca cattcaccat\n",
      "      301 cactgtgcgt cccaagaacg cagccctggg tcccgggagg gctcagtcca gcgagctgat\n",
      "      361 cctcaaacga tgccttccca tccaccagca gcacaactat gccgccccct ctccctacgt\n",
      "      421 ggagagtgag gatgcacccc cacagaagaa gataaagagc gaggcgtccc cacgtccgct\n",
      "      481 caagagtgtc atccccccaa aggctaagag cttgagcccc cgaaactctg actcggagga\n",
      "      541 cagtgagcgt cgcagaaacc acaacatcct ggagcgccag cgccgcaacg accttcggtc\n",
      "      601 cagctttctc acgctcaggg accacgtgcc ggagttggta aagaatgaga aggccgccaa\n",
      "      661 ggtggtcatt ttgaaaaagg ccactgagta tgtccactcc ctccaggccg aggagcacca\n",
      "      721 gcttttgctg gaaaaggaaa aattgcaggc aagacagcag cagttgctaa agaaaattga\n",
      "      781 acacgctcgg acttgctaga cgcttctcaa aactggacag tcactgccac tttgcacatt\n",
      "      841 ttgatttttt ttttaaacaa acattgtgtt gacattaaga atgttggttt actttcaaat\n",
      "      901 cggtcccctg tcgagttcgg ctctgggtgg gcagtaggac caccagtgtg gggttctgct\n",
      "      961 gggaccttgg agagcctgca tcccaggatg ctgggtggcc ctgcagcctc ctccacctca\n",
      "     1021 cctccatgac agcgctaaac gttggtgacg gttgggagcc tctggggctg ttgaagtcac\n",
      "     1081 cttgtgtgtt ccaagtttcc aaacaacaga aagtcattcc ttctttttaa aatggtgctt\n",
      "     1141 aagttccagc agatgccaca taaggggttt gccatttgat acccctgggg aacatttctg\n",
      "     1201 taaataccat tgacacatcc gccttttgta tacatcctgg gtaatgagag gtggcttttg\n",
      "     1261 cggccagtat tagactggaa gttcatacct aagtactgta ataatacctc aatgtttgag\n",
      "     1321 gagcatgttt tgtatacaaa tatattgtta atctctgtta tgtactgtac taattcttac\n",
      "     1381 actgcctgta tactttagta tgacgctgat acataactaa atttgatact tatattttcg\n",
      "     1441 tatgaaaatg agttgtgaaa gttttgagta gatattactt tatcactttt tgaactaaga\n",
      "     1501 aacttttgta aagaaattta ctatatatat atgccttttt cctagcctgt ttcttcctgt\n",
      "     1561 taatgtattt gttcatgttt ggtgcataga actgggtaaa tgcaaagttc tgtgtttaat\n",
      "     1621 ttcttcaaaa tgtatatatt tagtgctgca tcttatagca ctttgaaata cctcatgttt\n",
      "     1681 atgaaaataa atagcttaaa attaaa\n",
      "//\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# simple query\n",
    "gene_id = 1676324825\n",
    "url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id={gene_id}&rettype=gb\"\n",
    "response = requests.get(url)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8662014-1a5a-47ee-bcf6-148c89c042e9",
   "metadata": {},
   "source": [
    "**Your turn:** create an improved and more readable version of the above code to work with the `params` syntax."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b5b5ff4-4ed7-4a4c-ac1b-3ffc1934d221",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "58065de0-d912-48f2-858d-c7bd9f883c86",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOCUS       NM_001293231            1706 bp    mRNA    linear   PRI 28-APR-2025\n",
      "DEFINITION  Homo sapiens MYCN proto-oncogene, bHLH transcription factor (MYCN),\n",
      "            transcript variant 3, mRNA.\n",
      "ACCESSION   NM_001293231\n",
      "VERSION     NM_001293231.2\n",
      "KEYWORDS    RefSeq.\n",
      "SOURCE      Homo sapiens (human)\n",
      "  ORGANISM  Homo sapiens\n",
      "            Eukaryota; Metazoa; Chordata; Craniata; Vertebrata; Euteleostomi;\n",
      "            Mammalia; Eutheria; Euarchontoglires; Primates; Haplorrhini;\n",
      "            Catarrhini; Hominidae; Homo.\n",
      "REFERENCE   1  (bases 1 to 1706)\n",
      "  AUTHORS   Vempuluru,V.S., Maniar,A., Bakal,K. and Kaliki,S.\n",
      "  TITLE     Role of MYCN in retinoblastoma: A review of current literature\n",
      "  JOURNAL   Surv Ophthalmol 69 (5), 697-706 (2024)\n",
      "   PUBMED   38796108\n",
      "  REMARK    GeneRIF: Role of MYCN in retinoblastoma: A review of current\n",
      "            literature.\n",
      "            Review article\n",
      "REFERENCE   2  (bases 1 to 1706)\n",
      "  AUTHORS   Sundaramoorthy,S., Colombo,D.F., Sanalkumar,R., Broye,L., Balmas\n",
      "            Bourloud,K., Boulay,G., Cironi,L., Stamenkovic,I., Renella,R.,\n",
      "            Kuttler,F., Turcatti,G., Rivera,M.N., Muhlethaler-Mottet,A.,\n",
      "            Bardet,A.F. and Riggi,N.\n",
      "  TITLE     Preclinical spheroid models identify BMX as a therapeutic target\n",
      "            for metastatic MYCN nonamplified neuroblastoma\n",
      "  JOURNAL   JCI Insight 9 (14), e169647 (2024)\n",
      "   PUBMED   39133652\n",
      "  REMARK    GeneRIF: Preclinical spheroid models identify BMX as a therapeutic\n",
      "            target for metastatic MYCN nonamplified neuroblastoma.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   3  (bases 1 to 1706)\n",
      "  AUTHORS   Murray,J.E., Valli,E., Milazzo,G., Mayoh,C., Gifford,A.J.,\n",
      "            Fletcher,J.I., Xue,C., Jayatilleke,N., Salehzadeh,F., Gamble,L.D.,\n",
      "            Rouaen,J.R.C., Carter,D.R., Forgham,H., Sekyere,E.O., Keating,J.,\n",
      "            Eden,G., Allan,S., Alfred,S., Kusuma,F.K., Clark,A., Webber,H.,\n",
      "            Russell,A.J., de Weck,A., Kile,B.T., Santulli,M., De Rosa,P.,\n",
      "            Fleuren,E.D.G., Gao,W., Wilkinson-White,L., Low,J.K.K.,\n",
      "            Mackay,J.P., Marshall,G.M., Hilton,D.J., Giorgi,F.M., Koster,J.,\n",
      "            Perini,G., Haber,M. and Norris,M.D.\n",
      "  TITLE     The transcriptional co-repressor Runx1t1 is essential for\n",
      "            MYCN-driven neuroblastoma tumorigenesis\n",
      "  JOURNAL   Nat Commun 15 (1), 5585 (2024)\n",
      "   PUBMED   38992040\n",
      "  REMARK    GeneRIF: The transcriptional co-repressor Runx1t1 is essential for\n",
      "            MYCN-driven neuroblastoma tumorigenesis.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   4  (bases 1 to 1706)\n",
      "  AUTHORS   Besancon,R., Valsesia-Wittmann,S., Locher,C., Delloye-Bourgeois,C.,\n",
      "            Furhman,L., Tutrone,G., Bertrand,C., Jallas,A.C., Garin,E. and\n",
      "            Puisieux,A.\n",
      "  TITLE     Upstream ORF affects MYCN translation depending on exon 1b\n",
      "            alternative splicing\n",
      "  JOURNAL   BMC Cancer 9, 445 (2009)\n",
      "   PUBMED   20017904\n",
      "  REMARK    GeneRIF: Existence of upstream ORF in MYCN transcripts leads to a\n",
      "            new level of MYCN regulation.\n",
      "            Publication Status: Online-Only\n",
      "REFERENCE   5  (bases 1 to 1706)\n",
      "  AUTHORS   Marcelis,C.L.M. and de Brouwer,A.P.M.\n",
      "  TITLE     Feingold Syndrome 1\n",
      "  JOURNAL   (in) Adam MP, Feldman J, Mirzaa GM, Pagon RA, Wallace SE and\n",
      "            Amemiya A (Eds.);\n",
      "            GENEREVIEWS(R);\n",
      "            (1993)\n",
      "   PUBMED   20301770\n",
      "REFERENCE   6  (bases 1 to 1706)\n",
      "  AUTHORS   Scott,D.A.\n",
      "  TITLE     Esophageal Atresia / Tracheoesophageal Fistula Overview - RETIRED\n",
      "            CHAPTER, FOR HISTORICAL REFERENCE ONLY\n",
      "  JOURNAL   (in) Adam MP, Feldman J, Mirzaa GM, Pagon RA, Wallace SE and\n",
      "            Amemiya A (Eds.);\n",
      "            GENEREVIEWS(R);\n",
      "            (1993)\n",
      "   PUBMED   20301753\n",
      "REFERENCE   7  (bases 1 to 1706)\n",
      "  AUTHORS   Hagiwara,T., Nakaya,K., Nakamura,Y., Nakajima,H., Nishimura,S. and\n",
      "            Taya,Y.\n",
      "  TITLE     Specific phosphorylation of the acidic central region of the N-myc\n",
      "            protein by casein kinase II\n",
      "  JOURNAL   Eur J Biochem 209 (3), 945-950 (1992)\n",
      "   PUBMED   1425701\n",
      "REFERENCE   8  (bases 1 to 1706)\n",
      "  AUTHORS   Armstrong,B.C. and Krystal,G.W.\n",
      "  TITLE     Isolation and characterization of complementary DNA for N-cym, a\n",
      "            gene encoded by the DNA strand opposite to N-myc\n",
      "  JOURNAL   Cell Growth Differ 3 (6), 385-390 (1992)\n",
      "   PUBMED   1419902\n",
      "REFERENCE   9  (bases 1 to 1706)\n",
      "  AUTHORS   Fougerousse,F., Meloni,R., Roudaut,C. and Beckmann,J.S.\n",
      "  TITLE     Dinucleotide repeat polymorphism at the human hemoglobin alpha-1\n",
      "            pseudo-gene (HBAP1)\n",
      "  JOURNAL   Nucleic Acids Res 20 (5), 1165 (1992)\n",
      "   PUBMED   1549498\n",
      "REFERENCE   10 (bases 1 to 1706)\n",
      "  AUTHORS   Stanton,L.W. and Bishop,J.M.\n",
      "  TITLE     Alternative processing of RNA transcribed from NMYC\n",
      "  JOURNAL   Mol Cell Biol 7 (12), 4266-4272 (1987)\n",
      "   PUBMED   3437890\n",
      "COMMENT     REVIEWED REFSEQ: This record has been curated by NCBI staff. The\n",
      "            reference sequence was derived from AC010145.10 and BC002712.2.\n",
      "            \n",
      "            On May 31, 2019 this sequence version replaced NM_001293231.1.\n",
      "            \n",
      "            Summary: This gene is a member of the MYC family and encodes a\n",
      "            protein with a basic helix-loop-helix (bHLH) domain. This protein\n",
      "            is located in the nucleus and must dimerize with another bHLH\n",
      "            protein in order to bind DNA. Amplification of this gene is\n",
      "            associated with a variety of tumors, most notably neuroblastomas.\n",
      "            Multiple alternatively spliced transcript variants encoding\n",
      "            different isoforms have been found for this gene. [provided by\n",
      "            RefSeq, Jun 2014].\n",
      "            \n",
      "            Transcript Variant: This variant (3) lacks segment 1b and exon 2,\n",
      "            which results in an upstream AUG start codon, compared to variant\n",
      "            1. The resulting isoform (2) has a shorter and distinct N-terminus,\n",
      "            compared to isoform 1.\n",
      "            \n",
      "            Sequence Note: This RefSeq record was created from transcript and\n",
      "            genomic sequence data to make the sequence consistent with the\n",
      "            reference genome assembly. The genomic coordinates used for the\n",
      "            transcript record were based on transcript alignments.\n",
      "            \n",
      "            Publication Note:  This RefSeq record includes a subset of the\n",
      "            publications that are available for this gene. Please see the Gene\n",
      "            record to access additional publications.\n",
      "            \n",
      "            ##Evidence-Data-START##\n",
      "            Transcript exon combination :: BE382888.1, BI194921.1 [ECO:0000332]\n",
      "            ##Evidence-Data-END##\n",
      "PRIMARY     REFSEQ_SPAN         PRIMARY_IDENTIFIER PRIMARY_SPAN        COMP\n",
      "            1-194               AC010145.10        65895-66088\n",
      "            195-1706            BC002712.2         932-2443\n",
      "FEATURES             Location/Qualifiers\n",
      "     source          1..1706\n",
      "                     /organism=\"Homo sapiens\"\n",
      "                     /mol_type=\"mRNA\"\n",
      "                     /db_xref=\"taxon:9606\"\n",
      "                     /chromosome=\"2\"\n",
      "                     /map=\"2p24.3\"\n",
      "     gene            1..1706\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /note=\"MYCN proto-oncogene, bHLH transcription factor\"\n",
      "                     /db_xref=\"GeneID:4613\"\n",
      "                     /db_xref=\"HGNC:HGNC:7559\"\n",
      "                     /db_xref=\"MIM:164840\"\n",
      "     exon            1..194\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /inference=\"alignment:Splign:2.1.0\"\n",
      "     CDS             38..799\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /note=\"isoform 2 is encoded by transcript variant 3;\n",
      "                     neuroblastoma MYC oncogene; N-myc proto-oncogene protein;\n",
      "                     oncogene NMYC; neuroblastoma-derived v-myc avian\n",
      "                     myelocytomatosis viral related oncogene; class E basic\n",
      "                     helix-loop-helix protein 37; v-myc avian myelocytomatosis\n",
      "                     viral oncogene neuroblastoma derived homolog\"\n",
      "                     /codon_start=1\n",
      "                     /product=\"N-myc proto-oncogene protein isoform 2\"\n",
      "                     /protein_id=\"NP_001280160.1\"\n",
      "                     /db_xref=\"CCDS:CCDS86823.1\"\n",
      "                     /db_xref=\"GeneID:4613\"\n",
      "                     /db_xref=\"HGNC:HGNC:7559\"\n",
      "                     /db_xref=\"MIM:164840\"\n",
      "                     /translation=\"MRGAPGNCVGAEQALARRKRAQTVAIRGHPRPPGPPGDTRAESP\n",
      "                     PDPLQSAGDDEDDEEEDEEEEIDVVTVEKRRSSSNTKAVTTFTITVRPKNAALGPGRA\n",
      "                     QSSELILKRCLPIHQQHNYAAPSPYVESEDAPPQKKIKSEASPRPLKSVIPPKAKSLS\n",
      "                     PRNSDSEDSERRRNHNILERQRRNDLRSSFLTLRDHVPELVKNEKAAKVVILKKATEY\n",
      "                     VHSLQAEEHQLLLEKEKLQARQQQLLKKIEHARTC\"\n",
      "     exon            195..1706\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /inference=\"alignment:Splign:2.1.0\"\n",
      "     regulatory      1686..1691\n",
      "                     /regulatory_class=\"polyA_signal_sequence\"\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /note=\"hexamer: AATAAA\"\n",
      "     polyA_site      1706\n",
      "                     /gene=\"MYCN\"\n",
      "                     /gene_synonym=\"bHLHe37; FGLDS1; MODED; MPAPA; MYCNsORF;\n",
      "                     MYCNsPEP; N-myc; NMYC; ODED\"\n",
      "                     /note=\"major polyA site\"\n",
      "ORIGIN      \n",
      "        1 aggctgtgac agtcatctgt ctggacgcgc tgggtggatg cggggggctc ctgggaactg\n",
      "       61 tgttggagcc gagcaagcgc tagccaggcg caagcgcgca cagactgtag ccatccgagg\n",
      "      121 acacccccgc ccccccggcc cacccggaga cacccgcgca gaatcgcctc cggatcccct\n",
      "      181 gcagtcggcg ggagatgatg aagatgatga agaggaagat gaagaggaag aaatcgacgt\n",
      "      241 ggtcactgtg gagaagcggc gttcctcctc caacaccaag gctgtcacca cattcaccat\n",
      "      301 cactgtgcgt cccaagaacg cagccctggg tcccgggagg gctcagtcca gcgagctgat\n",
      "      361 cctcaaacga tgccttccca tccaccagca gcacaactat gccgccccct ctccctacgt\n",
      "      421 ggagagtgag gatgcacccc cacagaagaa gataaagagc gaggcgtccc cacgtccgct\n",
      "      481 caagagtgtc atccccccaa aggctaagag cttgagcccc cgaaactctg actcggagga\n",
      "      541 cagtgagcgt cgcagaaacc acaacatcct ggagcgccag cgccgcaacg accttcggtc\n",
      "      601 cagctttctc acgctcaggg accacgtgcc ggagttggta aagaatgaga aggccgccaa\n",
      "      661 ggtggtcatt ttgaaaaagg ccactgagta tgtccactcc ctccaggccg aggagcacca\n",
      "      721 gcttttgctg gaaaaggaaa aattgcaggc aagacagcag cagttgctaa agaaaattga\n",
      "      781 acacgctcgg acttgctaga cgcttctcaa aactggacag tcactgccac tttgcacatt\n",
      "      841 ttgatttttt ttttaaacaa acattgtgtt gacattaaga atgttggttt actttcaaat\n",
      "      901 cggtcccctg tcgagttcgg ctctgggtgg gcagtaggac caccagtgtg gggttctgct\n",
      "      961 gggaccttgg agagcctgca tcccaggatg ctgggtggcc ctgcagcctc ctccacctca\n",
      "     1021 cctccatgac agcgctaaac gttggtgacg gttgggagcc tctggggctg ttgaagtcac\n",
      "     1081 cttgtgtgtt ccaagtttcc aaacaacaga aagtcattcc ttctttttaa aatggtgctt\n",
      "     1141 aagttccagc agatgccaca taaggggttt gccatttgat acccctgggg aacatttctg\n",
      "     1201 taaataccat tgacacatcc gccttttgta tacatcctgg gtaatgagag gtggcttttg\n",
      "     1261 cggccagtat tagactggaa gttcatacct aagtactgta ataatacctc aatgtttgag\n",
      "     1321 gagcatgttt tgtatacaaa tatattgtta atctctgtta tgtactgtac taattcttac\n",
      "     1381 actgcctgta tactttagta tgacgctgat acataactaa atttgatact tatattttcg\n",
      "     1441 tatgaaaatg agttgtgaaa gttttgagta gatattactt tatcactttt tgaactaaga\n",
      "     1501 aacttttgta aagaaattta ctatatatat atgccttttt cctagcctgt ttcttcctgt\n",
      "     1561 taatgtattt gttcatgttt ggtgcataga actgggtaaa tgcaaagttc tgtgtttaat\n",
      "     1621 ttcttcaaaa tgtatatatt tagtgctgca tcttatagca ctttgaaata cctcatgttt\n",
      "     1681 atgaaaataa atagcttaaa attaaa\n",
      "//\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# improved version with params (more readable)\n",
    "gene_id = 1676324825\n",
    "url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi\"\n",
    "params = {\"db\": \"nucleotide\", \"id\": gene_id, \"rettype\": \"gb\"}\n",
    "\n",
    "response = requests.get(url, params=params)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efbfb646-bd80-44a9-bdbf-5ff5e1f6e04d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# more complete example (check for status code)\n",
    "def fetch_gene_sequence(gene_id):\n",
    "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id={gene_id}&rettype=fasta&retmode=text\"\n",
    "    response = requests.get(url)\n",
    "    if response.status_code == 200:\n",
    "        return response.text\n",
    "    else:\n",
    "        print(f\"Failed to retrieve data: {response.status_code}\")\n",
    "\n",
    "\n",
    "# Usage\n",
    "gene_id = \"NC_000852\"  # or 9790228\n",
    "gene_sequence = fetch_gene_sequence(gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b6ed303-5d39-42e7-8dfb-a970367a822c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_sequence[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e9d2a6e-addd-4acc-8a89-7dccf743278a",
   "metadata": {},
   "source": [
    "We'll see later how we could improve on the check for status code."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bc79843-0363-41ce-aa33-ac885cfa3522",
   "metadata": {},
   "source": [
    "## Scrape data from web pages\n",
    "\n",
    "Here we use `requests` to get the HTML source code of a web page (`www.example.com`), and then we extract links from it using [Beautiful Soup](http://www.crummy.com/software/BeautifulSoup/), a great Python library for pulling data out of HTML and XML files.\n",
    "\n",
    "If needed, install Beautiful Soup by `conda install -y bs4` or `pip install -y bs4`.\n",
    "\n",
    "You can do that without leaving Jupyter Lab: just start a code cell with `!`, immediately followed by the command, which will be run in the shell.\n",
    "\n",
    "`!conda install -y bs4` or `!pip install -y bs4`\n",
    "\n",
    "*When you use this interface, which is non-interactive, remember to apply the `-y` argument, i.e., to set any confirmation values to 'yes' automatically.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1de20d85-8e1f-4e65-b6a7-5111a3c64ff1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8a75fcf-2852-484b-a8d1-dd5ddb9baa56",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bs4 import BeautifulSoup\n",
    "\n",
    "url = \"http://www.example.com/\"\n",
    "# make a request\n",
    "response = requests.get(url)\n",
    "# display the response\n",
    "print(response.content)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "063bb4be-00a3-49bd-96ec-bad8c71d282c",
   "metadata": {},
   "source": [
    "The content looks like HTML: let's access it using the `text` attribute"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac9fad02-19e4-4f7a-9a5b-3e4235190465",
   "metadata": {},
   "outputs": [],
   "source": [
    "html = response.text\n",
    "print(html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb6851b8-8665-4bca-af48-1b8907b7e543",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create soup\n",
    "soup = BeautifulSoup(html, \"html.parser\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6665b394-625e-4018-816f-b9ff62ccdaf3",
   "metadata": {},
   "source": [
    "The `soup` variable is a data structure representing a parsed HTML document."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "902de808-ba46-4be1-a2db-6a914503373a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract page title from the HTML\n",
    "print(f\"Found title: {soup.title.text}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9747f4c8-58b1-4a1b-ad9d-f3cb7e1b03b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract links (hrefs) from the HTML\n",
    "for link in soup.find_all(\"a\"):\n",
    "    print(f\"Found link: {link.get('href')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90b9508d-119c-48ca-bad5-6acd55d62859",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract all text from the HTML\n",
    "print(f\"Found text: {soup.get_text()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "664d2887-77b2-41ce-85b3-91804a5d0030",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Interact with FTP servers\n",
    "\n",
    "FTP servers may look old-fashioned next to fancy HTTP requests, API queries, and web scraping. But a lot of bioinformatics-related services still use them: for example, reference genome builds on the UCSC hub, NCBI Short Read Archive (SRA), Gene Expression Omnibus (GEO) files, and more.\n",
    "\n",
    "Python provides tools to deal with FTP: as an example, we'll see here two of them.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e47bb6ab-e076-49b3-a104-83d3c0977010",
   "metadata": {},
   "source": [
    "## `urllib.request`\n",
    "\n",
    "This solution is based on native Python's URL handling, which supports FTP URLs out of the box.\n",
    "\n",
    "General syntax:\n",
    "\n",
    "```python\n",
    "from urllib.request import urlopen\n",
    "\n",
    "with open(urlopen('ftp://username:password@server/path/to/file') as resource:\n",
    "    data = resource.read()\n",
    "    with open('file', 'wb') as f:\n",
    "        f.write(data)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be9e436d-3f63-45c4-a32e-5776c13d22cd",
   "metadata": {},
   "source": [
    "Let's download a FASTA sequence for the human mitochondrial chromosome from the [UCSC FTP server](https://genome.ucsc.edu/goldenPath/help/ftp.html).\n",
    "\n",
    "From the above help page, we take note of the FTP server we need to use: `ftp://hgdownload.soe.ucsc.edu`\n",
    "\n",
    "To obtain the file path and location, we may browse [UCSC downloads page](http://hgdownload.soe.ucsc.edu/) and look for the file of interest. Let's go ahead and look for the hg38 assembly > sequence data by chromosome > scroll down to chrM.fa.gz\n",
    "\n",
    "The HTTP URL of the file is `https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chrM.fa.gz`, so the path relative to the server is `goldenPath/hg38/chromosomes/chr22.fa.gz`.\n",
    "\n",
    "Let's put all this together and download the file with Python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "8f82ae07-f343-4b43-9b5b-68705218b881",
   "metadata": {},
   "outputs": [],
   "source": [
    "from urllib.request import urlopen\n",
    "\n",
    "SERVER = \"ftp://hgdownload.soe.ucsc.edu\"\n",
    "path = \"goldenPath/hg38/chromosomes\"\n",
    "filename = \"chrM.fa.gz\"\n",
    "\n",
    "with urlopen(f'{SERVER}/{path}/{filename}') as resource:\n",
    "    data = resource.read()\n",
    "    with open(filename, 'wb') as f:\n",
    "        f.write(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ea0d612-f6a5-4468-a22c-805260dce6b3",
   "metadata": {},
   "source": [
    "## `ftplib`\n",
    "\n",
    "You can use `ftplib` if you want more control over the FTP file actions: for example, change directories, list files, or download many files in a loop."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "999c661c-709e-40d0-ba00-a51f67f6432f",
   "metadata": {},
   "outputs": [
    {
     "ename": "EOFError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEOFError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m SERVER \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhgdownload.soe.ucsc.edu\u001b[39m\u001b[38;5;124m\"\u001b[39m  \u001b[38;5;66;03m# no \"ftp://\" here\u001b[39;00m\n\u001b[1;32m      4\u001b[0m path \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mgoldenPath\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 6\u001b[0m ftp \u001b[38;5;241m=\u001b[39m \u001b[43mFTP\u001b[49m\u001b[43m(\u001b[49m\u001b[43mSERVER\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m ftp\u001b[38;5;241m.\u001b[39mlogin()  \u001b[38;5;66;03m# anonymous by default\u001b[39;00m\n\u001b[1;32m      8\u001b[0m ftp\u001b[38;5;241m.\u001b[39mcwd(path)\n",
      "File \u001b[0;32m~/miniforge3/envs/pyadv2025/lib/python3.12/ftplib.py:121\u001b[0m, in \u001b[0;36mFTP.__init__\u001b[0;34m(self, host, user, passwd, acct, timeout, source_address, encoding)\u001b[0m\n\u001b[1;32m    119\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout \u001b[38;5;241m=\u001b[39m timeout\n\u001b[1;32m    120\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m host:\n\u001b[0;32m--> 121\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43mhost\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    122\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m user:\n\u001b[1;32m    123\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlogin(user, passwd, acct)\n",
      "File \u001b[0;32m~/miniforge3/envs/pyadv2025/lib/python3.12/ftplib.py:162\u001b[0m, in \u001b[0;36mFTP.connect\u001b[0;34m(self, host, port, timeout, source_address)\u001b[0m\n\u001b[1;32m    160\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maf \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock\u001b[38;5;241m.\u001b[39mfamily\n\u001b[1;32m    161\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfile \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock\u001b[38;5;241m.\u001b[39mmakefile(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124m'\u001b[39m, encoding\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mencoding)\n\u001b[0;32m--> 162\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwelcome \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetresp\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwelcome\n",
      "File \u001b[0;32m~/miniforge3/envs/pyadv2025/lib/python3.12/ftplib.py:244\u001b[0m, in \u001b[0;36mFTP.getresp\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    243\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgetresp\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 244\u001b[0m     resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetmultiline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    245\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdebugging:\n\u001b[1;32m    246\u001b[0m         \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m*resp*\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msanitize(resp))\n",
      "File \u001b[0;32m~/miniforge3/envs/pyadv2025/lib/python3.12/ftplib.py:230\u001b[0m, in \u001b[0;36mFTP.getmultiline\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    229\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mgetmultiline\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 230\u001b[0m     line \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mgetline\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    231\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m line[\u001b[38;5;241m3\u001b[39m:\u001b[38;5;241m4\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m    232\u001b[0m         code \u001b[38;5;241m=\u001b[39m line[:\u001b[38;5;241m3\u001b[39m]\n",
      "File \u001b[0;32m~/miniforge3/envs/pyadv2025/lib/python3.12/ftplib.py:218\u001b[0m, in \u001b[0;36mFTP.getline\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    216\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m*get*\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msanitize(line))\n\u001b[1;32m    217\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m line:\n\u001b[0;32m--> 218\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mEOFError\u001b[39;00m\n\u001b[1;32m    219\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m line[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m2\u001b[39m:] \u001b[38;5;241m==\u001b[39m CRLF:\n\u001b[1;32m    220\u001b[0m     line \u001b[38;5;241m=\u001b[39m line[:\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m2\u001b[39m]\n",
      "\u001b[0;31mEOFError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "from ftplib import FTP\n",
    "\n",
    "SERVER = \"hgdownload.soe.ucsc.edu\"  # no \"ftp://\" here\n",
    "path = \"goldenPath\"\n",
    "\n",
    "ftp = FTP(SERVER)\n",
    "ftp.login()  # anonymous by default\n",
    "ftp.cwd(path)\n",
    "files = ftp.nlst()  # a Python list of the directory content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00caf397-77a2-4550-91e4-39d4bfd45048",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted(files)[1:20]\n",
    "\n",
    "# for fn in files:\n",
    "#     print(\" \", fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "719a3b12-1f2c-4038-a3f1-d79459d2b6e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "path = \"hg38/chromosomes\"  # relative to the previous path!\n",
    "ftp.cwd(path)\n",
    "files = ftp.nlst()\n",
    "sorted(files)[1:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3ebe497-32eb-40e6-b27b-dd77ef95727d",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"chrM.fa.gz\"\n",
    "\n",
    "with open(filename, \"wb\") as fout:\n",
    "    ftp.retrbinary(f\"RETR {filename}\", fout.write)\n",
    "\n",
    "ftp.quit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a3be27b-e845-4293-8e0f-90a4834cb10f",
   "metadata": {},
   "source": [
    "As another example, we'll use `ftplib` to list and download available files from NCBI's Gene Expression Omnibus (GEO) repository.\n",
    "\n",
    "The data set of interest has a GEO accession number GSE186032 (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE186032) and involves gene expression profiling by RNA-seq to identify changes in mouse breast tumor tissues resistant to anti-PD-L1.\n",
    "\n",
    "The main files of a GEO data set are:\n",
    "- a SOFT formatted file\n",
    "- a MINiML formatted file\n",
    "- a Series Matrix file\n",
    "\n",
    "In addition to these, supplementary files may be available. They can be, for example, clinical information, READMEs, raw read counts, etc. and they can be in different formats.\n",
    "\n",
    "In our case, there is one supplementary file consisting of RSEM estimated counts.\n",
    "\n",
    "By hovering on `ftp` in the data set's page on GEO, we could simply copy the FTP URL of the file:\n",
    "\n",
    "`ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE186nnn/GSE186032/suppl/GSE186032%5FEMT6%2DRSEM%2Destimated%2Dcounts.txt.gz`\n",
    "\n",
    "But we are more interested in finding it programmatically!\n",
    "\n",
    "GEO has a convention for forming its FTP URLs. The base URL for any accession is:\n",
    "\n",
    "`ftp://ftp.ncbi.nlm.nih.gov/geo/series/<PREFIX>nnn/<ACCESSION>/<FILETYPE>`\n",
    "\n",
    "- where `<ACCESSION>` is the full GEO accession, i.e., GSE186032\n",
    "- `<PREFIX>` is a substring of the accession minus its last 3 digits, i.e., GSE186\n",
    "- `<FILETYPE>` is one of `soft`, `matrix`, `miniml`, or `suppl`.\n",
    "\n",
    "With this information, we can connect to the host, change into the correct directory, list the available supplementary files, and grab the one we want."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a964439-f31a-4d0d-8908-af18114bbc3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ACCESSION = \"GSE186032\"\n",
    "PREFIX = ACCESSION[:-3]\n",
    "FTP_HOST = \"ftp.ncbi.nlm.nih.gov\"\n",
    "FTP_PATH = f\"geo/series/{PREFIX}nnn/{ACCESSION}/suppl\"\n",
    "\n",
    "ftp = FTP(FTP_HOST)\n",
    "ftp.login()\n",
    "ftp.cwd(FTP_PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9e64eab-52b9-436b-bfa5-6ad602badd00",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppl_files = ftp.nlst()\n",
    "print(f\"Available supplementary files for accession {ACCESSION}:\")\n",
    "\n",
    "for fn in suppl_files:\n",
    "    print(\" \", fn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51171715-8404-4c3d-8156-3c02132a79a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# there is only 1 file\n",
    "my_file = suppl_files[0]\n",
    "\n",
    "with open(my_file, \"wb\") as fout:\n",
    "    ftp.retrbinary(f\"RETR {my_file}\", fout.write)\n",
    "\n",
    "ftp.quit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e48fa533",
   "metadata": {},
   "source": [
    "The file is now on our local computer, ready for downstream analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fedddf9-c7af-465a-a0fb-3f143d36e940",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Assertions\n",
    "\n",
    "Assertions are useful sanity checks in Python: they allow you to check if some specific condition is True.\n",
    "They are particularly used while debugging your code: when the code goes into production, they are typically switched off.\n",
    "\n",
    "General syntax:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d7846d5-8541-467b-a43c-a1639412d35e",
   "metadata": {},
   "source": [
    "```\n",
    "assert condition[, assertion_message]\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81a66781-c876-451a-8cc7-607311f3056b",
   "metadata": {},
   "source": [
    "The `assertion_message` is optional but highly recommended: after all, it should be helpful!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51233e6a-de3a-432f-b4c4-fa92ba1d8c88",
   "metadata": {},
   "outputs": [],
   "source": [
    "number = 42\n",
    "assert number > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92f0d7bf-3507-43ca-a823-cd4ad3467495",
   "metadata": {},
   "outputs": [],
   "source": [
    "number = -42\n",
    "assert number > 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8030762-15e1-4408-a22b-fd4e80a716b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "number = -42\n",
    "assert number > 0, f\"Positive number requested, got {number}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "466b6bbb-64b2-48c2-9ebe-9fff1e089be3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# comparison\n",
    "assert 3>2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c584659-2019-4af9-bee0-7ca32831bdd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# membership\n",
    "numbers = [1, 3, 5, 7, 9]\n",
    "assert 10 in numbers, \"Input number not found in list\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e50715c1-94fd-4873-860d-22ac40e5ed16",
   "metadata": {},
   "outputs": [],
   "source": [
    "# type check\n",
    "x = 1\n",
    "y = x\n",
    "assert x is y\n",
    "assert x is not y"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02809ed3-9c44-433c-9760-116491ae1d79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# identity\n",
    "number = 42.0\n",
    "assert isinstance(number, int), f\"{number} is not integer\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "834cdcfe-7498-46a6-8b4d-1aa6a08ce918",
   "metadata": {},
   "source": [
    "Let's check whether an input nucleotide sequence contains only valid nucleotide letters: 'A', 'C', 'G', or 'T'.\n",
    "As it often happens, this could be accomplished in many alternative ways. I choose here to check if each of the letters belongs to the set of characters 'ACGT'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4666f651-6899-4ae9-b43c-8175377ef0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "nucleotide_sequence = \"ACCGAGTACG\"\n",
    "assert all(base in 'ACGT' for base in nucleotide_sequence), f\"Invalid nucleotide detected in {nucleotide_sequence}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "873b1d32-3ded-4b09-88ff-0915e6e1f27c",
   "metadata": {},
   "outputs": [],
   "source": [
    "nucleotide_sequence = \"ACCGBGTACG\"\n",
    "assert all(base in 'ACGT' for base in nucleotide_sequence), f\"Invalid nucleotide detected in {nucleotide_sequence}\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f62b6ab-7344-4c7e-9918-02b559fabfe2",
   "metadata": {},
   "source": [
    "Another common application is to check the response code of a HTTP requests.\n",
    "Remember the UniProt example?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95cc4ef7-8b96-4059-8349-a38c89fc8a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_id = \"P33993\"\n",
    "url = f\"https://rest.uniprot.org/uniprotkb/{uniprot_id}.fasta\"\n",
    "response = requests.get(url)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b4757d5-2940-4782-b4b9-ceba471ed9f1",
   "metadata": {},
   "source": [
    "Suppose now you used a different endpoint, because you found it in an old script or a Stackexchange post."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb081a2c-673d-4332-a36d-b3eef46ba714",
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot_id = \"P33993\"\n",
    "url = f\"https://www.uniprot.org/uniprot/?query=id:{uniprot_id}&columns=sequence&format=tab\"\n",
    "response = requests.get(url)\n",
    "print(response.text)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "decc49bd-3741-494a-92cc-d3edfb6b3eac",
   "metadata": {},
   "source": [
    "Not the output we expected, right?\n",
    "It is not so uncommon actually: APIs change during time, and because of this we can't expect that a server responses always in the same way. We need to check its response. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ff68d7b-e332-4849-a927-f7c02f04a23a",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert response.status_code == '200', 'Something went wrong during the query: check the API'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45905dd9-74fa-40f1-a059-a0c8630e8281",
   "metadata": {},
   "source": [
    "# Exceptions\n",
    "\n",
    "Exceptions are commonly referred to as \"errors\". In Python, there are many types of exceptions, depending on the specific error."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f84b764e-dd09-4bdc-932c-1c3d71a74787",
   "metadata": {},
   "outputs": [],
   "source": [
    "fruits = [\"apple\", \"orange\", \"banana\", \"apricot\"]\n",
    "fruits[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a347591-6bce-47b6-b52d-aa29fc19fe00",
   "metadata": {},
   "outputs": [],
   "source": [
    "fruits.remove(\"tomato\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ee8e503-9755-4d6e-9b9c-8b85d7577496",
   "metadata": {},
   "outputs": [],
   "source": [
    "fruits[0] + 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a36ed48-b096-415a-902d-79b3467ce35b",
   "metadata": {},
   "source": [
    "## Exception handling\n",
    "\n",
    "Exceptions terminate the execution of your program. To avoid this, it is necessary to write code to \"handle\" the exceptions gracefully and make it possible to continue evaluating our script."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c23d4f7-5582-4db7-b606-f4d149504da3",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = 10\n",
    "b = 0\n",
    "c = a / b\n",
    "print(\"a / b = \", c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef4c5396-03e3-4085-90ee-abc4043924e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = 10\n",
    "b = 0\n",
    "try:\n",
    "    c = a/b\n",
    "    print(\"a / b = \", c)\n",
    "except:\n",
    "    print(\"Can't divide by zero.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "116061d6-0aee-4e00-b11d-f8a25daaedac",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = int(input(\"a = \"))\n",
    "b = int(input(\"b = \"))\n",
    "c = a / b\n",
    "print(\"a / b = \", c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "255efbdb-37d6-4da9-af99-6714edda1d8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    a = int(input(\"a = \"))\n",
    "    b = int(input(\"b = \"))\n",
    "    c = a / b\n",
    "    print(\"a / b = \", c)\n",
    "except ValueError:\n",
    "    print(\"Entered value must be a number.\")\n",
    "except ZeroDivisionError:\n",
    "    print(\"Can't divide by zero.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b710434e-90b7-4189-9ee5-7356e0af152c",
   "metadata": {},
   "source": [
    "Another typical situation is where we want to read from a file, but that file is actually missing!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7da7d1c-65d2-4abd-b373-4556833e2fe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"file.log\") as file:\n",
    "    read_data = file.read()\n",
    "\n",
    "print(\"Here we are\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c3c8f08-82e0-471a-b9a5-e330735e499e",
   "metadata": {},
   "source": [
    "Note that \"Here we are\" is never evaluated if the block above (in particular `open('file.log')`) throws an exception.\n",
    "\n",
    "The code can be improved by handling the situation in which a file is potentially missing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7a9a717-8392-4eac-b5f8-b37d249dde3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    with open(\"file.log\") as file:\n",
    "        read_data = file.read()\n",
    "except:\n",
    "    print(\"Could not open file.log\")\n",
    "\n",
    "print(\"Here we are\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e08188d5-b385-4651-9775-8d193b9e080d",
   "metadata": {},
   "source": [
    "It is always good practice to specify the exact exception that the `except` clause should catch: in this case, `FileNotFoundError` (see above)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "674b61b7-6208-4a14-b8c5-73a06d852e7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    with open(\"file.log\") as file:\n",
    "        read_data = file.read()\n",
    "except FileNotFoundError as nf_err:\n",
    "    print(nf_err)\n",
    "\n",
    "print(\"Here we are\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9b31026-cb12-4e47-9e8a-d6e16fb81c18",
   "metadata": {},
   "source": [
    "An example application is to be sure that a dictionary with gene expression values actually contains valid items."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfa4add2-4703-4e89-81e8-05c18d53c1ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expr = {\"Rb1\": 2.5, \"P53\": None, \"Ras\": \"missing\"}\n",
    "\n",
    "for gene, expression in gene_expr.items():\n",
    "    try:\n",
    "        value = float(expression)\n",
    "        print(f\"expression level of {gene} = {value}\")\n",
    "    except TypeError as te:\n",
    "        print(f\"Expression for {gene} is missing!\")\n",
    "    except ValueError:\n",
    "        print(f\"Invalid expression for {gene}!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb5897d5-cf93-40d0-9556-847ed2ed3bd2",
   "metadata": {},
   "source": [
    "We could use the exception handling mechanism to check if a nucleotide sequence is valid, i.e., contains only 'A', 'C', 'G', or 'T' letters. We elaborate on the previous example by creating a function `validate_sequence()` and using it in a `try...except` construct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3817ee91-fac1-4b73-b9ce-9b85431b9b33",
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate_sequence(seq):\n",
    "    assert all(base in \"ATGC\" for base in seq), \"Invalid nucleotide detected\"\n",
    "    return True\n",
    "\n",
    "\n",
    "sequence = \"ATGBC\"\n",
    "try:\n",
    "    validate_sequence(sequence)\n",
    "except AssertionError as e:\n",
    "    print(f\"Validation failed: {e}\")\n",
    "\n",
    "print(\"Here we are!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ff7f2d4-1e27-4b84-a14b-6d0339617e3f",
   "metadata": {},
   "source": [
    "Let's revisit our previous example of querying the NCBI Eutils. We got this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6839de9a-2e10-4a5c-912c-3d4e795e4dfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fetch_gene_sequence(gene_id):\n",
    "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id={gene_id}&rettype=fasta&retmode=text\"\n",
    "    response = requests.get(url)\n",
    "    if response.status_code == 200:\n",
    "        return response.text\n",
    "    else:\n",
    "        print(f\"Failed to retrieve data: {response.status_code}\")\n",
    "\n",
    "\n",
    "# Usage\n",
    "gene_id = \"NC_000852\"  # or try with an invalid ID such as \"NM_00005\"\n",
    "gene_sequence = fetch_gene_sequence(gene_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b45b036-19af-41cf-b742-549092ea47c2",
   "metadata": {},
   "source": [
    "We can now improve the code above by using a `try...except` construct, which is the preferred way to deal with this kind of situations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9fea9fe-df44-4060-8346-7c269ccce8c7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# better exception handling\n",
    "def fetch_gene_sequence(gene_id):\n",
    "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id={gene_id}&rettype=fasta\"\n",
    "    response = requests.get(url)\n",
    "    try:\n",
    "        response.raise_for_status()  # throws an HTTPError exception if something bad happens\n",
    "        return response.text\n",
    "    except requests.HTTPError:\n",
    "        print(\n",
    "            f\"Failed to retrieve sequence for accession {gene_id}: code {response.status_code}\"\n",
    "        )\n",
    "        return None\n",
    "\n",
    "\n",
    "# Usage\n",
    "gene_id = \"NC_000852\"  # or try with an invalid ID such as \"NM_00005\"\n",
    "sequence = fetch_gene_sequence(gene_id)\n",
    "if sequence:\n",
    "    print(sequence)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9af125f5-5488-46c1-9f71-e6027242c4d1",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Iterators and Generators\n",
    "\n",
    "Iterators are objects that can be iterated upon, meaning that they return one action or item at a time. To be considered an iterator, objects need to implement two methods: `__iter__()` and `__next__()`.\n",
    "\n",
    "We already know a few Python iterators: `for` loops and list comprehensions.\n",
    "\n",
    "Generators are a Pythonic implementation of iterators, without needing to explicitly implement a class with `__iter__()` and `__next__()` methods. Similarly, and most importantly, you don't need to keep track of the object's internal state. An important thing to note is that generators iterate over an object **lazily**, meaning they do not store their contents in memory.\n",
    "\n",
    "The `yield` statement controls the flow of a generator. The statement goes further to handle the state of the generator function, pausing it until it's called again, using the `next()` function on the generator object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b19fd79b-2e58-4189-80cd-ada1764cf20d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating a simple generator in Python\n",
    "def return_n_values(n):\n",
    "    num = 0\n",
    "    while num < n:\n",
    "        yield num\n",
    "        num += 1\n",
    "\n",
    "\n",
    "values = return_n_values(10)\n",
    "\n",
    "print(type(values))\n",
    "print(values)\n",
    "\n",
    "# start unpacking the generator\n",
    "print(next(values))\n",
    "print(next(values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb6104b9-3d06-49a0-b106-5a4710580424",
   "metadata": {},
   "outputs": [],
   "source": [
    "# unpacking generators with a for loop\n",
    "def return_n_values(n):\n",
    "    num = 0\n",
    "    while num < n:\n",
    "        yield num\n",
    "        num += 1\n",
    "\n",
    "\n",
    "values = return_n_values(5)\n",
    "\n",
    "for val in values:\n",
    "    print(val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8b810c2-7ebb-406d-b9a6-7cbc1224ee56",
   "metadata": {},
   "outputs": [],
   "source": [
    "# creating a generator with a for loop\n",
    "def return_n_values(n):\n",
    "    for i in range(n):\n",
    "        yield i\n",
    "\n",
    "\n",
    "values = return_n_values(3)\n",
    "\n",
    "for val in values:\n",
    "    print(val, end=\" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77d2c68-c83f-4594-9bf6-a9542bcee579",
   "metadata": {},
   "source": [
    "## Exercise: from functions to generators"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "636074c6-d2d6-4fb2-a27c-3315317c3b32",
   "metadata": {},
   "source": [
    "Rewrite as a generator the following function:\n",
    "\n",
    "```python\n",
    "def even_numbers(n):\n",
    "    evens = []\n",
    "    for i in range(n):\n",
    "        if i % 2 == 0:\n",
    "            evens.append(i)\n",
    "    return evens\n",
    "```\n",
    "\n",
    "Test both the function and the generator with `n=10`, printing their outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7510d60f-7f94-42d3-866b-49caed011fa5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e993e36-4c30-41cf-ae29-f9f68a9d188a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def even_numbers(n):\n",
    "    evens = []\n",
    "    for i in range(n):\n",
    "        if i % 2 == 0:\n",
    "            evens.append(i)\n",
    "    return evens\n",
    "\n",
    "\n",
    "evens = even_numbers(10)\n",
    "print(evens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1da2caa4-9028-4837-835c-e9e8629ea8d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def even_numbers(n):\n",
    "    for i in range(n):\n",
    "        if i % 2 == 0:\n",
    "            yield i\n",
    "\n",
    "\n",
    "evens = even_numbers(10)\n",
    "print(evens)  # generator object\n",
    "\n",
    "for even in evens:\n",
    "    print(even, end=\" \")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c76790aa-1aeb-43cf-af1c-f07659659568",
   "metadata": {},
   "source": [
    "## Exercise: codon generator\n",
    "\n",
    "Write a simple generator function that yields the codons from a given DNA sequence.\n",
    "\n",
    "Example input: \"ATGCGTATGCCTAATGATCTAAGCTAGCTGATCGATCTAGCTAGATGTAG\"\n",
    "\n",
    "Example output:\n",
    "```\n",
    "ATG\n",
    "CGT\n",
    "ATG\n",
    "CCT\n",
    "(truncated)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5d8ffe-99de-4073-a1b1-40e69009b643",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "825886ec-2d65-4d6b-91dd-eb43feb033de",
   "metadata": {},
   "outputs": [],
   "source": [
    "def codon_generator(dna_sequence):\n",
    "    for i in range(0, len(dna_sequence), 3):\n",
    "        # Ensure slicing does not go beyond the sequence length\n",
    "        if i + 3 <= len(dna_sequence):\n",
    "            yield dna_sequence[i : i + 3]\n",
    "\n",
    "\n",
    "sequence = \"ATGCGTATGCCTAATGATCTAAGCTAGCTGATCGATCTAGCTAGATGTAG\"\n",
    "\n",
    "for codon in codon_generator(sequence):\n",
    "    print(codon)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2496d90-73a6-4017-9237-d028cf778836",
   "metadata": {},
   "source": [
    "## Applications of generators\n",
    "\n",
    "Since generators are lazily evaluated, they are useful when iterating through large datasets, avoiding the need to create a duplicate of the dataset in memory. Large datasets include infinite sequences such as the Fibonacci sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85ab9f36-1c55-4664-820d-74a9467e6e66",
   "metadata": {},
   "source": [
    "### Reading large files\n",
    "\n",
    "If you have a large file that you need to process line by line, you can use a generator to read the file one line at a time, instead of reading the entire file into memory at once. This can save a lot of memory, especially if the file is very large."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4b03fa2-ee3b-426d-921b-e288561949c2",
   "metadata": {},
   "source": [
    "```python\n",
    "def read_file(filename):\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            yield line\n",
    "\n",
    "for line in read_file('large_file.txt'):\n",
    "    print(line)\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "048af1ef-714e-4d06-8f2e-f562b830bafb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# large file reader: inefficient version\n",
    "def file_reader(file_path):\n",
    "    rows = []\n",
    "    for row in open(file_path, \"r\"):\n",
    "        rows.append(row)\n",
    "    return rows\n",
    "  \n",
    "def print_row_count(file_path):\n",
    "    count = 0\n",
    "    for row in file_reader(file_path):\n",
    "        count += 1\n",
    "    print(f\"Total count is {count}\")\n",
    "\n",
    "print_row_count('data/ecoli_K12_genomic.fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6358153-4c74-4969-a7d7-40a72c93609e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# large file reader: better version\n",
    "def gen_file_reader(file_path):\n",
    "    for row in open(file_path, \"r\"):\n",
    "         yield row\n",
    "        \n",
    "def gen_print_row_count(file_path):\n",
    "    count = 0\n",
    "    for row in gen_file_reader(file_path):\n",
    "        count += 1\n",
    "    print(f\"Total count is {count}\")\n",
    "    \n",
    "gen_print_row_count('data/ecoli_K12_genomic.fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fd419a5-2b5d-4447-a54f-624ce41ba705",
   "metadata": {},
   "source": [
    "### Generating infinite sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1187964-1d7e-4d82-8552-a25487f9086a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fibonacci():\n",
    "    a, b = 0, 1\n",
    "    while True:\n",
    "        yield a\n",
    "        a, b = b, a + b\n",
    "\n",
    "\n",
    "for i, fib in enumerate(fibonacci()):\n",
    "    print(fib)\n",
    "    if i > 10:\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a31cd4d-2686-474f-a69c-6db23d918e71",
   "metadata": {},
   "outputs": [],
   "source": [
    "fib = fibonacci()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8c0509e-d51e-47fc-8c6e-0445e7e2906d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# try re-executing this code cell over and over again (CMD+ENTER or CRTL+ENTER)\n",
    "next(fib)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c6c7f58-6571-400e-88dd-f95b4abb076f",
   "metadata": {},
   "source": [
    "We can revise our Fibonacci generator with a parameter reprenting an upper bound:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbe540cc-c599-4c25-8ebe-7e9bcfe3a57f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fibonacci(nums):\n",
    "    a, b = 0, 1\n",
    "    for _ in range(nums):\n",
    "        a, b = b, a + b\n",
    "        yield a\n",
    "\n",
    "\n",
    "for fib in fibonacci(10):\n",
    "    print(fib)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d09b7f1-aeac-4e8f-8936-06b01003ae28",
   "metadata": {},
   "source": [
    "Similarly to list comprehension, there is a shorthand syntax for generators!\n",
    "\n",
    "Use the generator expression syntax to create simple generators. For example, you can use the following syntax to create a generator that yields the even numbers between 0 and n:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eafba1d3-92d4-4885-9ac3-f5db40cd652c",
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 10\n",
    "evens = (i for i in range(n) if i % 2 == 0)\n",
    "print(list(evens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02c6d923-0dd9-452c-a1b3-1745e53631d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "evens_squared = (i**2 for i in range(n) if i % 2 == 0)\n",
    "print(list(evens_squared))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "948bd110-9198-40db-aa9c-2941acbf98e7",
   "metadata": {},
   "source": [
    "### Pipelines: Filtering and transforming data\n",
    "\n",
    "Generators can be used to filter and transform data in a single step, which can make your code more readable."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e47b62fb-2123-47de-b5d9-002d1ae46ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def even_numbers(numbers):\n",
    "    for number in numbers:\n",
    "        if number % 2 == 0:\n",
    "            yield number\n",
    "\n",
    "\n",
    "def square(numbers):\n",
    "    for number in numbers:\n",
    "        yield number**2\n",
    "\n",
    "\n",
    "numbers = [1, 2, 3, 4, 5, 6]\n",
    "even_squares = square(even_numbers(numbers))\n",
    "print(list(even_squares))  # [4, 16, 36]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c36ac42-faaf-467a-99f1-a56576eb17f3",
   "metadata": {},
   "source": [
    "We can combine our `square` and (revised) `fibonacci` generators to print the sum of squares of the first N=10 numbers in the Fibonacci series."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2781e23-64dc-478b-a924-1fe3756dcac6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(square(fibonacci(10)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05dd9bef-9e55-4836-8884-c56d7d440505",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Credits\n",
    "\n",
    "Partially abridged from work by Lee Stott (MIT License), Sebastian Bassi (MIT License), and datanagy.io"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
