{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "373e4706-8b43-439c-a57b-d3ff64997219",
   "metadata": {},
   "source": [
    "# Capstone Project: Neuroblastoma gene expression data analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9e4bb2f7-6da8-49f6-b6af-f42a2839dd22",
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "import os\n",
    "import itertools\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9aef2b1",
   "metadata": {},
   "source": [
    "# Working on the metadata table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "0ce4ba5d-5021-420b-8661-275fccaacbe3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "08-Jul-2025 14:53:33 INFO GEOparse - Parsing data/GSE49711_family.soft.gz: \n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SERIES: GSE49711\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - PLATFORM: GPL17553\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205736\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205737\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205738\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205739\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205740\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205741\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205742\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205743\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205744\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205745\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205746\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205747\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205748\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205749\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205750\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205751\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205752\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205753\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205754\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205755\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205756\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205757\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205758\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205759\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205760\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205761\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205762\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205763\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205764\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205765\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205766\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205767\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205768\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205769\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205770\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205771\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205772\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205773\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205774\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205775\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205776\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205777\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205778\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205779\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205780\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205781\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205782\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205783\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205784\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205785\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205786\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205787\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205788\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205789\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205790\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205791\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205792\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205793\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205794\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205795\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205796\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205797\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205798\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205799\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205800\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205801\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205802\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205803\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205804\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205805\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205806\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205807\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205808\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205809\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205810\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205811\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205812\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205813\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205814\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205815\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205816\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205817\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205818\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205819\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205820\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205821\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205822\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205823\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205824\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205825\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205826\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205827\n",
      "08-Jul-2025 14:53:33 DEBUG GEOparse - SAMPLE: GSM1205828\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205829\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205830\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205831\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205832\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205833\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205834\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205835\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205836\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205837\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205838\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205839\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205840\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205841\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205842\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205843\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205844\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205845\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205846\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205847\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205848\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205849\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205850\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205851\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205852\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205853\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205854\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205855\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205856\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205857\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205858\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205859\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205860\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205861\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205862\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205863\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205864\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205865\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205866\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205867\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205868\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205869\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205870\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205871\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205872\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205873\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205874\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205875\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205876\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205877\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205878\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205879\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205880\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205881\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205882\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205883\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205884\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205885\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205886\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205887\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205888\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205889\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205890\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205891\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205892\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205893\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205894\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205895\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205896\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205897\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205898\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205899\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205900\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205901\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205902\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205903\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205904\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205905\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205906\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205907\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205908\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205909\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205910\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205911\n",
      "08-Jul-2025 14:53:34 DEBUG GEOparse - SAMPLE: GSM1205912\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205913\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205914\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205915\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205916\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205917\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205918\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205919\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205920\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205921\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205922\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205923\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205924\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205925\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205926\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205927\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205928\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205929\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205930\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205931\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205932\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205933\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205934\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205935\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205936\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205937\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205938\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205939\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205940\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205941\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205942\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205943\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205944\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205945\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205946\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205947\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205948\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205949\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205950\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205951\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205952\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205953\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205954\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205955\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205956\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205957\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205958\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205959\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205960\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205961\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205962\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205963\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205964\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205965\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205966\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205967\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205968\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205969\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205970\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205971\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205972\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205973\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205974\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205975\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205976\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205977\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205978\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205979\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205980\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205981\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205982\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205983\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205984\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205985\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205986\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205987\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205988\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205989\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205990\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205991\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205992\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205993\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205994\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205995\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205996\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205997\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205998\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1205999\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206000\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206001\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206002\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206003\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206004\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206005\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206006\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206007\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206008\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206009\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206010\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206011\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206012\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206013\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206014\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206015\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206016\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206017\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206018\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206019\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206020\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206021\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206022\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206023\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206024\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206025\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206026\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206027\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206028\n",
      "08-Jul-2025 14:53:35 DEBUG GEOparse - SAMPLE: GSM1206029\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206030\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206031\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206032\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206033\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206034\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206035\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206036\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206037\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206038\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206039\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206040\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206041\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206042\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206043\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206044\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206045\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206046\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206047\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206048\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206049\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206050\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206051\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206052\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206053\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206054\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206055\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206056\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206057\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206058\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206059\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206060\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206061\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206062\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206063\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206064\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206065\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206066\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206067\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206068\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206069\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206070\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206071\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206072\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206073\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206074\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206075\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206076\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206077\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206078\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206079\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206080\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206081\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206082\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206083\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206084\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206085\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206086\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206087\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206088\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206089\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206090\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206091\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206092\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206093\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206094\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206095\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206096\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206097\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206098\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206099\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206100\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206101\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206102\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206103\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206104\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206105\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206106\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206107\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206108\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206109\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206110\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206111\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206112\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206113\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206114\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206115\n",
      "08-Jul-2025 14:53:36 DEBUG GEOparse - SAMPLE: GSM1206116\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206117\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206118\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206119\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206120\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206121\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206122\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206123\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206124\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206125\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206126\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206127\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206128\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206129\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206130\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206131\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206132\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206133\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206134\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206135\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206136\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206137\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206138\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206139\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206140\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206141\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206142\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206143\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206144\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206145\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206146\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206147\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206148\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206149\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206150\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206151\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206152\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206153\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206154\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206155\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206156\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206157\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206158\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206159\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206160\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206161\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206162\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206163\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206164\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206165\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206166\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206167\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206168\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206169\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206170\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206171\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206172\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206173\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206174\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206175\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206176\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206177\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206178\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206179\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206180\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206181\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206182\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206183\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206184\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206185\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206186\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206187\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206188\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206189\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206190\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206191\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206192\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206193\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206194\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206195\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206196\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206197\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206198\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206199\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206200\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206201\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206202\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206203\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206204\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206205\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206206\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206207\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206208\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206209\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206210\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206211\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206212\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206213\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206214\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206215\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206216\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206217\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206218\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206219\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206220\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206221\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206222\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206223\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206224\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206225\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206226\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206227\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206228\n",
      "08-Jul-2025 14:53:37 DEBUG GEOparse - SAMPLE: GSM1206229\n",
      "08-Jul-2025 14:53:38 DEBUG GEOparse - SAMPLE: GSM1206230\n",
      "08-Jul-2025 14:53:38 DEBUG GEOparse - SAMPLE: GSM1206231\n",
      "08-Jul-2025 14:53:38 DEBUG GEOparse - SAMPLE: GSM1206232\n",
      "08-Jul-2025 14:53:38 DEBUG GEOparse - SAMPLE: GSM1206233\n"
     ]
    }
   ],
   "source": [
    "gse_id = \"GSE49711\"\n",
    "geo_dir = \"data\"\n",
    "os.makedirs(geo_dir, exist_ok=True)\n",
    "# if you need to download from GEO:\n",
    "# gse = GEOparse.get_GEO(geo=gse_id, destdir=geo_dir)\n",
    "# if you already have the SOFT file locally:\n",
    "gse = GEOparse.get_GEO(filepath=os.path.join(geo_dir, gse_id + \"_family.soft.gz\"))\n",
    "meta_df = gse.phenotype_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ba605dae-1dbd-4d3f-b77d-68db3f1027e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['title', 'tissue', 'dataset', 'Sex', 'age_at_diagnosis', 'mycn_status',\n",
       "       'high_risk', 'inss_stage', 'class_label', 'progression',\n",
       "       'death_from_disease'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "16913bd6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>tissue</th>\n",
       "      <th>dataset</th>\n",
       "      <th>Sex</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>mycn_status</th>\n",
       "      <th>high_risk</th>\n",
       "      <th>inss_stage</th>\n",
       "      <th>class_label</th>\n",
       "      <th>progression</th>\n",
       "      <th>death_from_disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1205736</th>\n",
       "      <td>SEQC_NB001</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>987</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205737</th>\n",
       "      <td>SEQC_NB002</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>M</td>\n",
       "      <td>1808</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205738</th>\n",
       "      <td>SEQC_NB003</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205739</th>\n",
       "      <td>SEQC_NB004</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>F</td>\n",
       "      <td>335</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205740</th>\n",
       "      <td>SEQC_NB005</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>536</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 title         tissue dataset Sex age_at_diagnosis  \\\n",
       "GSM1205736  SEQC_NB001  neuroblastoma       1   M              987   \n",
       "GSM1205737  SEQC_NB002  neuroblastoma       2   M             1808   \n",
       "GSM1205738  SEQC_NB003  neuroblastoma       1   F              625   \n",
       "GSM1205739  SEQC_NB004  neuroblastoma       2   F              335   \n",
       "GSM1205740  SEQC_NB005  neuroblastoma       1   F              536   \n",
       "\n",
       "           mycn_status high_risk inss_stage class_label progression  \\\n",
       "GSM1205736           0         1          4           1           1   \n",
       "GSM1205737           0         1          4           1           1   \n",
       "GSM1205738           1         1          4           1           1   \n",
       "GSM1205739           1         1          2         N/A           0   \n",
       "GSM1205740           1         1          4           1           1   \n",
       "\n",
       "           death_from_disease  \n",
       "GSM1205736                  1  \n",
       "GSM1205737                  1  \n",
       "GSM1205738                  1  \n",
       "GSM1205739                  0  \n",
       "GSM1205740                  1  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# optionally, drop columns\n",
    "meta_cols_to_keep = meta_df.columns[(meta_df.columns == 'title') | (meta_df.columns.str.startswith('characteristics'))]\n",
    "meta_df = meta_df[meta_cols_to_keep]\n",
    "\n",
    "# cleanup column names a bit\n",
    "# e.g. \"characteristics_ch1.3.age at diagnosis\" -> \"age_at_diagnosis\"\n",
    "newcols = meta_df.columns.str.replace(r'^.*\\.', '', regex=True)\n",
    "newcols\n",
    "newcols = newcols.str.replace(' ', '_')\n",
    "meta_df.columns = newcols\n",
    "\n",
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "64a2f2cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>tissue</th>\n",
       "      <th>dataset</th>\n",
       "      <th>Sex</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>mycn_status</th>\n",
       "      <th>high_risk</th>\n",
       "      <th>inss_stage</th>\n",
       "      <th>class_label</th>\n",
       "      <th>progression</th>\n",
       "      <th>death_from_disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1205736</th>\n",
       "      <td>SEQC_NB001</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>987</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205737</th>\n",
       "      <td>SEQC_NB002</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>M</td>\n",
       "      <td>1808</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205738</th>\n",
       "      <td>SEQC_NB003</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205739</th>\n",
       "      <td>SEQC_NB004</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>F</td>\n",
       "      <td>335</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205740</th>\n",
       "      <td>SEQC_NB005</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>536</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 title         tissue dataset Sex age_at_diagnosis  \\\n",
       "GSM1205736  SEQC_NB001  neuroblastoma       1   M              987   \n",
       "GSM1205737  SEQC_NB002  neuroblastoma       2   M             1808   \n",
       "GSM1205738  SEQC_NB003  neuroblastoma       1   F              625   \n",
       "GSM1205739  SEQC_NB004  neuroblastoma       2   F              335   \n",
       "GSM1205740  SEQC_NB005  neuroblastoma       1   F              536   \n",
       "\n",
       "           mycn_status high_risk inss_stage class_label progression  \\\n",
       "GSM1205736           0         1          4           1           1   \n",
       "GSM1205737           0         1          4           1           1   \n",
       "GSM1205738           1         1          4           1           1   \n",
       "GSM1205739           1         1          2         N/A           0   \n",
       "GSM1205740           1         1          4           1           1   \n",
       "\n",
       "           death_from_disease  \n",
       "GSM1205736                  1  \n",
       "GSM1205737                  1  \n",
       "GSM1205738                  1  \n",
       "GSM1205739                  0  \n",
       "GSM1205740                  1  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# makin ga copy of the original meta_df, to avoid SettingWithCopyWarning\n",
    "meta_tr = meta_df.copy()\n",
    "meta_tr.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "2c117f95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['title', 'tissue', 'dataset', 'Sex', 'age_at_diagnosis', 'mycn_status',\n",
       "       'high_risk', 'inss_stage', 'class_label', 'progression',\n",
       "       'death_from_disease'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_tr.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "a518a339",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>tissue</th>\n",
       "      <th>dataset</th>\n",
       "      <th>Sex</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>mycn_status</th>\n",
       "      <th>high_risk</th>\n",
       "      <th>inss_stage</th>\n",
       "      <th>class_label</th>\n",
       "      <th>progression</th>\n",
       "      <th>death_from_disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1205736</th>\n",
       "      <td>SEQC_NB001</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>987</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205737</th>\n",
       "      <td>SEQC_NB002</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>M</td>\n",
       "      <td>1808</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205738</th>\n",
       "      <td>SEQC_NB003</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>625</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205739</th>\n",
       "      <td>SEQC_NB004</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>F</td>\n",
       "      <td>335</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>N/A</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205740</th>\n",
       "      <td>SEQC_NB005</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>536</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 title         tissue dataset Sex age_at_diagnosis  \\\n",
       "GSM1205736  SEQC_NB001  neuroblastoma       1   M              987   \n",
       "GSM1205737  SEQC_NB002  neuroblastoma       2   M             1808   \n",
       "GSM1205738  SEQC_NB003  neuroblastoma       1   F              625   \n",
       "GSM1205739  SEQC_NB004  neuroblastoma       2   F              335   \n",
       "GSM1205740  SEQC_NB005  neuroblastoma       1   F              536   \n",
       "\n",
       "           mycn_status high_risk inss_stage class_label progression  \\\n",
       "GSM1205736           0         1          4           1           1   \n",
       "GSM1205737           0         1          4           1           1   \n",
       "GSM1205738           1         1          4           1           1   \n",
       "GSM1205739           1         1          2         N/A           0   \n",
       "GSM1205740           1         1          4           1           1   \n",
       "\n",
       "           death_from_disease  \n",
       "GSM1205736                  1  \n",
       "GSM1205737                  1  \n",
       "GSM1205738                  1  \n",
       "GSM1205739                  0  \n",
       "GSM1205740                  1  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_tr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "4911b3aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class_label\n",
       "Neither        226\n",
       "Favorable      181\n",
       "Unfavorable     91\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#clean up table - remove NAs from class_label\n",
    "meta_tr[\"class_label\"] = meta_tr[\"class_label\"].map({\"N/A\": \"Neither\", \"0\": \"Favorable\", \"1\": \"Unfavorable\"})\n",
    "\n",
    "meta_tr.class_label.value_counts()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "acaa9cc5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mycn_status\n",
       "Normal       401\n",
       "Amplified     92\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# redefine mycn_status\n",
    "meta_tr[\"mycn_status\"] = meta_tr[\"mycn_status\"].map({\"0\": \"Normal\", \"1\": \"Amplified\"})\n",
    "\n",
    "meta_tr[\"mycn_status\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "49102a81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "title                 0\n",
       "tissue                0\n",
       "dataset               0\n",
       "Sex                   0\n",
       "age_at_diagnosis      0\n",
       "mycn_status           5\n",
       "high_risk             0\n",
       "inss_stage            0\n",
       "class_label           0\n",
       "progression           0\n",
       "death_from_disease    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_tr.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "895d84d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_tr = meta_tr.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "5115c29f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dataset\n",
       "2    247\n",
       "1    246\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_tr.dataset.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "101d4a71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>tissue</th>\n",
       "      <th>dataset</th>\n",
       "      <th>Sex</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>mycn_status</th>\n",
       "      <th>high_risk</th>\n",
       "      <th>inss_stage</th>\n",
       "      <th>class_label</th>\n",
       "      <th>progression</th>\n",
       "      <th>death_from_disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1205736</th>\n",
       "      <td>SEQC_NB001</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>987</td>\n",
       "      <td>Normal</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Unfavorable</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205737</th>\n",
       "      <td>SEQC_NB002</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>M</td>\n",
       "      <td>1808</td>\n",
       "      <td>Normal</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Unfavorable</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205738</th>\n",
       "      <td>SEQC_NB003</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>625</td>\n",
       "      <td>Amplified</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Unfavorable</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205739</th>\n",
       "      <td>SEQC_NB004</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>F</td>\n",
       "      <td>335</td>\n",
       "      <td>Amplified</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>Neither</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 title         tissue dataset Sex age_at_diagnosis  \\\n",
       "GSM1205736  SEQC_NB001  neuroblastoma       1   M              987   \n",
       "GSM1205737  SEQC_NB002  neuroblastoma       2   M             1808   \n",
       "GSM1205738  SEQC_NB003  neuroblastoma       1   F              625   \n",
       "GSM1205739  SEQC_NB004  neuroblastoma       2   F              335   \n",
       "\n",
       "           mycn_status high_risk inss_stage  class_label progression  \\\n",
       "GSM1205736      Normal         1          4  Unfavorable           1   \n",
       "GSM1205737      Normal         1          4  Unfavorable           1   \n",
       "GSM1205738   Amplified         1          4  Unfavorable           1   \n",
       "GSM1205739   Amplified         1          2      Neither           0   \n",
       "\n",
       "           death_from_disease  \n",
       "GSM1205736                  1  \n",
       "GSM1205737                  1  \n",
       "GSM1205738                  1  \n",
       "GSM1205739                  0  "
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_tr.head(4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac97b9d8",
   "metadata": {},
   "source": [
    "# Working on the expression data table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7261376d",
   "metadata": {},
   "source": [
    "## download the table and read in as an object df_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc07f35b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DOWNLOADING EXPRESSION TABLE\n",
    "'''url = \"https://ftp.ncbi.nlm.nih.gov/geo/series/GSE49nnn/GSE49711/matrix/GSE49711_SEQC_NB_MAV_G_log2.20121127.txt.gz\"\n",
    "output_file = \"GSE49711_SEQC_NB_MAV_G_log2.20121127.txt.gz\"\n",
    "\n",
    "response = requests.get(url)\n",
    "with open(output_file, \"wb\") as:\n",
    "    f.write(response.content)\n",
    "\n",
    "print(\"Download complete.\")'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b311e7c6-c898-47b3-a87a-262371b17181",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_4492/2812194724.py:1: DtypeWarning: Columns (1,2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_expression = pd.read_csv(\"/home/gracie/Downloads/day4-20250702T073025Z-1-001/data/GSE49711_SEQC_NB_MAV_G_log2.20121127.txt\", sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "df_expression = pd.read_csv(\"/home/gracie/Downloads/day4-20250702T073025Z-1-001/data/GSE49711_SEQC_NB_MAV_G_log2.20121127.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "48c72f36-383a-4fc4-9441-c070d46a048f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#Gene</th>\n",
       "      <th>#NCBI GeneId</th>\n",
       "      <th>#RefSeq transcript Id</th>\n",
       "      <th>#Chromosome</th>\n",
       "      <th>#Strand</th>\n",
       "      <th>#from base</th>\n",
       "      <th>#to base</th>\n",
       "      <th>#Title</th>\n",
       "      <th>#Measured object</th>\n",
       "      <th>SEQC_NB001</th>\n",
       "      <th>...</th>\n",
       "      <th>SEQC_NB489</th>\n",
       "      <th>SEQC_NB490</th>\n",
       "      <th>SEQC_NB491</th>\n",
       "      <th>SEQC_NB492</th>\n",
       "      <th>SEQC_NB493</th>\n",
       "      <th>SEQC_NB494</th>\n",
       "      <th>SEQC_NB495</th>\n",
       "      <th>SEQC_NB496</th>\n",
       "      <th>SEQC_NB497</th>\n",
       "      <th>SEQC_NB498</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALB</td>\n",
       "      <td>213</td>\n",
       "      <td>NM_000477.5</td>\n",
       "      <td>4</td>\n",
       "      <td>+</td>\n",
       "      <td>74249329</td>\n",
       "      <td>74287139</td>\n",
       "      <td>complex locus ALB, encoding albumin.</td>\n",
       "      <td>ALB:Gene_AceView</td>\n",
       "      <td>9.29</td>\n",
       "      <td>...</td>\n",
       "      <td>9.89</td>\n",
       "      <td>6.48</td>\n",
       "      <td>7.73</td>\n",
       "      <td>9.02</td>\n",
       "      <td>8.47</td>\n",
       "      <td>7.78</td>\n",
       "      <td>8.58</td>\n",
       "      <td>8.41</td>\n",
       "      <td>8.62</td>\n",
       "      <td>10.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD24L4.1</td>\n",
       "      <td>938;100133941</td>\n",
       "      <td>NM_013230.2</td>\n",
       "      <td>Y</td>\n",
       "      <td>-</td>\n",
       "      <td>21152409</td>\n",
       "      <td>21154677</td>\n",
       "      <td>gene CD24L4.1, encoding CD24 molecule-like 4 p...</td>\n",
       "      <td>CD24L4.1:Gene_AceView</td>\n",
       "      <td>18.82</td>\n",
       "      <td>...</td>\n",
       "      <td>19.99</td>\n",
       "      <td>20.17</td>\n",
       "      <td>18.94</td>\n",
       "      <td>20.42</td>\n",
       "      <td>20.75</td>\n",
       "      <td>20.86</td>\n",
       "      <td>20.57</td>\n",
       "      <td>21.04</td>\n",
       "      <td>20.13</td>\n",
       "      <td>20.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RPS11</td>\n",
       "      <td>6205</td>\n",
       "      <td>NM_001015.4</td>\n",
       "      <td>19</td>\n",
       "      <td>+</td>\n",
       "      <td>49999634</td>\n",
       "      <td>50002944</td>\n",
       "      <td>ribosomal protein S11</td>\n",
       "      <td>RPS11:Gene_RefSeq</td>\n",
       "      <td>21.17</td>\n",
       "      <td>...</td>\n",
       "      <td>22.98</td>\n",
       "      <td>23.31</td>\n",
       "      <td>21.14</td>\n",
       "      <td>22.85</td>\n",
       "      <td>20.08</td>\n",
       "      <td>20.91</td>\n",
       "      <td>20.67</td>\n",
       "      <td>21.10</td>\n",
       "      <td>21.04</td>\n",
       "      <td>20.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RPS18</td>\n",
       "      <td>6222</td>\n",
       "      <td>NM_022551.2</td>\n",
       "      <td>6</td>\n",
       "      <td>+</td>\n",
       "      <td>33239799</td>\n",
       "      <td>33244290</td>\n",
       "      <td>gene RPS18, encoding ribosomal protein S18.</td>\n",
       "      <td>RPS18:Gene_AceView</td>\n",
       "      <td>20.90</td>\n",
       "      <td>...</td>\n",
       "      <td>22.28</td>\n",
       "      <td>22.87</td>\n",
       "      <td>20.93</td>\n",
       "      <td>22.39</td>\n",
       "      <td>20.71</td>\n",
       "      <td>20.95</td>\n",
       "      <td>21.03</td>\n",
       "      <td>21.66</td>\n",
       "      <td>20.49</td>\n",
       "      <td>21.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C5orf13</td>\n",
       "      <td>9315</td>\n",
       "      <td>NM_001142474.1;NM_001142475.1;NM_001142476.1;N...</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>110998323</td>\n",
       "      <td>111312627</td>\n",
       "      <td>gene C5orf13, encoding chromosome 5 open readi...</td>\n",
       "      <td>C5orf13:Gene_AceView</td>\n",
       "      <td>20.02</td>\n",
       "      <td>...</td>\n",
       "      <td>21.55</td>\n",
       "      <td>21.04</td>\n",
       "      <td>20.57</td>\n",
       "      <td>21.06</td>\n",
       "      <td>22.16</td>\n",
       "      <td>21.98</td>\n",
       "      <td>21.56</td>\n",
       "      <td>21.80</td>\n",
       "      <td>22.06</td>\n",
       "      <td>21.31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  507 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      #Gene   #NCBI GeneId                              #RefSeq transcript Id  \\\n",
       "0       ALB            213                                        NM_000477.5   \n",
       "1  CD24L4.1  938;100133941                                        NM_013230.2   \n",
       "2     RPS11           6205                                        NM_001015.4   \n",
       "3     RPS18           6222                                        NM_022551.2   \n",
       "4   C5orf13           9315  NM_001142474.1;NM_001142475.1;NM_001142476.1;N...   \n",
       "\n",
       "  #Chromosome #Strand  #from base   #to base  \\\n",
       "0           4       +    74249329   74287139   \n",
       "1           Y       -    21152409   21154677   \n",
       "2          19       +    49999634   50002944   \n",
       "3           6       +    33239799   33244290   \n",
       "4           5       -   110998323  111312627   \n",
       "\n",
       "                                              #Title       #Measured object  \\\n",
       "0               complex locus ALB, encoding albumin.       ALB:Gene_AceView   \n",
       "1  gene CD24L4.1, encoding CD24 molecule-like 4 p...  CD24L4.1:Gene_AceView   \n",
       "2                              ribosomal protein S11      RPS11:Gene_RefSeq   \n",
       "3        gene RPS18, encoding ribosomal protein S18.     RPS18:Gene_AceView   \n",
       "4  gene C5orf13, encoding chromosome 5 open readi...   C5orf13:Gene_AceView   \n",
       "\n",
       "   SEQC_NB001  ...  SEQC_NB489  SEQC_NB490  SEQC_NB491  SEQC_NB492  \\\n",
       "0        9.29  ...        9.89        6.48        7.73        9.02   \n",
       "1       18.82  ...       19.99       20.17       18.94       20.42   \n",
       "2       21.17  ...       22.98       23.31       21.14       22.85   \n",
       "3       20.90  ...       22.28       22.87       20.93       22.39   \n",
       "4       20.02  ...       21.55       21.04       20.57       21.06   \n",
       "\n",
       "   SEQC_NB493  SEQC_NB494  SEQC_NB495  SEQC_NB496  SEQC_NB497  SEQC_NB498  \n",
       "0        8.47        7.78        8.58        8.41        8.62       10.20  \n",
       "1       20.75       20.86       20.57       21.04       20.13       20.02  \n",
       "2       20.08       20.91       20.67       21.10       21.04       20.82  \n",
       "3       20.71       20.95       21.03       21.66       20.49       21.11  \n",
       "4       22.16       21.98       21.56       21.80       22.06       21.31  \n",
       "\n",
       "[5 rows x 507 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76fb710f",
   "metadata": {},
   "source": [
    "## Clean up the table and remove duplicated genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "2672bd76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>SEQC_NB001</th>\n",
       "      <th>SEQC_NB002</th>\n",
       "      <th>SEQC_NB003</th>\n",
       "      <th>SEQC_NB004</th>\n",
       "      <th>SEQC_NB005</th>\n",
       "      <th>SEQC_NB006</th>\n",
       "      <th>SEQC_NB007</th>\n",
       "      <th>SEQC_NB008</th>\n",
       "      <th>SEQC_NB009</th>\n",
       "      <th>...</th>\n",
       "      <th>SEQC_NB489</th>\n",
       "      <th>SEQC_NB490</th>\n",
       "      <th>SEQC_NB491</th>\n",
       "      <th>SEQC_NB492</th>\n",
       "      <th>SEQC_NB493</th>\n",
       "      <th>SEQC_NB494</th>\n",
       "      <th>SEQC_NB495</th>\n",
       "      <th>SEQC_NB496</th>\n",
       "      <th>SEQC_NB497</th>\n",
       "      <th>SEQC_NB498</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALB</td>\n",
       "      <td>9.29</td>\n",
       "      <td>7.99</td>\n",
       "      <td>9.25</td>\n",
       "      <td>9.35</td>\n",
       "      <td>8.99</td>\n",
       "      <td>9.26</td>\n",
       "      <td>8.15</td>\n",
       "      <td>9.54</td>\n",
       "      <td>10.72</td>\n",
       "      <td>...</td>\n",
       "      <td>9.89</td>\n",
       "      <td>6.48</td>\n",
       "      <td>7.73</td>\n",
       "      <td>9.02</td>\n",
       "      <td>8.47</td>\n",
       "      <td>7.78</td>\n",
       "      <td>8.58</td>\n",
       "      <td>8.41</td>\n",
       "      <td>8.62</td>\n",
       "      <td>10.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD24L4.1</td>\n",
       "      <td>18.82</td>\n",
       "      <td>19.80</td>\n",
       "      <td>20.25</td>\n",
       "      <td>20.44</td>\n",
       "      <td>20.09</td>\n",
       "      <td>20.38</td>\n",
       "      <td>19.75</td>\n",
       "      <td>20.21</td>\n",
       "      <td>19.66</td>\n",
       "      <td>...</td>\n",
       "      <td>19.99</td>\n",
       "      <td>20.17</td>\n",
       "      <td>18.94</td>\n",
       "      <td>20.42</td>\n",
       "      <td>20.75</td>\n",
       "      <td>20.86</td>\n",
       "      <td>20.57</td>\n",
       "      <td>21.04</td>\n",
       "      <td>20.13</td>\n",
       "      <td>20.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RPS11</td>\n",
       "      <td>21.17</td>\n",
       "      <td>20.42</td>\n",
       "      <td>22.44</td>\n",
       "      <td>21.22</td>\n",
       "      <td>22.09</td>\n",
       "      <td>21.23</td>\n",
       "      <td>21.00</td>\n",
       "      <td>22.60</td>\n",
       "      <td>21.43</td>\n",
       "      <td>...</td>\n",
       "      <td>22.98</td>\n",
       "      <td>23.31</td>\n",
       "      <td>21.14</td>\n",
       "      <td>22.85</td>\n",
       "      <td>20.08</td>\n",
       "      <td>20.91</td>\n",
       "      <td>20.67</td>\n",
       "      <td>21.10</td>\n",
       "      <td>21.04</td>\n",
       "      <td>20.82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RPS18</td>\n",
       "      <td>20.90</td>\n",
       "      <td>20.36</td>\n",
       "      <td>22.00</td>\n",
       "      <td>21.08</td>\n",
       "      <td>21.71</td>\n",
       "      <td>20.74</td>\n",
       "      <td>20.88</td>\n",
       "      <td>22.55</td>\n",
       "      <td>21.63</td>\n",
       "      <td>...</td>\n",
       "      <td>22.28</td>\n",
       "      <td>22.87</td>\n",
       "      <td>20.93</td>\n",
       "      <td>22.39</td>\n",
       "      <td>20.71</td>\n",
       "      <td>20.95</td>\n",
       "      <td>21.03</td>\n",
       "      <td>21.66</td>\n",
       "      <td>20.49</td>\n",
       "      <td>21.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C5orf13</td>\n",
       "      <td>20.02</td>\n",
       "      <td>21.70</td>\n",
       "      <td>21.05</td>\n",
       "      <td>21.89</td>\n",
       "      <td>21.65</td>\n",
       "      <td>21.14</td>\n",
       "      <td>21.48</td>\n",
       "      <td>20.64</td>\n",
       "      <td>21.81</td>\n",
       "      <td>...</td>\n",
       "      <td>21.55</td>\n",
       "      <td>21.04</td>\n",
       "      <td>20.57</td>\n",
       "      <td>21.06</td>\n",
       "      <td>22.16</td>\n",
       "      <td>21.98</td>\n",
       "      <td>21.56</td>\n",
       "      <td>21.80</td>\n",
       "      <td>22.06</td>\n",
       "      <td>21.31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  499 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ID  SEQC_NB001  SEQC_NB002  SEQC_NB003  SEQC_NB004  SEQC_NB005  \\\n",
       "0       ALB        9.29        7.99        9.25        9.35        8.99   \n",
       "1  CD24L4.1       18.82       19.80       20.25       20.44       20.09   \n",
       "2     RPS11       21.17       20.42       22.44       21.22       22.09   \n",
       "3     RPS18       20.90       20.36       22.00       21.08       21.71   \n",
       "4   C5orf13       20.02       21.70       21.05       21.89       21.65   \n",
       "\n",
       "   SEQC_NB006  SEQC_NB007  SEQC_NB008  SEQC_NB009  ...  SEQC_NB489  \\\n",
       "0        9.26        8.15        9.54       10.72  ...        9.89   \n",
       "1       20.38       19.75       20.21       19.66  ...       19.99   \n",
       "2       21.23       21.00       22.60       21.43  ...       22.98   \n",
       "3       20.74       20.88       22.55       21.63  ...       22.28   \n",
       "4       21.14       21.48       20.64       21.81  ...       21.55   \n",
       "\n",
       "   SEQC_NB490  SEQC_NB491  SEQC_NB492  SEQC_NB493  SEQC_NB494  SEQC_NB495  \\\n",
       "0        6.48        7.73        9.02        8.47        7.78        8.58   \n",
       "1       20.17       18.94       20.42       20.75       20.86       20.57   \n",
       "2       23.31       21.14       22.85       20.08       20.91       20.67   \n",
       "3       22.87       20.93       22.39       20.71       20.95       21.03   \n",
       "4       21.04       20.57       21.06       22.16       21.98       21.56   \n",
       "\n",
       "   SEQC_NB496  SEQC_NB497  SEQC_NB498  \n",
       "0        8.41        8.62       10.20  \n",
       "1       21.04       20.13       20.02  \n",
       "2       21.10       21.04       20.82  \n",
       "3       21.66       20.49       21.11  \n",
       "4       21.80       22.06       21.31  \n",
       "\n",
       "[5 rows x 499 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# keep only genes with a NCBI gene ID\n",
    "df = df_expression.loc[~df_expression[\"#NCBI GeneId\"].isna() & ~df_expression['#RefSeq transcript Id'].isna()]\n",
    "# select relevant columns from expression df before merging with sample metadata\n",
    "expr = df[ df.columns[df.columns.str.contains('#Gene') | df.columns.str.startswith(\"SEQC_\")] ].rename(columns={\"#Gene\": \"ID\"})\n",
    "expr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0952a116",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>SEQC_NB001</th>\n",
       "      <th>SEQC_NB002</th>\n",
       "      <th>SEQC_NB003</th>\n",
       "      <th>SEQC_NB004</th>\n",
       "      <th>SEQC_NB005</th>\n",
       "      <th>SEQC_NB006</th>\n",
       "      <th>SEQC_NB007</th>\n",
       "      <th>SEQC_NB008</th>\n",
       "      <th>SEQC_NB009</th>\n",
       "      <th>...</th>\n",
       "      <th>SEQC_NB489</th>\n",
       "      <th>SEQC_NB490</th>\n",
       "      <th>SEQC_NB491</th>\n",
       "      <th>SEQC_NB492</th>\n",
       "      <th>SEQC_NB493</th>\n",
       "      <th>SEQC_NB494</th>\n",
       "      <th>SEQC_NB495</th>\n",
       "      <th>SEQC_NB496</th>\n",
       "      <th>SEQC_NB497</th>\n",
       "      <th>SEQC_NB498</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>INS-IGF2</td>\n",
       "      <td>15.67</td>\n",
       "      <td>16.07</td>\n",
       "      <td>15.25</td>\n",
       "      <td>14.51</td>\n",
       "      <td>18.72</td>\n",
       "      <td>17.67</td>\n",
       "      <td>21.72</td>\n",
       "      <td>15.94</td>\n",
       "      <td>16.77</td>\n",
       "      <td>...</td>\n",
       "      <td>15.37</td>\n",
       "      <td>19.42</td>\n",
       "      <td>13.16</td>\n",
       "      <td>15.23</td>\n",
       "      <td>17.46</td>\n",
       "      <td>18.49</td>\n",
       "      <td>16.32</td>\n",
       "      <td>18.20</td>\n",
       "      <td>15.28</td>\n",
       "      <td>17.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>RPS21</td>\n",
       "      <td>19.16</td>\n",
       "      <td>18.34</td>\n",
       "      <td>20.40</td>\n",
       "      <td>18.89</td>\n",
       "      <td>20.23</td>\n",
       "      <td>18.95</td>\n",
       "      <td>18.81</td>\n",
       "      <td>20.53</td>\n",
       "      <td>20.37</td>\n",
       "      <td>...</td>\n",
       "      <td>20.70</td>\n",
       "      <td>20.91</td>\n",
       "      <td>20.09</td>\n",
       "      <td>20.69</td>\n",
       "      <td>18.10</td>\n",
       "      <td>18.71</td>\n",
       "      <td>19.20</td>\n",
       "      <td>19.47</td>\n",
       "      <td>18.52</td>\n",
       "      <td>19.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>RPS21</td>\n",
       "      <td>19.25</td>\n",
       "      <td>18.39</td>\n",
       "      <td>20.49</td>\n",
       "      <td>18.96</td>\n",
       "      <td>20.31</td>\n",
       "      <td>19.03</td>\n",
       "      <td>18.89</td>\n",
       "      <td>20.59</td>\n",
       "      <td>20.45</td>\n",
       "      <td>...</td>\n",
       "      <td>20.80</td>\n",
       "      <td>21.01</td>\n",
       "      <td>20.17</td>\n",
       "      <td>20.80</td>\n",
       "      <td>18.18</td>\n",
       "      <td>18.82</td>\n",
       "      <td>19.32</td>\n",
       "      <td>19.52</td>\n",
       "      <td>18.60</td>\n",
       "      <td>19.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>INS-IGF2</td>\n",
       "      <td>14.64</td>\n",
       "      <td>14.95</td>\n",
       "      <td>14.06</td>\n",
       "      <td>13.28</td>\n",
       "      <td>17.44</td>\n",
       "      <td>16.52</td>\n",
       "      <td>20.60</td>\n",
       "      <td>14.73</td>\n",
       "      <td>15.59</td>\n",
       "      <td>...</td>\n",
       "      <td>14.29</td>\n",
       "      <td>18.29</td>\n",
       "      <td>11.99</td>\n",
       "      <td>14.02</td>\n",
       "      <td>16.45</td>\n",
       "      <td>17.47</td>\n",
       "      <td>15.26</td>\n",
       "      <td>17.13</td>\n",
       "      <td>14.44</td>\n",
       "      <td>16.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>256</th>\n",
       "      <td>EIF4A1</td>\n",
       "      <td>17.79</td>\n",
       "      <td>18.22</td>\n",
       "      <td>19.36</td>\n",
       "      <td>18.56</td>\n",
       "      <td>19.12</td>\n",
       "      <td>19.16</td>\n",
       "      <td>17.96</td>\n",
       "      <td>19.44</td>\n",
       "      <td>19.17</td>\n",
       "      <td>...</td>\n",
       "      <td>19.49</td>\n",
       "      <td>19.73</td>\n",
       "      <td>18.88</td>\n",
       "      <td>19.33</td>\n",
       "      <td>18.51</td>\n",
       "      <td>18.61</td>\n",
       "      <td>18.60</td>\n",
       "      <td>18.91</td>\n",
       "      <td>18.55</td>\n",
       "      <td>17.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56218</th>\n",
       "      <td>LOC728728</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56427</th>\n",
       "      <td>OR1D5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3.85</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>6.93</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56428</th>\n",
       "      <td>OR1D5</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>4.44</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56787</th>\n",
       "      <td>SFTPA1</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>5.38</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57119</th>\n",
       "      <td>TMPRSS11E</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>374 rows  499 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID  SEQC_NB001  SEQC_NB002  SEQC_NB003  SEQC_NB004  SEQC_NB005  \\\n",
       "66      INS-IGF2       15.67       16.07       15.25       14.51       18.72   \n",
       "105        RPS21       19.16       18.34       20.40       18.89       20.23   \n",
       "106        RPS21       19.25       18.39       20.49       18.96       20.31   \n",
       "167     INS-IGF2       14.64       14.95       14.06       13.28       17.44   \n",
       "256       EIF4A1       17.79       18.22       19.36       18.56       19.12   \n",
       "...          ...         ...         ...         ...         ...         ...   \n",
       "56218  LOC728728        0.00        0.00        0.00        0.00        0.00   \n",
       "56427      OR1D5        0.00        0.00        0.00        0.00        0.00   \n",
       "56428      OR1D5        0.00        0.00        0.00        0.00        0.00   \n",
       "56787     SFTPA1        0.00        0.00        0.00        0.00        0.00   \n",
       "57119  TMPRSS11E        0.00        0.00        0.00        0.00        0.00   \n",
       "\n",
       "       SEQC_NB006  SEQC_NB007  SEQC_NB008  SEQC_NB009  ...  SEQC_NB489  \\\n",
       "66          17.67       21.72       15.94       16.77  ...       15.37   \n",
       "105         18.95       18.81       20.53       20.37  ...       20.70   \n",
       "106         19.03       18.89       20.59       20.45  ...       20.80   \n",
       "167         16.52       20.60       14.73       15.59  ...       14.29   \n",
       "256         19.16       17.96       19.44       19.17  ...       19.49   \n",
       "...           ...         ...         ...         ...  ...         ...   \n",
       "56218        0.00        0.00        0.00        0.00  ...        0.00   \n",
       "56427        0.00        0.00        0.00        3.85  ...        0.00   \n",
       "56428        0.00        0.00        0.00        0.00  ...        0.00   \n",
       "56787        0.00        0.00        0.00        0.00  ...        0.00   \n",
       "57119        0.00        0.00        0.00        0.00  ...        0.00   \n",
       "\n",
       "       SEQC_NB490  SEQC_NB491  SEQC_NB492  SEQC_NB493  SEQC_NB494  SEQC_NB495  \\\n",
       "66          19.42       13.16       15.23       17.46       18.49       16.32   \n",
       "105         20.91       20.09       20.69       18.10       18.71       19.20   \n",
       "106         21.01       20.17       20.80       18.18       18.82       19.32   \n",
       "167         18.29       11.99       14.02       16.45       17.47       15.26   \n",
       "256         19.73       18.88       19.33       18.51       18.61       18.60   \n",
       "...           ...         ...         ...         ...         ...         ...   \n",
       "56218        0.00        0.00        0.00        0.00        0.00        0.00   \n",
       "56427        0.00        6.93        0.00        0.00        0.00        0.00   \n",
       "56428        0.00        4.44        0.00        0.00        0.00        0.00   \n",
       "56787        0.00        0.00        0.00        0.00        0.00        5.38   \n",
       "57119        0.00        0.00        0.00        0.00        0.00        0.00   \n",
       "\n",
       "       SEQC_NB496  SEQC_NB497  SEQC_NB498  \n",
       "66          18.20       15.28       17.28  \n",
       "105         19.47       18.52       19.02  \n",
       "106         19.52       18.60       19.13  \n",
       "167         17.13       14.44       16.34  \n",
       "256         18.91       18.55       17.95  \n",
       "...           ...         ...         ...  \n",
       "56218        0.00        0.00        0.00  \n",
       "56427        0.00        0.00        0.00  \n",
       "56428        0.00        0.00        0.00  \n",
       "56787        0.16        0.00        0.00  \n",
       "57119        0.00        0.00        0.00  \n",
       "\n",
       "[374 rows x 499 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check for duplicated gene names...\n",
    "dup_id = expr.loc[expr.ID.duplicated(), :].ID\n",
    "\n",
    "# check the expression levels of these dup genes\n",
    "expr.loc[expr[\"ID\"].isin(dup_id)]\n",
    "# expression values of dup genes in the same samples are not always identical\n",
    "# for the sake of simplicity, remove dup genes using ~expr.ID.duplicated()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0e59cb33",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "before removing dup: (22861, 499)\n",
      "after removing dup: (22674, 499)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print(f\"before removing dup: {expr.shape}\")\n",
    "expr = expr.loc[~expr.ID.duplicated(), :].copy()\n",
    "print(f\"after removing dup: {expr.shape}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70d166aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ID</th>\n",
       "      <th>ALB</th>\n",
       "      <th>CD24L4.1</th>\n",
       "      <th>RPS11</th>\n",
       "      <th>RPS18</th>\n",
       "      <th>C5orf13</th>\n",
       "      <th>CCT2</th>\n",
       "      <th>COL1A1</th>\n",
       "      <th>DDX1</th>\n",
       "      <th>EEF1A1</th>\n",
       "      <th>FLT3LG_</th>\n",
       "      <th>...</th>\n",
       "      <th>USP17L5</th>\n",
       "      <th>USP17L7</th>\n",
       "      <th>USP17L8</th>\n",
       "      <th>VTRNA2</th>\n",
       "      <th>WFDC11</th>\n",
       "      <th>WFDC9</th>\n",
       "      <th>XAGE2</th>\n",
       "      <th>XAGE2B</th>\n",
       "      <th>ZFATAS</th>\n",
       "      <th>ZP4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SEQC_NB001</th>\n",
       "      <td>9.29</td>\n",
       "      <td>18.82</td>\n",
       "      <td>21.17</td>\n",
       "      <td>20.90</td>\n",
       "      <td>20.02</td>\n",
       "      <td>16.31</td>\n",
       "      <td>18.60</td>\n",
       "      <td>15.73</td>\n",
       "      <td>21.71</td>\n",
       "      <td>20.02</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.74</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SEQC_NB002</th>\n",
       "      <td>7.99</td>\n",
       "      <td>19.80</td>\n",
       "      <td>20.42</td>\n",
       "      <td>20.36</td>\n",
       "      <td>21.70</td>\n",
       "      <td>16.74</td>\n",
       "      <td>18.76</td>\n",
       "      <td>15.62</td>\n",
       "      <td>21.08</td>\n",
       "      <td>19.23</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SEQC_NB003</th>\n",
       "      <td>9.25</td>\n",
       "      <td>20.25</td>\n",
       "      <td>22.44</td>\n",
       "      <td>22.00</td>\n",
       "      <td>21.05</td>\n",
       "      <td>17.06</td>\n",
       "      <td>19.39</td>\n",
       "      <td>22.84</td>\n",
       "      <td>22.72</td>\n",
       "      <td>21.26</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SEQC_NB004</th>\n",
       "      <td>9.35</td>\n",
       "      <td>20.44</td>\n",
       "      <td>21.22</td>\n",
       "      <td>21.08</td>\n",
       "      <td>21.89</td>\n",
       "      <td>16.83</td>\n",
       "      <td>20.87</td>\n",
       "      <td>15.81</td>\n",
       "      <td>22.18</td>\n",
       "      <td>20.02</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SEQC_NB005</th>\n",
       "      <td>8.99</td>\n",
       "      <td>20.09</td>\n",
       "      <td>22.09</td>\n",
       "      <td>21.71</td>\n",
       "      <td>21.65</td>\n",
       "      <td>16.85</td>\n",
       "      <td>23.02</td>\n",
       "      <td>15.79</td>\n",
       "      <td>22.24</td>\n",
       "      <td>20.75</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  22674 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "ID           ALB  CD24L4.1  RPS11  RPS18  C5orf13   CCT2  COL1A1   DDX1  \\\n",
       "SEQC_NB001  9.29     18.82  21.17  20.90    20.02  16.31   18.60  15.73   \n",
       "SEQC_NB002  7.99     19.80  20.42  20.36    21.70  16.74   18.76  15.62   \n",
       "SEQC_NB003  9.25     20.25  22.44  22.00    21.05  17.06   19.39  22.84   \n",
       "SEQC_NB004  9.35     20.44  21.22  21.08    21.89  16.83   20.87  15.81   \n",
       "SEQC_NB005  8.99     20.09  22.09  21.71    21.65  16.85   23.02  15.79   \n",
       "\n",
       "ID          EEF1A1  FLT3LG_  ...  USP17L5  USP17L7  USP17L8  VTRNA2  WFDC11  \\\n",
       "SEQC_NB001   21.71    20.02  ...      0.0      0.0      0.0     0.0    4.74   \n",
       "SEQC_NB002   21.08    19.23  ...      0.0      0.0      0.0     0.0    2.98   \n",
       "SEQC_NB003   22.72    21.26  ...      0.0      0.0      0.0     0.0    0.00   \n",
       "SEQC_NB004   22.18    20.02  ...      0.0      0.0      0.0     0.0    5.14   \n",
       "SEQC_NB005   22.24    20.75  ...      0.0      0.0      0.0     0.0    0.00   \n",
       "\n",
       "ID          WFDC9  XAGE2  XAGE2B  ZFATAS  ZP4  \n",
       "SEQC_NB001    0.0    0.0     0.0     0.0  0.0  \n",
       "SEQC_NB002    0.0    0.0     0.0     0.0  0.0  \n",
       "SEQC_NB003    0.0    0.0     0.0     0.0  0.0  \n",
       "SEQC_NB004    0.0    0.0     0.0     0.0  0.0  \n",
       "SEQC_NB005    0.0    0.0     0.0     0.0  0.0  \n",
       "\n",
       "[5 rows x 22674 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transpose to have patient IDs on the rows\n",
    "# gene names as column names\n",
    "expr = expr.set_index('ID').T\n",
    "expr.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fdaaa4bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ID\n",
       "ALB         0\n",
       "CD24L4.1    0\n",
       "RPS11       0\n",
       "RPS18       0\n",
       "C5orf13     0\n",
       "           ..\n",
       "WFDC9       0\n",
       "XAGE2       0\n",
       "XAGE2B      0\n",
       "ZFATAS      0\n",
       "ZP4         0\n",
       "Length: 22674, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2810d15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expr.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a72899b",
   "metadata": {},
   "source": [
    "If many NAs, one needs to impute:\n",
    "- replace NAs by 0\n",
    "- replace by median/mean\n",
    "- ..."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fa657d3",
   "metadata": {},
   "source": [
    "# merge expression table and metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "1648377a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(498, 22674)\n",
      "(498, 11)\n"
     ]
    }
   ],
   "source": [
    "print(expr.shape)\n",
    "print(meta_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "a418ca4d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>title</th>\n",
       "      <th>tissue</th>\n",
       "      <th>dataset</th>\n",
       "      <th>Sex</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>mycn_status</th>\n",
       "      <th>high_risk</th>\n",
       "      <th>inss_stage</th>\n",
       "      <th>class_label</th>\n",
       "      <th>progression</th>\n",
       "      <th>...</th>\n",
       "      <th>USP17L5</th>\n",
       "      <th>USP17L7</th>\n",
       "      <th>USP17L8</th>\n",
       "      <th>VTRNA2</th>\n",
       "      <th>WFDC11</th>\n",
       "      <th>WFDC9</th>\n",
       "      <th>XAGE2</th>\n",
       "      <th>XAGE2B</th>\n",
       "      <th>ZFATAS</th>\n",
       "      <th>ZP4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM1205736</th>\n",
       "      <td>SEQC_NB001</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>987</td>\n",
       "      <td>Normal</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Unfavorable</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.74</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205737</th>\n",
       "      <td>SEQC_NB002</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>M</td>\n",
       "      <td>1808</td>\n",
       "      <td>Normal</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Unfavorable</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.98</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205738</th>\n",
       "      <td>SEQC_NB003</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>1</td>\n",
       "      <td>F</td>\n",
       "      <td>625</td>\n",
       "      <td>Amplified</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>Unfavorable</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM1205739</th>\n",
       "      <td>SEQC_NB004</td>\n",
       "      <td>neuroblastoma</td>\n",
       "      <td>2</td>\n",
       "      <td>F</td>\n",
       "      <td>335</td>\n",
       "      <td>Amplified</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>Neither</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows  22685 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 title         tissue dataset Sex age_at_diagnosis  \\\n",
       "GSM1205736  SEQC_NB001  neuroblastoma       1   M              987   \n",
       "GSM1205737  SEQC_NB002  neuroblastoma       2   M             1808   \n",
       "GSM1205738  SEQC_NB003  neuroblastoma       1   F              625   \n",
       "GSM1205739  SEQC_NB004  neuroblastoma       2   F              335   \n",
       "\n",
       "           mycn_status high_risk inss_stage  class_label progression  ...  \\\n",
       "GSM1205736      Normal         1          4  Unfavorable           1  ...   \n",
       "GSM1205737      Normal         1          4  Unfavorable           1  ...   \n",
       "GSM1205738   Amplified         1          4  Unfavorable           1  ...   \n",
       "GSM1205739   Amplified         1          2      Neither           0  ...   \n",
       "\n",
       "           USP17L5  USP17L7  USP17L8  VTRNA2  WFDC11  WFDC9  XAGE2  XAGE2B  \\\n",
       "GSM1205736     0.0      0.0      0.0     0.0    4.74    0.0    0.0     0.0   \n",
       "GSM1205737     0.0      0.0      0.0     0.0    2.98    0.0    0.0     0.0   \n",
       "GSM1205738     0.0      0.0      0.0     0.0    0.00    0.0    0.0     0.0   \n",
       "GSM1205739     0.0      0.0      0.0     0.0    5.14    0.0    0.0     0.0   \n",
       "\n",
       "            ZFATAS  ZP4  \n",
       "GSM1205736     0.0  0.0  \n",
       "GSM1205737     0.0  0.0  \n",
       "GSM1205738     0.0  0.0  \n",
       "GSM1205739     0.0  0.0  \n",
       "\n",
       "[4 rows x 22685 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge expression and metadata\n",
    "expr_tr = meta_tr.merge(expr, left_on=\"title\", right_index=True)\n",
    "expr_tr.head(4)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
