{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "463ff290-d018-40a8-8dca-540b8e56c148",
   "metadata": {},
   "source": [
    "# Data Visualization\n",
    "## Advanced Python for Life Sciences @ Physalia courses (Summer 2025)\n",
    "### Marco Chierici, Fondazione Bruno Kessler - Data Science For Health"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ee18046",
   "metadata": {},
   "source": [
    "# Objectives\n",
    "\n",
    "- Be able to create simple plots with Matplotlib and tweak them\n",
    "- Know about object-oriented vs pyplot interfaces of Matplotlib\n",
    "- Be able to adapt gallery examples\n",
    "- Know that other tools exist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa922d59-76a1-4b51-a7cf-a6326508841e",
   "metadata": {},
   "source": [
    "## Repeatability/reproducibility\n",
    "\n",
    "From [Claus O. Wilke: \"Fundamentals of Data Visualization\"](https://clauswilke.com/dataviz/):\n",
    "\n",
    "> *One thing I have learned over the years is that automation is your friend. I\n",
    "> think figures should be autogenerated as part of the data analysis pipeline\n",
    "> (which should also be automated), and they should come out of the pipeline\n",
    "> ready to be sent to the printer, no manual post-processing needed.*\n",
    "\n",
    "- **No manual post-processing**. This will bite you when you need to regenerate 50\n",
    "  figures one day before submission deadline or regenerate a set of figures\n",
    "  after the person who created them left the group.\n",
    "- There is not the one perfect language and **not the one perfect library** for everything.\n",
    "- Within Python, many libraries exist:\n",
    "  - [Matplotlib](https://matplotlib.org/gallery.html):\n",
    "    probably the most standard and most widely used\n",
    "  - [Seaborn](https://seaborn.pydata.org/examples/index.html):\n",
    "    high-level interface to Matplotlib, statistical functions built in\n",
    "  - [Altair](https://altair-viz.github.io/gallery/index.html):\n",
    "    declarative visualization (R users will be more at home), statistics built in\n",
    "  - [Plotly](https://plotly.com/python/):\n",
    "    interactive graphs\n",
    "  - [Bokeh](https://demo.bokeh.org/):\n",
    "    also good for interactivity\n",
    "  - [plotnine](https://plotnine.readthedocs.io/):\n",
    "    implementation of a grammar of graphics in Python, it is based on [ggplot2](https://ggplot2.tidyverse.org/)\n",
    "  - [ggplot](https://yhat.github.io/ggpy/):\n",
    "    R users will be more at home\n",
    "  - ...\n",
    "- Two main families of libraries: procedural (e.g. Matplotlib) and declarative\n",
    "  (using grammar of graphics).\n",
    "\n",
    "\n",
    "## Why Matplotlib?\n",
    "\n",
    "- Matplotlib is perhaps the most \"standard\" Python plotting library.\n",
    "- Many libraries build on top of Matplotlib.\n",
    "- MATLAB users will feel familiar.\n",
    "- Even if you choose to use another library (see above list), chances are high\n",
    "  that you need to adapt a Matplotlib plot of somebody else.\n",
    "- Libraries that are built on top of Matplotlib may need knowledge of Matplotlib\n",
    "  for custom adjustments.\n",
    "\n",
    "However it is a relatively low-level interface for\n",
    "drawing (in terms of abstractions, not in terms of quality) and does not\n",
    "provide statistical functions. Some figures require typing and tweaking many lines of code.\n",
    "\n",
    "Many other visualization libraries exist with their own strengths, it is also a\n",
    "matter of personal preferences. **Later we will also try other libraries.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a5e3ffd-c1ce-450c-aff3-89018df9249a",
   "metadata": {},
   "source": [
    "# Getting started with Matplotlib\n",
    "\n",
    "Let us create our first plot:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b85ab6bf-dd4a-4311-becb-1e1fb17665a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# canonical import for matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# this line tells Jupyter to display matplotlib figures in the notebook\n",
    "%matplotlib inline\n",
    "\n",
    "# this is Anscombe's quartet dataset 1 from\n",
    "# https://en.wikipedia.org/wiki/Anscombe%27s_quartet\n",
    "data_x = [10.0, 8.0, 13.0, 9.0, 11.0, 14.0, 6.0, 4.0, 12.0, 7.0, 5.0]\n",
    "data_y = [8.04, 6.95, 7.58, 8.81, 8.33, 9.96, 7.24, 4.26, 10.84, 4.82, 5.68]\n",
    "\n",
    "# create a blank figure (canvas)\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# add a plot\n",
    "ax.scatter(x=data_x, y=data_y, c=\"#E69F00\")\n",
    "\n",
    "# customize aesthetics\n",
    "ax.set_xlabel(\"we should label the x axis\")\n",
    "ax.set_ylabel(\"we should label the y axis\")\n",
    "ax.set_title(\"some title\")\n",
    "\n",
    "plt.show()\n",
    "# uncomment the next line if you would like to save the figure to disk\n",
    "# fig.savefig(\"my-first-plot.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f1fdaa2",
   "metadata": {},
   "source": [
    "When running a Matplotlib script on a remote server without a\n",
    "\"display\" (e.g. compute cluster), you may need to add this line:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5561f9ce",
   "metadata": {},
   "source": [
    "```python\n",
    "import matplotlib.pyplot as plt\n",
    "plt.use(\"Agg\")\n",
    "\n",
    "# ... rest of the script\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "398820d6",
   "metadata": {},
   "source": [
    "To add semi-automatic legends, we first need to add a label to the data used in the scatterplot. Then, we call `ax.legend()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9ac10ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.scatter(x=data_x, y=data_y, c=\"#E69F00\", label=\"set 1\")\n",
    "\n",
    "ax.set_xlabel(\"we should label the x axis\")\n",
    "ax.set_ylabel(\"we should label the y axis\")\n",
    "ax.set_title(\"some title\")\n",
    "ax.legend()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97384d2d",
   "metadata": {},
   "source": [
    "Not really informative or necessary with just one variable, but it will become handy with more data points."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b310186",
   "metadata": {},
   "source": [
    "## Exercise"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5d16c48",
   "metadata": {},
   "source": [
    "Extend the previous plot by also plotting this set of values but this time\n",
    "  using a different color (`#56B4E9`):\n",
    "  \n",
    "  ```python\n",
    "  data2_y = [9.14, 8.14, 8.74, 8.77, 9.26, 8.10, 6.13, 3.10, 9.13, 7.26, 4.74]\n",
    "  ```\n",
    "\n",
    "Then add another color (`#009E73`) which plots the second dataset, scaled\n",
    "  by 2.0.\n",
    "  ```python\n",
    "  # here we multiply all elements of data2_y by 2.0\n",
    "  data2_y_scaled = [y*2.0 for y in data2_y]\n",
    "  ```\n",
    "\n",
    "Try to add a legend to the plot with `ax.legend()`\n",
    "\n",
    "Optionally save the plot to a file.\n",
    "\n",
    "At the end it should look like [this one](https://drive.google.com/uc?export=view&id=1BRUlUkRArpVrFZwlnISccI4OuvTvsiVG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3166e293-dbe8-4931-9bfc-55393c5c4f05",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a150544",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this is dataset 2\n",
    "data2_y = [9.14, 8.14, 8.74, 8.77, 9.26, 8.10, 6.13, 3.10, 9.13, 7.26, 4.74]\n",
    "# here we multiply all elements of data2_y by 2.0\n",
    "data2_y_scaled = [y * 2.0 for y in data2_y]\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.scatter(x=data_x, y=data_y, c=\"#E69F00\", label=\"set 1\")\n",
    "ax.scatter(x=data_x, y=data2_y, c=\"#56B4E9\", label=\"set 2\")\n",
    "ax.scatter(x=data_x, y=data2_y_scaled, c=\"#009E73\", label=\"set 2 (scaled)\")\n",
    "\n",
    "ax.set_xlabel(\"we should label the x axis\")\n",
    "ax.set_ylabel(\"we should label the y axis\")\n",
    "ax.set_title(\"some title\")\n",
    "ax.legend()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea3be949",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### Why these colors?\n",
    "\n",
    "This qualitative color palette is opimized for all color-vision\n",
    "deficiencies, see\n",
    "- <https://clauswilke.com/dataviz/color-pitfalls.html> and\n",
    "- [Okabe, M., and K. Ito. 2008. \"Color Universal Design (CUD):\n",
    "How to Make Figures and Presentations That Are Friendly to Colorblind People.\"](http://jfly.iam.u-tokyo.ac.jp/color/)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d04d798f",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Matplotlib has two different interfaces\n",
    "\n",
    "When plotting with Matplotlib, it is useful to know and understand that there are **two approaches** even though the reasons of this dual approach is outside the scope of this lesson.\n",
    "\n",
    "The more modern option is an **object-oriented interface** (OO) - the one we just used. The `fig` and `ax` objects can be configured separately and passed around to functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dd05390",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.scatter(x=data_x, y=data_y, c=\"#E69F00\")\n",
    "\n",
    "ax.set_xlabel(\"we should label the x axis\")\n",
    "ax.set_ylabel(\"we should label the y axis\")\n",
    "ax.set_title(\"some title\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b42fc73",
   "metadata": {},
   "source": [
    "The more traditional option mimics MATLAB plotting and uses the **`pyplot` interface** - also known as MATLAB-like interface, where `plt` carries the global settings:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd935ecd",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(8, 4), dpi=100)\n",
    "\n",
    "plt.scatter(x=data_x, y=data_y, c=\"#E69F00\")\n",
    "\n",
    "plt.xlabel(\"we should label the x axis\")\n",
    "plt.ylabel(\"we should label the y axis\")\n",
    "plt.title(\"some title\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f39a9ec",
   "metadata": {},
   "source": [
    "When searching for help on the internet, you will find both approaches.\n",
    "\n",
    "Although the `pyplot` interface looks more compact, **I recommend to learn and use the object-oriented interface.**\n",
    "\n",
    "Regardless of the interface you choose, **stick to it** in your code and avoid mixing interfaces."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "120225f8",
   "metadata": {},
   "source": [
    "### Why do I emphasize this?\n",
    "\n",
    "One day you may want to write functions which wrap around Matplotlib function calls and then you can send `fig` and `ax` into these functions: there is less risk that adjusting figures changes global settings also for unrelated figures created in other functions.\n",
    "\n",
    "When using the `pyplot` interface, settings are modified for the entire `plt` package. \n",
    "\n",
    "The latter is acceptable for linear scripts but may yield surprising results when introducing functions to enhance/abstract Matplotlib calls."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1bf943f",
   "metadata": {},
   "source": [
    "### Key principles\n",
    "\n",
    "- Avoid manual post-processing, script everything.\n",
    "- Browse a number of example galleries to help you choose the library\n",
    "  that fits best your work/style.\n",
    "- Figures for presentation slides and figures for manuscripts have\n",
    "  different requirements.\n",
    "- Think about color-vision deficiencies when choosing colors. Use\n",
    "  existing solutions for this problem."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed5278f8",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Further steps with Matplotlib\n",
    "\n",
    "## Reading data\n",
    "\n",
    "We will experiment with some example weather data obtained from [Norsk KlimaServiceSenter](https://seklima.met.no/observations/), Meteorologisk institutt (MET) (CC BY 4.0).\n",
    "\n",
    "In this example, we will load a CSV file directly from the web instead of using a local file name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec92d561",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "url = \"https://raw.githubusercontent.com/coderefinery/data-visualization-python/main/data/tromso-daily.csv\"\n",
    "df_tromso = pd.read_csv(url)\n",
    "\n",
    "df_tromso"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11c4e9ec",
   "metadata": {},
   "source": [
    "## Combine dataframes\n",
    "\n",
    "Using pandas we can **merge, join, concatenate, and compare** dataframes, see our dedicated lecture and also <https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html>.\n",
    "\n",
    "Let us try to concatenate two dataframes: one for Tromsø weather data (which we already loaded into `df_tromso`) and one for Oslo, which we load right now:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a089b846",
   "metadata": {},
   "outputs": [],
   "source": [
    "url = \"https://raw.githubusercontent.com/coderefinery/data-visualization-python/main/data/oslo-daily.csv\"\n",
    "df_oslo = pd.read_csv(url)\n",
    "df_oslo"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24a20894",
   "metadata": {},
   "source": [
    "Now we can concatenate them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5764b7b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.concat([df_tromso, df_oslo], axis=0, ignore_index=True)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de048b40-03da-4a07-b239-a85b7add11f2",
   "metadata": {},
   "source": [
    "`axis=0` above means to combine dataframes along the 1st axis, i.e., the rows.\n",
    "\n",
    "**Mini-exercise** Once the dataframes are combined, we can do interesting queries: for example, compute the summary statistics of all variables, breaking down by the \"name\" column. (hint: `describe`, `groupby`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "496db889-6cef-463c-a7ff-16616dfa6612",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f707524",
   "metadata": {},
   "outputs": [],
   "source": [
    "# solution\n",
    "df.groupby(\"name\").describe().transpose()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8363b302",
   "metadata": {},
   "source": [
    "## Data cleaning\n",
    "\n",
    "Before plotting the data, we realize that the `date` data type is not inferred as date or time by pandas, but rather as \"object\" (in pandas lingo, that's a string)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36d56fb6-eb77-48bb-b93d-f55160b0b2e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"date\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2330a334-1e0f-49f6-b5c4-b2f2ad03d4bd",
   "metadata": {},
   "source": [
    "So we convert the dates into proper \"datetime\" format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "733810f8-0a43-4f54-890c-e022c2adf41a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace dd.mm.yyyy to date format\n",
    "df[\"date\"] = pd.to_datetime(df[\"date\"], format=\"%d.%m.%Y\")\n",
    "df[\"date\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed25744b",
   "metadata": {},
   "source": [
    "## Exercise: Plotting data with Matplotlib\n",
    "\n",
    "Consider now the Tromsø dataframe only.\n",
    "\n",
    "1. Convert its 'date' column to proper date and time format. Then, use Matplotlib to create a simple line plot to visualize the maximum temperature as a function of time.\n",
    "\n",
    "2. Add meaningful x & y axis labels and a title.\n",
    "\n",
    "3. (Optional) Save the output figure to a file named \"temperatures-tromso.png\".\n",
    "\n",
    "Hints:\n",
    "* besides `scatter`, maplotlib has a more general `plot` function with similar arguments\n",
    "* use the object-oriented interface (e.g., the one with `ax`), or the MATLAB-like interface (e.g., the one with `plt`), or both!\n",
    "* see the beginning of the notebook for how to save figures to files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7426166-cef9-4fc5-a2b7-3aa4f42a3a35",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "359f1aae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# object-oriented interface\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "# make sure the date axis is in the right format\n",
    "# we fixed it above for df but not for df_tromso !\n",
    "df_tromso[\"date\"] = pd.to_datetime(df_tromso[\"date\"], format=\"%d.%m.%Y\")\n",
    "\n",
    "ax.plot(df_tromso[\"date\"], df_tromso[\"max temperature\"])\n",
    "\n",
    "ax.set_xlabel(\"Date\")\n",
    "ax.set_ylabel(\"Max temperature\")\n",
    "ax.set_title(\"Temperatures in Tromsø\")\n",
    "\n",
    "fig.savefig(\"temperatures-tromso_oob.png\")\n",
    "\n",
    "# this avoid matplotlib to echo the last command, i.e. Text(0.5, 1.0, 'some title')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ef4d3b7-5057-44f1-b4e7-e3d7fc30f14a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# MATLAB-like interface\n",
    "fig = plt.figure()\n",
    "\n",
    "plt.plot(df_tromso[\"date\"], df_tromso[\"max temperature\"])\n",
    "\n",
    "plt.xlabel(\"Date\")\n",
    "plt.ylabel(\"Max temperature\")\n",
    "plt.title(\"Temperatures in Tromsø\")\n",
    "\n",
    "# uncomment the next line if you would like to save the figure to disk\n",
    "fig.savefig(\"temperatures-tromso.png\")\n",
    "\n",
    "# this avoid matplotlib to echo the last command, i.e. Text(0.5, 1.0, 'some title')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6141622f-1782-4a12-a48f-25eeda507fda",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Seaborn quickstart\n",
    "\n",
    "[Seaborn](https://seaborn.pydata.org/examples/index.html) is a visualization library built **on top of Matplotlib**, offering **more modern plot styles** and color defaults, as well as better-looking color palettes. In fact, the look-and-feel of Matplotlib plots is a little old-fashioned in the context of modern data visualization.\n",
    "\n",
    "Another advantage of Seaborn is its **native handling of Pandas** DataFrames, whereas Matplotlib is not designed for use with them.\n",
    "\n",
    "Moreover, Seaborn's API is more high-level than Matplotlib's, thus allowing you to create **even complex visualizations with less boilerplate code** than Matplotlib."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9c0fd0a-917a-4d70-865d-31724d987046",
   "metadata": {},
   "source": [
    "Let's start right away and compare vanilla Matplotlib with Seaborn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b60bc61a-cd4b-4c1e-bb5d-858ea70f28d7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data creation\n",
    "import numpy as np\n",
    "\n",
    "np.random.seed(0)\n",
    "x = np.linspace(0, 10, 500)\n",
    "y = np.cumsum(np.random.randn(500, 6), 0)\n",
    "\n",
    "# plot with Matplotlib defaults\n",
    "plt.plot(x, y)\n",
    "plt.title(\"Matplotlib lineplot\")\n",
    "plt.legend(\"ABCDEF\", ncol=2, loc=\"upper left\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b451540a-d7b9-4b8e-bbf9-dcd11dc99a48",
   "metadata": {},
   "source": [
    "Let's recreate the same plot in Seaborn. There is only one little extra step to put the input data into a pandas DataFrame - since Seaborn is designed to work with DataFrames! Moreover, the DataFrame should be **tidy**, so in long format."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0a64d3f-e57f-4b9c-b732-48c8cfe84bc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "x.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33182cf6-12a7-4087-ab2c-6b63beff14b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "x[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f2949c1-07bb-4ff0-bad9-3bf2358af7de",
   "metadata": {},
   "outputs": [],
   "source": [
    "y.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d7738bd-96a2-4159-96f2-65bbf65aff0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.DataFrame(y, index=x, columns=[\"A\", \"B\", \"C\", \"D\", \"E\", \"F\"])\n",
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2a3756b-24ba-452f-af36-c1bc85119bdf",
   "metadata": {},
   "source": [
    "Pandas `.melt()` takes a dataframe where one or more columns are identifier variables (in our case, the index) and all other columns are considered measured variables (in our case, A:F). These latter columns are unpivoted to the row axis, thus leaving just two non-identifier columns, \"variable\" and \"value\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "448c84ab-4263-4e4f-aea8-80fac4697684",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.reset_index().melt(id_vars=\"index\")\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42b5df5a-05a3-440c-9c1a-c56f899b426b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# canonical import\n",
    "import seaborn as sns\n",
    "\n",
    "plot = sns.lineplot(x=\"index\", y=\"value\", hue=\"variable\", data=data)\n",
    "plot.set_title(\"Seaborn lineplot\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b6fd673-5002-45b7-a68c-250fe89d4d6f",
   "metadata": {},
   "source": [
    "### Exercise\n",
    "\n",
    "Recreate the plot \"Temperatures in Tromsø\" using Seaborn instead of Matplotlib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a2357c4-fdda-42e6-8f1f-898cad679886",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e549e2b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = sns.lineplot(x=\"date\", y=\"max temperature\", data=df_tromso)\n",
    "\n",
    "plot.set_title(\"Temperatures in Tromso\")\n",
    "plot.set_xlabel(\"Date\")\n",
    "plot.set_ylabel(\"Max temp\")\n",
    "\n",
    "# uncomment next line to save the figure\n",
    "# plot.figure.savefig(\"temperatures.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6dae600-c203-40e1-8fa7-8eab8c59cf6c",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "Let's experiment now with *plotting multiple data* on the same plot.\n",
    "\n",
    "As we saw earlier, the `hue='name'` parameter of Seaborn plotting functions automatically breaks down the plots by the categories in the `name` variable of the dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "220251f7-f244-48d3-b32b-e1ddf1ab90e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = sns.lineplot(x=\"date\", y=\"max temperature\", hue=\"name\", data=df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4cc8a04-f788-42fd-bb48-b806e62b1e4b",
   "metadata": {},
   "source": [
    "**We got both timeseries nicely drawn in the same plot.**\n",
    "\n",
    "Let's experiment a little more. What if we wanted the two timeseries in separate panels?\n",
    "\n",
    "We use the versatile function `sns.relplot()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c67434cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = sns.relplot(\n",
    "    x=\"date\",\n",
    "    y=\"max temperature\",\n",
    "    hue=\"name\",\n",
    "    col=\"name\",  # try commenting this out or changing col to row\n",
    "    kind=\"line\",\n",
    "    data=df,\n",
    ")\n",
    "\n",
    "# uncomment next line to save the figure\n",
    "# plot.savefig(\"temperatures-side-by-side.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a553f7ee",
   "metadata": {},
   "source": [
    "Note how by changing two lines we can change the data plotted (`x` or `y`) and the plot type (`kind`).\n",
    "\n",
    "**Your turn!** \n",
    "\n",
    "Now plot the snow depth vs. time as a scatter plot instead of a line plot. Optionally save the figure to the file \"snow-depth.png\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e4da19c-b192-418f-a7b4-051dee50dc2c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ac30143",
   "metadata": {},
   "outputs": [],
   "source": [
    "# solution\n",
    "plot = sns.relplot(\n",
    "    x=\"date\",\n",
    "    y=\"snow depth\",  # we now plot the snow depth instead of the max temperature\n",
    "    hue=\"name\",\n",
    "    col=\"name\",\n",
    "    kind=\"scatter\",  # we changed \"line\" to \"scatter\"\n",
    "    data=df,\n",
    ")\n",
    "\n",
    "# uncomment next line to save the figure\n",
    "# plot.savefig(\"snow-depth.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcb1e006-ff87-42db-9248-8288f92f536d",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# More on Matplotlib\n",
    "\n",
    "## Anatomy of a Figure\n",
    "\n",
    "![](https://matplotlib.org/stable/_images/anatomy.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c1e56c",
   "metadata": {},
   "source": [
    "##  Axes\n",
    "\n",
    "A figure can have one or more subplots inside it called Axes, arranged in rows and columns. Every figure has at least one Axes. Don't confuse Axes with X and Y axis: they are different!\n",
    "\n",
    "The Axes objects, such as `ax1` and `ax2` above, are what you think of as 'a plot'. It is the region of the image with the data space. A given figure can contain many Axes, but a given Axes object can only be in one Figure. The **Axes** contains two (or three in the case of 3D) **Axis** objects (be aware of the difference between Axes and Axis!!) which take care of the data limits (the data limits can also be controlled via `set_xlim()` and `set_ylim()` Axes methods). \n",
    "\n",
    "Each Axes has:\n",
    "\n",
    "1. a title (set via `set_title()`);\n",
    "1. a x-label (set via `set_xlabel()`);\n",
    "1. a y-label (set via `set_ylabel()`).\n",
    "\n",
    "The Axes class and its member functions are the primary entry point to working with the object-oriented programming (OOP) interface.\n",
    "\n",
    "## Axis\n",
    "\n",
    "These are the number-line-like objects. They take care of setting the graph limits and generating the ticks (the marks on the axis) and ticklabels (strings labeling the ticks). The location of the ticks is determined by a Locator object and the ticklabel strings are formatted by a Formatter. The combination of the correct Locator and Formatter gives very fine control over the tick locations and labels.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2d6cb6a",
   "metadata": {},
   "source": [
    "## How to draw two scatterplots in different panels\n",
    "\n",
    "Suppose I want to draw our two sets of points in two separate plots side-by-side instead of the same plot. \n",
    "\n",
    "You can do that by creating two separate subplots, or Axes, using `plt.subplots(1, 2)`: this means to creates 1 row with 2 subplots. The command returns two objects:\n",
    "\n",
    "1. the figure\n",
    "1. the axes (subplots) inside the figure\n",
    "\n",
    "Previously, we used `plt.plot()`. Since there was only one axes by default, matplotlib drew the points on that axes itself.\n",
    "\n",
    "But now, since you want the points drawn on different subplots (axes), you have to call the plot function **on the respective axes**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f2685fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create some data\n",
    "x = np.linspace(0, 5, 10)\n",
    "y = x**2\n",
    "\n",
    "# Create Figure and Subplots, capturing them in separate variables\n",
    "fig, axes = plt.subplots(1, 2)\n",
    "\n",
    "# Grab the individual Axis from the Axes object\n",
    "ax1 = axes[0]\n",
    "ax2 = axes[1]\n",
    "\n",
    "# Plot\n",
    "ax1.plot(x, y, \"o\")\n",
    "ax2.plot(x, y, \"o\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4004e7c3-cb7e-4067-a960-6fe45ac92e95",
   "metadata": {},
   "source": [
    "The above code seems quite repetitive and can be further optimized:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa0ee95f-e3a5-4598-96f0-ac04a5c0369f",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 2)\n",
    "\n",
    "for ax in axes:\n",
    "    ax.plot(x, y, \"o\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0c787b7-5ea7-4d28-b8ca-324eab88cc56",
   "metadata": {},
   "source": [
    "Let's add titles and x-y labels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd3f132f-a736-40e0-bd12-e554af1f5ef4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 2)\n",
    "\n",
    "for ax in axes:\n",
    "    ax.plot(x, y, \"o\")\n",
    "    ax.set_xlabel(\"x\")\n",
    "    ax.set_ylabel(\"y\")\n",
    "    ax.set_title(\"Scatterplot\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e8f4f4f-765d-4ba2-a13e-00ea8f8fcdc0",
   "metadata": {},
   "source": [
    "Quite good, but figure axes and labels overlap: we fix this with the `fig.tight_layout` method, which automatically adjusts the positions of the axes on the figure canvas so that there is no overlapping content:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb43dbfe-f2fd-47de-995e-27162fcd7aed",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 2)\n",
    "\n",
    "for ax in axes:\n",
    "    ax.plot(x, y, \"o\")\n",
    "    ax.set_xlabel(\"x\")\n",
    "    ax.set_ylabel(\"y\")\n",
    "    ax.set_title(\"Scatterplot\")\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ba7c1b8-792e-4937-81e4-68b2f537375e",
   "metadata": {},
   "source": [
    "Better, but we notice that the y-axis labels of the right-hand panel are redundant: in the following example, we create two Axes sharing the y axis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "874d4c16-c8f6-433a-9fcc-c71a8b01f9bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 2, sharey=True)\n",
    "\n",
    "for ax in axes:\n",
    "    ax.plot(x, y, \"o\")\n",
    "    ax.set_xlabel(\"x\")\n",
    "    ax.set_ylabel(\"y\")\n",
    "    ax.set_title(\"Scatterplot\")\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e217c7c-339b-4a80-9fd8-208965eeff31",
   "metadata": {},
   "source": [
    "Thats sounds like a lot of functions to learn! It's actually quite easy to remember them.\n",
    "\n",
    "The `ax1` and `ax2` objects, like `plt`, have equivalent `set_title`, `set_xlabel` and `set_ylabel` functions. In fact, `plt.title()` actually calls the current axes `set_title()` to do the job.\n",
    "\n",
    "* `plt.xlabel()` → `ax.set_xlabel()`\n",
    "* `plt.ylabel()` → `ax.set_ylabel()`\n",
    "* `plt.xlim()` → `ax.set_xlim()`\n",
    "* `plt.ylim()` → `ax.set_ylim()`\n",
    "* `plt.title()` → `ax.set_title()`\n",
    "\n",
    "Alternately, to save keystrokes, you can set multiple things in one go using the `ax.set()`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f72972cb-ab20-4a69-95fc-488693ba54f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(1, 2, sharey=True)\n",
    "\n",
    "for ax in axes:\n",
    "    ax.plot(x, y, \"o\")\n",
    "    ax.set(title=\"Scatterplot\", xlabel=\"x\", ylabel=\"y\", xlim=(0, 6), ylim=(0, 30))\n",
    "\n",
    "fig.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c03d5879-93a0-4dbc-893a-b374ceacb6c9",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Plotting with Pandas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a8523f1-b0b9-4e34-a927-635d13850d55",
   "metadata": {},
   "source": [
    "We can plot pandas dataframes directly! This implicitly uses `matplotlib.pyplot`.\n",
    "\n",
    "Plotting in pandas is as simple as using a `.plot()` method on a series or dataframe.\n",
    "\n",
    "To demonstrate pandas plotting, let's reload the subset of the breast cancer dataset that we prepared earlier this week:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56a5b6e4-a238-4317-8156-7f014849345a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "DATADIR = Path(\"data\")\n",
    "\n",
    "bc_data = pd.read_csv(DATADIR / \"bc_new_table.txt\", sep=\"\\t\", index_col=0)\n",
    "bc_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95cb1f04-54b3-4241-b5d6-46902a9248eb",
   "metadata": {},
   "source": [
    "`pandas` incorporates part of the `matplotlib` library to create quick data visualizations. \n",
    "\n",
    "We use the `.plot(x, y, kind, data)` method directly on a dataframe `data` and specify the column names we want to plot (`x` and `y` arguments) and the type of plot (`kind`).\n",
    "\n",
    "Some of the supported `kind`s of plots are:\n",
    "- 'scatter' for scatterplots,\n",
    "- 'hist' for histograms,\n",
    "- 'bar' for vertical barplots,\n",
    "- 'barh' for horizontal barplots,\n",
    "- 'box' for boxplots,\n",
    "- 'kde' for kernel density estimate plots (or simply density plots)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de93cb1f-1ba0-4336-aab3-5b0d0250deee",
   "metadata": {},
   "source": [
    "Some kind of plots need both `x` and `y`, others just `y`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c65030f-5ab7-4835-967a-82bf8166a046",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to build a scatter plot of 'texture' vs 'radius'\n",
    "bc_data.plot(kind=\"scatter\", x=\"radius\", y=\"texture\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba2be4fc-ae04-4414-a716-fbe0d3c4e386",
   "metadata": {},
   "outputs": [],
   "source": [
    "# exploring the parameters we can customize the plot\n",
    "bc_data.plot(\n",
    "    kind=\"scatter\",\n",
    "    x=\"radius\",\n",
    "    y=\"texture\",\n",
    "    figsize=(6, 5),  # custom figure size\n",
    "    alpha=0.6,  # opacity\n",
    "    color=\"blue\",  # fill color\n",
    "    edgecolor=\"w\",\n",
    "    s=80,  # point size\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c1c90a8-b112-4571-8595-45c9d8ecd551",
   "metadata": {},
   "source": [
    "What about customizing titles and x/y axis labels? How to save a pandas plot to file?\n",
    "\n",
    "This has to be done using Matplotlib syntax, as below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1e9edfe-b344-4417-a575-2cec3f2e7626",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we need to assign the pandas plot to a variable\n",
    "plot = bc_data.plot(\n",
    "    kind=\"scatter\",\n",
    "    x=\"radius\",\n",
    "    y=\"texture\",\n",
    "    figsize=(6, 5),  # custom figure size\n",
    "    alpha=0.6,  # opacity\n",
    "    color=\"blue\",  # fill color\n",
    "    edgecolor=\"w\",\n",
    "    s=80,  # point size\n",
    ")\n",
    "\n",
    "plt.title(\"Custom Scatterplot\")\n",
    "plt.xlabel(\"Radius\")\n",
    "plt.ylabel(\"Texture\")\n",
    "\n",
    "# get the current figure from the plot\n",
    "fig = plot.get_figure()\n",
    "# save the current figure to file\n",
    "fig.savefig(\"test.png\")\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c771196-2548-4a15-89c8-61b3bad59710",
   "metadata": {},
   "source": [
    "### Exercise\n",
    "\n",
    "Experiment on pandas' `plot` capabilities. \n",
    "\n",
    "1. Following the rationale above, first create a histogram of the `radius` variable. Make the bars light blue with black borders."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5383e00-4a32-4674-9cc1-f532099d1193",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3df5919c-9d4d-4e28-a518-3fb1880a66e1",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "bc_data.plot(kind=\"hist\", y=\"radius\", color=\"lightblue\", edgecolor=\"black\", width=1.5);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e9e4297-5c4e-4997-a3a6-65ea23818770",
   "metadata": {},
   "source": [
    "2. Create an alternate version of the above by selecting the `radius` column from the Pandas dataframe just before calling the `.plot` method. Modify then the `.plot` arguments accordingly. Use the default color for the bars this time, with black borders."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b48fb6f9-0fb9-4476-a21a-cb868bd7831a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8c51c7f-0f22-44da-acee-91bbc53852ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc_data[\"radius\"].plot(kind=\"hist\", edgecolor=\"black\", width=1.5);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de5b4132-41fd-488a-be21-d6ae4ffa29b2",
   "metadata": {},
   "source": [
    "3. Try plotting multiple histograms on the same plot! One should be light blue and the other orange. \n",
    "\n",
    "Hint: (version 1) you can pass more than one column name to the `y` and `color` arguments of the `plot` method. Or (version 2) you can select more than one column from the dataframe and then call `plot`.\n",
    "\n",
    "Add some transparency (`alpha`) to help discriminating both histograms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5efe2f3a-e2cf-4077-9e1f-b19410261a4c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0340f00-02d2-402c-ba0c-2e446a0e2cb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# version 1\n",
    "bc_data.plot(\n",
    "    kind=\"hist\",\n",
    "    y=[\"radius\", \"texture\"],\n",
    "    bins=15,\n",
    "    alpha=0.8,\n",
    "    width=1.8,\n",
    "    color=[\"lightblue\", \"orange\"],\n",
    "    edgecolor=\"black\",\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3673b40a-a600-4f27-be72-56a04d776cc1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98713c80-e478-404a-aa4f-d0a235cd92aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# version 2\n",
    "bc_data[[\"radius\", \"texture\"]].plot(\n",
    "    kind=\"hist\",\n",
    "    bins=15,\n",
    "    alpha=0.8,\n",
    "    width=1.8,\n",
    "    color=[\"lightblue\", \"orange\"],\n",
    "    edgecolor=\"black\",\n",
    ");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c4c93aa-f84a-4575-91ee-bdec05cb777b",
   "metadata": {},
   "source": [
    "4. (optional) It is also possible to put the `.plot()` right after some other functions applied to the dataframe: e.g. `df.some_func().plot()`. For example, create a bar plot to visualize the number of benign and malignant diagnoses in the dataset.\n",
    "\n",
    "Hint: `value_counts()` or `value_counts(normalize=True)`\n",
    "\n",
    "Make the bars light blue with blue edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b78e03c-7595-487a-9af9-e2fe51063f6f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "050e4995-efcc-4e05-9c01-d5213e41151a",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc_data[\"diagnosis\"].value_counts(normalize=True).plot(\n",
    "    kind=\"bar\", color=\"lightblue\", edgecolor=\"blue\", fontsize=12, width=0.5, rot=0\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81966d39-bc8d-443b-af0c-9e28aabd42f3",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Seaborn plot gallery\n",
    "\n",
    "With matplotlib, you may have much more control over colors, axes, sizes and shapes; on the other hand, it is much more user friendly to use Seaborn or Pandas plotting.\n",
    "\n",
    "## Heatmaps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "671e46c2-43b7-4517-b4ce-2954ac068d58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# correlation matrix\n",
    "corr_table = bc_data.corr(numeric_only=True)\n",
    "\n",
    "# plot it as a heatmap with searborn\n",
    "sns.heatmap(corr_table);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ce930c0-c61b-4599-84a6-0788eef1cb03",
   "metadata": {},
   "outputs": [],
   "source": [
    "# change a few parameters\n",
    "sns.heatmap(corr_table, cmap=\"Blues\", linewidth=0.1);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8d9ae98-531d-439f-a4fb-47cbc2262865",
   "metadata": {},
   "source": [
    "The argument `cmaps` accepts a [matplotlib colormap name](https://matplotlib.org/stable/users/explain/colors/colormaps.html), or list of colors."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f48e807-5587-4f88-9369-a6f4e39fd36b",
   "metadata": {},
   "source": [
    "## Pair plots\n",
    "\n",
    "Pair plots, also called \"scatterplot matrices\", are useful to explore correlations in multidimensional data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16ca34eb-695a-4555-8c84-3b7aa75ad4c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.pairplot(bc_data);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43830723-8ad2-4707-9d87-4e7d10ba9490",
   "metadata": {},
   "source": [
    "**Your turn!**\n",
    "\n",
    "Remake the scatterplot matrix above by coloring the points according to the diagnosis status."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "695135f1-b7c6-48b5-9d8f-5b4470c5ec4b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76209c54-8fa7-4acf-8013-f02a10422af3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.pairplot(bc_data, hue=\"diagnosis\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "788f6998-489b-41a4-83df-8dc484af533f",
   "metadata": {},
   "source": [
    "## Boxplots\n",
    "\n",
    "`seaborn` can be extremely helpful as a first sneak peek at the correlation between multiple variables. In addition, seaborn also provides an efficient way to plot box plots directly from our dataframes using `sns.boxplot` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbf92ff5-b0ae-477f-94f3-9f6347877244",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a simple boxplot with seaborn:\n",
    "# by default they already look good\n",
    "\n",
    "sns.boxplot(\n",
    "    y=\"area\",\n",
    "    x=\"diagnosis\",\n",
    "    hue=\"diagnosis\",\n",
    "    palette=[\n",
    "        \"tomato\",\n",
    "        \"lightgreen\",\n",
    "    ],  # colors to use for the levels of the ``hue`` variable\n",
    "    data=bc_data,\n",
    "    width=0.5,\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cecd322d-16a2-4485-8ed7-32a18f996d51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we can also add a stripplot to make it more informative\n",
    "sns.boxplot(\n",
    "    y=\"area\",\n",
    "    x=\"diagnosis\",\n",
    "    hue=\"diagnosis\",\n",
    "    palette=[\"tomato\", \"lightgreen\"],\n",
    "    data=bc_data,\n",
    "    width=0.5,\n",
    ")\n",
    "\n",
    "sns.stripplot(\n",
    "    y=\"area\",\n",
    "    x=\"diagnosis\",\n",
    "    hue=\"diagnosis\",\n",
    "    palette=[\"tomato\", \"lightgreen\"],\n",
    "    alpha=0.6,\n",
    "    data=bc_data,\n",
    ")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bf82c90-9e71-4348-81e0-d6d50f1e279b",
   "metadata": {},
   "source": [
    "Notice how Seaborn \"adds\" plots to the same figure above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d46f717-f7bb-417a-a56c-d326a7f40e25",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tweak the plotting parameters to make it more appealing\n",
    "sns.boxplot(\n",
    "    y=\"area\",\n",
    "    x=\"diagnosis\",\n",
    "    hue=\"diagnosis\",\n",
    "    palette=[\"tomato\", \"lightgreen\"],\n",
    "    data=bc_data,\n",
    "    width=0.5,\n",
    ")\n",
    "\n",
    "sns.stripplot(\n",
    "    y=\"area\",\n",
    "    x=\"diagnosis\",\n",
    "    hue=\"diagnosis\",\n",
    "    palette=\"dark:black\",\n",
    "    alpha=0.5,\n",
    "    size=4,\n",
    "    edgecolor=\"w\",\n",
    "    linewidth=0.5,\n",
    "    data=bc_data,\n",
    ");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2eaae0e-837f-43a8-9c0a-aa1af28f7020",
   "metadata": {},
   "source": [
    "## Histograms and density plots\n",
    "\n",
    "Seaborn produces histograms with `sns.histplot` and density plots `sns.kdeplot`.\n",
    "\n",
    "### Exercise\n",
    "\n",
    "Using the same rationale above, create a histogram of the `radius` variable with Seaborn."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "302ab748-00a8-464a-b9c1-7698f4ac9f4c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a51c72bd-a002-4811-bdbc-60e6684c1233",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(data=bc_data, x=\"radius\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76f92356-77c0-4073-8c2f-bae5e200342e",
   "metadata": {},
   "source": [
    "Now, add to the same plot the histogram of the variable `texture`. Hint: see the boxplot example above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "094a1386-ba7a-4741-b7e2-957002ab2188",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cbe8f15-942c-4561-a99d-c14e56a85c6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(data=bc_data, x=\"radius\")\n",
    "sns.histplot(data=bc_data, x=\"texture\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "845950e9-5b3b-4089-a433-0be559aa9124",
   "metadata": {},
   "source": [
    "See what happens if you add the optional argument `kde=True`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87fe9735-0f77-404e-84cd-a551f56fcefe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f918f1d4-b48f-49f3-bc5c-6f50a377c1ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.histplot(data=bc_data, x=\"radius\", kde=True)\n",
    "sns.histplot(data=bc_data, x=\"texture\", kde=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f422605a-ab71-4f50-8066-17fdba7bf4d3",
   "metadata": {},
   "source": [
    "Create now a density plot of the variable `area`. Try with and without the optional argument `fill=True`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b5a1d8e-1aa2-4683-95a4-fd2c48baf627",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75a48c53-646a-495a-bf03-e372bd5e7456",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.kdeplot(data=bc_data, x=\"area\", fill=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7d943bd-dcbd-47c2-98e0-85fe60ee610c",
   "metadata": {},
   "source": [
    "What happens if you create a density plot using two variables? e.g. `area` and `smoothness`. Again, try with and without the optional argument `fill=True`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "947a1a54-2aef-496b-8674-e588f9532e00",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f2b90e3-a17c-4376-b0cc-c09e4b427a5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# a bivariate plot\n",
    "sns.kdeplot(data=bc_data, x=\"area\", y=\"smoothness\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31558e0d-9484-461c-bb65-4b9434c7c5d3",
   "metadata": {},
   "source": [
    "## Plots with marginal distributions\n",
    "\n",
    "These are obtained with `sns.jointplot`.\n",
    "\n",
    "Assigning a `hue` variable will add conditional colors to the plots and an automatic legend."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a891f72c-c9fb-4fa2-bd6a-ebd222ff40a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.jointplot(x=\"concavity\", y=\"texture\", data=bc_data);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe25163b-d52f-4c13-8aaa-a3624132627e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.jointplot(x=\"concavity\", y=\"texture\", hue=\"diagnosis\",\n",
    "              data=bc_data);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10f660a9-0d62-4370-9ef5-ec3e817c46f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.jointplot(x=\"concavity\", y=\"texture\",\n",
    "              kind=\"kde\", # hist, hex, reg (regression)\n",
    "              fill=True,\n",
    "              data=bc_data);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21f686b5-862c-49d9-9bf5-8ba858b0d45b",
   "metadata": {},
   "source": [
    "## Faceted plots\n",
    "\n",
    "To demonstrate faceted plots, we load the dataset \"tips\" from Seaborn:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5a7ef19-3767-4df5-86ae-bd6160439391",
   "metadata": {},
   "outputs": [],
   "source": [
    "tips = sns.load_dataset(\"tips\")\n",
    "tips.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4732a7a-1878-49cc-841f-770d6def85e2",
   "metadata": {},
   "source": [
    "Before proceeding, we compute the tip percentage and add it as a new `tip_pct` column. Then, we plot it as a histogram using Seaborn, with light gray bars and a black edge."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd14c3e9-9a8f-45e5-b2b9-c9252ccf297c",
   "metadata": {},
   "outputs": [],
   "source": [
    "tips[\"tip_pct\"] = 100 * tips[\"tip\"] / tips[\"total_bill\"]\n",
    "sns.histplot(data=tips, x=\"tip_pct\", color=\"lightgray\", edgecolor=\"black\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3199a9b4-7be8-41ed-8cc3-46ac3b316538",
   "metadata": {},
   "source": [
    "Now we want to draw 4 histograms of the tip percentage broken down by both `time` (Lunch and Dinner) and `sex` (Male and Female): we would like to have `time` across the columns and `sex` across the rows.\n",
    "\n",
    "We use seaborn's `FacetGrid()` to create the layout and then we use `map()` on the grid to apply a plotting function (`sns.histplot`) to each facet's subset of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c78983f0-0882-421a-a9f9-91a0e88ed1f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "grid = sns.FacetGrid(tips, col=\"time\", row=\"sex\")\n",
    "grid.map(sns.histplot, \"tip_pct\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "549537e3-7d09-4b16-9c41-e40fb7ea8303",
   "metadata": {},
   "outputs": [],
   "source": [
    "# margin_titles may not work in all situations\n",
    "grid = sns.FacetGrid(tips, col=\"time\", row=\"sex\", margin_titles=True)\n",
    "grid.map(sns.histplot, \"tip_pct\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28caa00b-be2c-45f6-a100-9359fefaaa80",
   "metadata": {},
   "outputs": [],
   "source": [
    "grid = sns.FacetGrid(tips, row=\"sex\", col=\"time\")\n",
    "grid.map_dataframe(sns.histplot, x=\"tip_pct\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae77f2e7-0a43-4091-a23e-31b37c85fab2",
   "metadata": {},
   "source": [
    "This time I used `map_dataframe`, which acts like `map` but is made for plotting with functions that accept a long-form DataFrame as a `data` keyword argument and access the\n",
    "data in that DataFrame using string variable names (`x=\"tip_pct\"`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4237899-c0f6-4efb-9413-f730923ec01b",
   "metadata": {},
   "source": [
    "### Exercise (optional)\n",
    "\n",
    "Following the above examples, try plotting:\n",
    "\n",
    "1. two scatterplots in one row, showing `tip` vs. `total_bill` faceted by `sex` (on the columns), and colored by `smoker` (use `hue` in the call to `FacetGrid`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "216676a7-2849-43f8-9c7f-239be752a0b6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16c94afb-2fe7-4bb8-aee6-7c8efb2e5b16",
   "metadata": {},
   "outputs": [],
   "source": [
    "grid = sns.FacetGrid(tips, col=\"sex\", hue=\"smoker\")\n",
    "grid.map(sns.scatterplot, \"total_bill\", \"tip\", alpha=0.7)\n",
    "grid.add_legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acc6e7ae-4a5b-480f-b44a-34ee57924a68",
   "metadata": {},
   "source": [
    "2. four barplots in one row, showing `total_bill` by `sex`, faceted by `day` (on the columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93480c77-536a-41c8-a74c-d455d7c7a8ae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d524cc8f-f61f-4ee9-ba45-73f220f9a5f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.FacetGrid(tips, col=\"day\", height=4, aspect=.5)\n",
    "g.map(sns.barplot, \"sex\", \"total_bill\", order=[\"Male\", \"Female\"])\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f393781-7ec3-47f3-b005-2e7c3db93b38",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "79c4c0d0-2c14-4fcf-a6bb-ca0c2aba25bd",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Interactive plotting: plotly"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a1a48a9-b8a9-484d-a1b5-75df4a406f9c",
   "metadata": {},
   "source": [
    "Plotly is a powerful tool designed to create interactive, publication-quality graphs. \n",
    "\n",
    "From simple line charts to complex three-dimensional surfaces, plotly provides an extensive range of chart types and can be used in Jupyter Notebooks, as a standalone HTML, and integrated directly into web applications. \n",
    "\n",
    "What sets plotly apart is its ability to create **interactive plots** that the user can engage with; for example, by *zooming in* on areas of interest or *hovering* over data points to reveal more information.\n",
    "\n",
    "One of plotly's greatest strengths is its compatibility with numerous data science libraries such as Pandas, NumPy, and SciPy, allowing for seamless integration with the broader Python data science ecosystem.\n",
    "\n",
    "Plotly has various modules for different purposes:\n",
    "\n",
    "1. `plotly.graph_objects` (imported as `go`) is part of Plotly’s core functionality and is used for creating a wide range of charts and graphs. It offers fine-grained control over the appearance of the plots, including colors, layouts, and annotations.\n",
    "2. `plotly.express` (imported as `px`) is a higher level, function oriented API providing a simpler, more streamlined syntax for creating common types of charts and visualizations."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9f9f0d4-a92c-4491-84ca-967ca5fa5120",
   "metadata": {},
   "source": [
    "## Installation\n",
    "\n",
    "```\n",
    "conda install -c conda-forge plotly\n",
    "```\n",
    "\n",
    "or\n",
    "\n",
    "```\n",
    "pip install plotly\n",
    "```\n",
    "\n",
    "For use within Jupyter Lab, you'll also need anywidget\n",
    "\n",
    "```\n",
    "conda install -c conda-forge jupyterlab anywidget\n",
    "```\n",
    "\n",
    "```\n",
    "pip install jupyterlab anywidget\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1374eda-8bdc-41cb-9cb6-1b06efc6d75f",
   "metadata": {},
   "source": [
    "## First steps with plotly.express"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cfeace2-e021-4b46-95a8-f1aa00e4acc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import numpy as np\n",
    "# import pandas as pd\n",
    "# import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7f3d9ff-be03-4713-a514-0b32bb85f188",
   "metadata": {},
   "outputs": [],
   "source": [
    "# canonical import\n",
    "import plotly.express as px\n",
    "\n",
    "# x, y are list-like objects\n",
    "x = np.arange(5)\n",
    "y = x**2\n",
    "\n",
    "fig = px.scatter(x=x, y=y)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54aec927-b250-41c7-aaa6-ef273b083d55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# x, y are columns of a pandas dataframe\n",
    "df = px.data.iris()  # iris is a pandas DataFrame\n",
    "fig = px.scatter(df, x=\"sepal_width\", y=\"sepal_length\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c8dd34e-2e1d-48ac-a7e4-283b9fb9f3ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tweaking color and size options creates a bubble chart\n",
    "fig = px.scatter(df, x=\"sepal_width\", y=\"sepal_length\", color=\"species\",\n",
    "                 size=\"petal_length\", hover_data=[\"petal_width\"])\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23fa42e8-48b3-4ac3-9f7d-d231080ba806",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.scatter(df, x=\"sepal_width\", y=\"sepal_length\", color=\"species\", symbol=\"species\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f188122a-1ad5-494f-bead-bae292742c9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# marginal distribution plot\n",
    "fig = px.scatter(df, x=\"sepal_length\", y=\"sepal_width\", marginal_x=\"histogram\", marginal_y=\"rug\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "339ac0ac-299e-4c5a-a03e-0aebe8b861c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# a bar plot\n",
    "categories = [\"A\", \"B\", \"C\", \"D\", \"E\"]\n",
    "values = [15, 22, 18, 12, 28]\n",
    "\n",
    "fig = px.bar(x=categories, y=values)\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca11b118-6159-4c7a-97b2-d2ecaaad0619",
   "metadata": {},
   "source": [
    "Styling plotly express plots:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53962eda-063d-4b48-9a5a-dd58438c2d00",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = px.data.tips()\n",
    "fig = px.histogram(df, x=\"day\", y=\"total_bill\", color=\"sex\")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a89ffe5-e1d4-4c35-a81f-9a2f3fc11caa",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.histogram(df, x=\"day\", y=\"total_bill\", color=\"sex\",\n",
    "            title=\"Receipts by Payer Gender and Day of Week\",\n",
    "            width=600, height=400,\n",
    "            labels={  # replaces default labels by column name\n",
    "                \"sex\": \"Payer Gender\",  \"day\": \"Day of Week\", \"total_bill\": \"Receipts\"\n",
    "            },\n",
    "            category_orders={  # replaces default order by column name\n",
    "                \"day\": [\"Thur\", \"Fri\", \"Sat\", \"Sun\"], \"sex\": [\"Male\", \"Female\"]\n",
    "            },\n",
    "            template=\"simple_white\"\n",
    "            )\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1125ff17-0e85-4481-b448-b2020cf95407",
   "metadata": {},
   "source": [
    "Available themes and templates:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "735668f9-39b7-4f02-a351-93b403bc7f70",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.io as pio\n",
    "pio.templates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2aefd69-e289-4cb8-85ae-393b5c4b15a1",
   "metadata": {},
   "source": [
    "Update or modify plots:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8ae1ef0-c689-4ceb-b457-aa084b2bb339",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.histogram(df, x=\"day\", y=\"total_bill\", color=\"sex\",\n",
    "            title=\"Receipts by Payer Gender and Day of Week\",\n",
    "            width=600, height=400,\n",
    "            labels={ # replaces default labels by column name\n",
    "                \"sex\": \"Payer Gender\",  \"day\": \"Day of Week\", \"total_bill\": \"Receipts\"\n",
    "            },\n",
    "            category_orders={ # replaces default order by column name\n",
    "                \"day\": [\"Thur\", \"Fri\", \"Sat\", \"Sun\"], \"sex\": [\"Male\", \"Female\"]\n",
    "            },\n",
    "            template=\"simple_white\"\n",
    "            )\n",
    "\n",
    "fig.update_layout( # customize font and legend orientation & position\n",
    "    font_family=\"Rockwell\",\n",
    ")\n",
    "\n",
    "fig.add_shape( # add a horizontal line\n",
    "    type=\"line\", line_color=\"black\", line_width=3, opacity=1, line_dash=\"dot\",\n",
    "    x0=0, x1=1, xref=\"paper\", y0=950, y1=950, yref=\"y\"\n",
    ")\n",
    "\n",
    "fig.add_annotation( # add a text callout with arrow\n",
    "    text=\"below target\", x=\"Fri\", y=400, arrowhead=1, showarrow=True\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21bd6741-7c66-4965-bbf8-830cb962d4cd",
   "metadata": {},
   "source": [
    "## Saving and embedding Plotly charts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce7315d9-03a7-4e3c-9434-c64a8ac56619",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig.write_html(\"customized_histogram.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51b6eae0-c598-498d-86ee-a486d1abbf1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%html\n",
    "<iframe\n",
    "  src=\"customized_histogram.html\"\n",
    "  width=\"800\"\n",
    "  height=\"600\"\n",
    "  title=\"chart name\"\n",
    "  style=\"border:none\">\n",
    "</iframe>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14d400d4-65f9-4a0e-865c-3dd3428c6b12",
   "metadata": {},
   "source": [
    "## Saving static Plotly charts\n",
    "\n",
    "You can easily export a static version of your Plotly charts: for example, in PDF format.\n",
    "\n",
    "In order to do this, you have to first install a required package in your environment: Kaleido, which is a library for generating static images without the need of external dependencies.\n",
    "\n",
    "So type the following in a code cell:\n",
    "\n",
    "```\n",
    "!pip install -U kaleido\n",
    "```\n",
    "\n",
    "if you want to use `pip`, or\n",
    "\n",
    "```\n",
    "!conda install -y -c conda-forge python-kaleido\n",
    "```\n",
    "\n",
    "if you use `conda`.\n",
    "\n",
    "You may need to *restart the kernel* afterwards (i.e., in Jupyter Lab's menu: Kernel > Restart Kernel and run up to selected cell...).\n",
    "\n",
    "Then, you can go ahead and export any Plotly chart to a static format such as PDF or PNG."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4451dcdd-deed-4b6c-9df2-dc31f519bb21",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig.write_image(\"customized_histogram.pdf\")  # the format is inferred from file extension"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "214e961c-3f3d-4f26-8a83-0dc95fb5ffb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig.write_image(\"customized_histogram.png\")  # the format is inferred from file extension"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84a2f76a-54fd-44d7-9266-77d3d63756bb",
   "metadata": {},
   "source": [
    "## Exercise\n",
    "\n",
    "Use plotly with the `bc_data` to recreate the marginal distribution plot of `texture` vs. `concavity`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a6e3852-68ab-43d4-bf17-a5910563802a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff72f43a-9b5c-4845-a9ca-3880d06f0898",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = px.scatter(\n",
    "    bc_data,\n",
    "    x=\"concavity\",\n",
    "    y=\"texture\",\n",
    "    color=\"diagnosis\",  # optional, same as 'hue' in seaborn\n",
    "    marginal_x=\"histogram\",\n",
    "    marginal_y=\"histogram\",\n",
    ")\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77800122-f398-44ec-af25-bef5f74b5050",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Constructing biologically relevant plots\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39814035-f715-4df8-8804-cf92bea58db0",
   "metadata": {},
   "source": [
    "## Manhattan plot\n",
    "\n",
    "Manhattan plots are a variation of scatterplots used to display dense data. They typically represent the p-values of an entire genome-wide association study (GWAS) on a genomic scale: $-log10(P)$ on the y axis and chromosomes on the x axis. We'll use the package `qmplot`:\n",
    "\n",
    "```\n",
    "!pip install qmplot\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b84f1c1-9bae-40ad-a7ee-1a0a9cb9e82c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bc361b1-b15a-4113-a7e8-d3f13883c499",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"https://raw.githubusercontent.com/plotly/dash-bio-docs-files/master/manhattan_data.csv\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7be15b00-cad1-4b2d-b72e-0f52eff74c38",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.describe().transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ec5723a-d80c-416b-a49e-8a72f2e2c7a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from qmplot import manhattanplot\n",
    "\n",
    "ax = manhattanplot(\n",
    "    data=df,\n",
    "    chrom=\"CHR\",\n",
    "    pos=\"BP\",\n",
    "    pv=\"P\",\n",
    "    snp=\"SNP\",\n",
    "    xticklabel_kws={\"rotation\": \"vertical\"},\n",
    "    # suggestiveline=None,\n",
    "    # genomewideline=None,\n",
    "    # CHR=\"7\",  # single chromosome\n",
    "    sign_marker_p=1e-8,  # threshold for SNP annotation\n",
    "    is_annotate_topsnp=True,\n",
    ")\n",
    "\n",
    "plt.savefig(\"output_manhattan_plot.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eabb3c6-02cd-4d21-b0fb-7fc40c9095e0",
   "metadata": {},
   "source": [
    "## Volcano plot\n",
    "\n",
    "Volcano plots are a special case of scatterplots. They are typically used to display the results of a differential gene expression analysis in terms of statistical significance ($-log(P)$) vs. magnitude of change (fold change, effect size)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f8f8224-3d94-4623-9752-0241e565f52b",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"https://raw.githubusercontent.com/plotly/dash-bio-docs-files/master/volcano_data1.csv\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "802216e6-8bd2-4b17-a38c-b5e4a9de4609",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.describe().transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7eb78bef-5424-4ae6-bd19-084b1d4c0867",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"-log10P\"] = -np.log10(df[\"P\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f227a9dd-9737-4706-8c2b-e6f15094cf5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# determine \"top genes\" based on sensible thresholds\n",
    "condition = (df[\"EFFECTSIZE\"] >= 1) & (df[\"-log10P\"] > 7)\n",
    "# add a column\n",
    "df[\"signif\"] = \"ns\"\n",
    "df.loc[condition, \"signif\"] = \"Up\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90029ec1-030c-4932-ae22-8b15692bc257",
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = sns.scatterplot(\n",
    "    df,\n",
    "    x=\"EFFECTSIZE\",\n",
    "    y=\"-log10P\",\n",
    "    hue=\"signif\",\n",
    "    palette={\"Up\": \"red\", \"ns\": \"darkblue\"},  # optional palette\n",
    "    alpha=0.6,\n",
    "    legend=None,\n",
    ")\n",
    "\n",
    "ax.axvline(x=1, color=\"grey\", linestyle=\"--\")\n",
    "ax.axvline(x=-1, color=\"grey\", linestyle=\"--\")\n",
    "ax.axhline(y=7, color=\"grey\", linestyle=\"--\")\n",
    "\n",
    "df_top = df.loc[condition]\n",
    "for _, row in df_top.iterrows():\n",
    "    ax.text(row[\"EFFECTSIZE\"], row['-log10P']+0.1, row[\"GENE\"], fontsize=9)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c157b18b-bddb-4a48-837b-ad13ba1c8723",
   "metadata": {},
   "outputs": [],
   "source": [
    "# optional: automatic text repel with adjustText\n",
    "# pip install adjustText\n",
    "from adjustText import adjust_text\n",
    "\n",
    "ax = sns.scatterplot(\n",
    "    df,\n",
    "    x=\"EFFECTSIZE\",\n",
    "    y=\"-log10P\",\n",
    "    hue=\"signif\",\n",
    "    palette={\"Up\": \"red\", \"ns\": \"darkblue\"},  # optional palette\n",
    "    alpha=0.6,\n",
    "    legend=None,\n",
    ")\n",
    "\n",
    "ax.axvline(x=1, color=\"grey\", linestyle=\"--\")\n",
    "ax.axvline(x=-1, color=\"grey\", linestyle=\"--\")\n",
    "ax.axhline(y=7, color=\"grey\", linestyle=\"--\")\n",
    "\n",
    "df_top = df.loc[condition]\n",
    "texts = [ax.text(row[\"EFFECTSIZE\"], row['-log10P']+0.1, row[\"GENE\"], fontsize=9) for _, row in df_top.iterrows()]\n",
    "\n",
    "adjust_text(texts, arrowprops=dict(arrowstyle='-'));"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4d6bff2-36f9-4b35-83e0-0d0ce0a20ca5",
   "metadata": {},
   "source": [
    "## Clustergram: heatmap + dendrogram"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80aadc40-211e-4961-aa56-5ce8578d50dc",
   "metadata": {},
   "source": [
    "Clustergrams are typically used for gene expression data. The hierarchical clustering represented by the dendrograms can be used to identify groups of genes with related expression levels represented by the heatmap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40a0682a-2d6a-47ba-a9f0-7dddbe21f8ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"data/hbr_uhr_top_deg_normalized_counts.csv\", index_col=0)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7672b6e-ab5d-4c7f-948e-7e6cfb26644d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.describe().transpose().head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97705e4f-2256-4f60-a25d-71a88c93c8b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.clustermap(\n",
    "    df,\n",
    "    z_score=0,  # scale the rows by z-score (1 = scale the columns)\n",
    "    cmap=\"viridis\",  # custom color palette\n",
    "    figsize=(8, 8),\n",
    "    vmin=-1.5,  # min value on the colorbar\n",
    "    vmax=1.5,  # max value on the colorbar\n",
    "    cbar_kws=({\"label\": \"z score\"}),  # dict that specifies the title to the colorbar\n",
    "    cbar_pos=(0.855, 0.8, 0.025, 0.15),  # optional position of the colorbar axes in the figure (left, bottom, width, height)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0afa147f-fbe5-4a5c-9ef6-3fba49fc2338",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Enrichment analysis\n",
    "\n",
    "Suppose you have gene expression data for a cohort of patients. The data were collected before and after an intervention. You identified a bunch of differentially expressed genes (DEGs) between the two conditions.\n",
    "\n",
    "There are a lot of web tools that you can consider as query databases, such as:\n",
    "\n",
    "- [Database for Annotation, Visualization and Integrated Discovery (DAVID)](https://david.ncifcrf.gov/)\n",
    "- [Ingenuity Pathway Analysis (IPA)](https://digitalinsights.qiagen.com/products-overview/discovery-insights-portfolio/analysis-and-visualization/qiagen-ipa/)\n",
    "- [Metacore](https://clarivate.com/cortellis/solutions/early-research-intelligence-solutions/)\n",
    "- [PandaOmics](https://insilico.com/pandaomics)\n",
    "- [Enrichr](https://maayanlab.cloud/Enrichr/)\n",
    "- [GSEA](https://www.gsea-msigdb.org/gsea/index.jsp)\n",
    "- [STRING](https://string-db.org/)\n",
    "- [Panther](http://www.pantherdb.org/)\n",
    "- [Reactome](https://reactome.org/)\n",
    "- [KEGG](https://www.genome.jp/kegg/)\n",
    "\n",
    "...to name a few!\n",
    "\n",
    "Here we'll set up an enrichment pipeline based on two ingredients: Enrichr and [GSEApy](https://gseapy.readthedocs.io/en/latest/introduction.html).\n",
    "\n",
    "Enrichr is a user-friendly and free to use database and allows querying of gene lists against a repertoire of gene-set libraries. GSEApy is a Python wrapper for Enrichr, allowing users to query the characteristics of your DEGs within Python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f87f95ec-65df-433e-a433-378711acb2ca",
   "metadata": {},
   "source": [
    "To install GSEApy, run this command in the below code cell:\n",
    "\n",
    "```\n",
    "!pip install gseapy\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88cdca82-d678-411d-8cb6-aafabe43d40a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff28daaa-1c33-4ce8-bd48-9ba2b6d24a5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa364cea-a59e-45c9-a734-7762dd4fb9b2",
   "metadata": {},
   "source": [
    "Enrichr allows you to query across multiple databases. To understand the gene set libraries that are available for data query, you can type in the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6f66fd2-9aaf-4ed3-92d1-0ec0445bc703",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = gp.get_library_name()\n",
    "print(names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4229fa0d-97e0-400e-8aaf-bc25e0cd6c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(names)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "2a29a73e-6684-4009-a16b-d956470d9426",
   "metadata": {},
   "source": [
    "As a proof of concept, we will use the transcriptomics dataset published by [Zak et al., PNAS, 2012](https://www.pnas.org/content/109/50/E3503), examining how seropositive and seronegative subjects respond to the Ad5 vaccine across various time points.\n",
    "\n",
    "The fold change, ratio, p-value and adjusted p-values (q-value) are calculated with respect to baseline (timepoint=0). We will filter for \"up\" and \"down\" DEGs based on fold-change > 1.5, q-value < 0.05 and fold-change < -1.5, q-value < 0.05 respectively at day 1 post-vaccination using the commands as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "029493a2-c059-45bd-9c3d-305e51c21dd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"data/Ad5_seroneg.csv\", index_col=0)\n",
    "\n",
    "DEGs_up = (df[(df[\"fc_1d\"] > 1.5) & (df[\"qval_1d\"] < 0.05)]).index.tolist()\n",
    "DEGs_down = (df[(df[\"fc_1d\"] < -1.5) & (df[\"qval_1d\"] < 0.05)]).index.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e1da2e4-814b-4aa6-b277-d33657bcd40a",
   "metadata": {},
   "source": [
    "Very popular databases are Gene Ontology (GO) and Reactome. GO Biological Processes (GOBP), GO Molecular Functions (GOMF) and GO Cellular Components (GOCC) usually provide users a rough idea of what the DEGs do, and their localisation within the cell. To query the upregulated DEGs against all of these 4 databases, the following commands can be executed:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a40ee2fe-6b19-4348-9ec6-99087c86239a",
   "metadata": {},
   "outputs": [],
   "source": [
    "enr_GOBP_up = gp.enrichr(\n",
    "    gene_list=DEGs_up,\n",
    "    gene_sets=[\"GO_Biological_Process_2025\"],\n",
    "    organism=\"Human\",\n",
    "    outdir=\"out/enr_DEGs_GOBP_up\",\n",
    "    cutoff=0.05,\n",
    ")\n",
    "\n",
    "enr_GOMF_up = gp.enrichr(\n",
    "    gene_list=DEGs_up,\n",
    "    gene_sets=[\"GO_Molecular_Function_2025\"],\n",
    "    organism=\"Human\",\n",
    "    outdir=\"out/enr_DEGs_GOMF_up\",\n",
    "    cutoff=0.05,\n",
    ")\n",
    "\n",
    "enr_GOCC_up = gp.enrichr(\n",
    "    gene_list=DEGs_up,\n",
    "    gene_sets=[\"GO_Cellular_Component_2025\"],\n",
    "    organism=\"Human\",\n",
    "    outdir=\"out/enr_DEGs_GOCC_up\",\n",
    "    cutoff=0.05,\n",
    ")\n",
    "\n",
    "enr_Reactome_up = gp.enrichr(\n",
    "    gene_list=DEGs_up,\n",
    "    gene_sets=[\"Reactome_2022\"],\n",
    "    organism=\"Human\",\n",
    "    outdir=\"out/enr_DEGs_Reactome_up\",\n",
    "    cutoff=0.05,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a84298ba-d406-4325-a744-f6b5ceb9b3f0",
   "metadata": {},
   "source": [
    "Let's break these commands down individually to understand what they do.\n",
    "\n",
    "Firstly, `gp.enrichr` calls Enrichr to query against the gene list that is assigned to `DEGs_up`. We then specify the gene sets to query against (GOBP, GOMF, GOCC or Reactome), which is why the syntaxes are broken into 4 distinct function calls.\n",
    "Since the data is from Humans, we assign the organism as Human.\n",
    "Next we assign the output path with `outdir`. In this example, we specify a folder named `'out'` to store the output. Finally, we set a cutoff of $p=0.05$ for analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8fa14051-4adf-4f52-85e0-d902d374f0fe",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "enr_GOBP_up.results.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "938a134b-ee5b-48d2-9547-d4dd96a80cc9",
   "metadata": {},
   "source": [
    "The above table provides important statistics that allows you to have a quick sensing of the pathways that are enriched."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6da07c3-1ef3-4843-86f8-35bae9b650b4",
   "metadata": {},
   "source": [
    "A bar plot is among the most widely used methods to visualize enriched terms, depicting the enrichment scores (e.g. p-values) of the top terms."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e33f3b49-7952-4749-b86b-a72d38d434ea",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from gseapy.plot import barplot, dotplot\n",
    "\n",
    "barplot(enr_GOBP_up.res2d, title='GO BP seroneg day 1 (up)', color='r');"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "285bc1e7-a111-4d4d-a300-744817eafa23",
   "metadata": {},
   "source": [
    "The `barplot()` function plots the top 10 enriched molecular pathways with adjusted p-value < 0.05.\n",
    "\n",
    "To determine the enriched pathways in another database, for example Reactome:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2580ac51-8f9e-4a68-8065-e5959b2fc2c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "barplot(enr_Reactome_up.res2d, title='Reactome seroneg day 1 (up)', color='r');"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3637e5ba-77f1-4aec-af1b-94f84edf8b86",
   "metadata": {},
   "source": [
    "Dot plots are similar to bar plots with the capability to encode one additional score as dot size."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbbd7de3-f5ae-40da-a390-add6fe8fef3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "dotplot(enr_GOBP_up.res2d)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e57fd6d-ef9d-41fb-a6a7-ff77f395c77f",
   "metadata": {},
   "source": [
    "To save your figure, make sure that you pass a file name to `ofname`. The image format is automatically determined by the extension you use:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c4f8e6b-8697-428b-ae53-14a03d398883",
   "metadata": {},
   "outputs": [],
   "source": [
    "dotplot(enr_GOBP_up.res2d, ofname=\"dotplot_test.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8020fd99-4cb0-4a3f-942d-b9e68cb85dbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "dotplot(enr_GOBP_up.res2d, ofname=\"dotplot_test.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "309e5ed3",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Credits\n",
    "\n",
    "Partially abridged from \"Python for Scientific Computing\" course (Aalto Scientific Computing, 2020), [Top 50 matplotlib Visualizations – The Master Plots](https://www.machinelearningplus.com/plots/top-50-matplotlib-visualizations-the-master-plots-python/)), [Arthur Turrell](https://aeturrell.github.io/coding-for-economists/vis-common-plots.html), [Machine Learning Plus](https://www.machinelearningplus.com/plots/matplotlib-tutorial-complete-guide-python-plot-examples/), [matplotlib.org](https://matplotlib.org/tutorials/introductory/usage.html#sphx-glr-tutorials-introductory-usage-py).\n",
    "\n",
    "---\n",
    "\n",
    "# Further resources\n",
    "\n",
    "- [The Python graph gallery](https://www.python-graph-gallery.com/): a great site with 100s of charts made with Python using mostly Matplotlib and Seaborn, organized in multiple sections with their associated reproducible code.\n",
    "- [Matplotlib tutorials](https://matplotlib.org/stable/tutorials/index)\n",
    "- [Seaborn gallery](https://seaborn.pydata.org/examples/index.html)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
