{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "439f9c95-b36b-44a8-b6be-36a8a9f6cf1c",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "# Scientific Python and Data Manipulation\n",
    "## Advanced Python for Life Sciences @ Physalia courses (Summer 2025)\n",
    "### Marco Chierici, Fondazione Bruno Kessler"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8497383-3b12-4f1a-9f0e-f975b20681bf",
   "metadata": {},
   "source": [
    "# NumPy\n",
    "\n",
    "## Motivation\n",
    "\n",
    "Suppose you have to work with matrices and perform a matrix multiplication. You could create matrices in pure Python by using nested lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62f63adc-feb4-49dd-b097-451bfb97f109",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8707a33-7137-45ec-9e13-0faa21d4a7b0",
   "metadata": {},
   "source": [
    "This creates a 3x3 matrix that should look like this:\n",
    "\n",
    "```\n",
    "1 2 3\n",
    "4 5 6\n",
    "7 8 9\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f356c41-25bb-421e-86c6-c56dee45b44a",
   "metadata": {},
   "source": [
    "And then you can access the individual items by their indices."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a98f9b0-9177-4cf5-ba8a-66aab50f2cef",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix[0][1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3053d56c-a7df-4142-85ab-1c751b1ded18",
   "metadata": {},
   "source": [
    "How could you multiply each item by 2? For example, you can use a nested `for` loop."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd7e6444-54c5-47e6-ac8a-ae7acc16b96a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in matrix:\n",
    "    for i in range(len(row)):\n",
    "        row[i] = row[i] * 2\n",
    "\n",
    "matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e627891-ddef-410d-b21a-19bdddbcf829",
   "metadata": {},
   "source": [
    "The output is correct; the problem is that all of this is done in pure Python and you thus need to reimplement every operation you need to use!\n",
    "\n",
    "**NumPy** brings N-dimensional arrays and linear algebra routines to Python.\n",
    "\n",
    "More info and full documentation: https://numpy.org"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3e15cf9-2070-410e-b4e1-2afbc2507438",
   "metadata": {},
   "source": [
    "## Overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "3759c47d-b617-4c25-8dec-68523f1d66fa",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "# canonical import\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "cd517a18-f622-4d67-9263-e29b2da18251",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1 2 3]\n",
      " [4 5 6]\n",
      " [7 8 9]]\n"
     ]
    }
   ],
   "source": [
    "# a one-dimensional array\n",
    "a = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])\n",
    "print(a)\n",
    "#print(type(a))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fe20da0-731e-4f86-98e2-b66b2f8dc40c",
   "metadata": {},
   "source": [
    "You can index a Numpy array like you would to with a nested list:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "e7e3e37e-4347-4c8a-9db7-83e27e392b93",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a[0][1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a63ae07b-dfd9-4002-96d3-9a02548875d0",
   "metadata": {},
   "source": [
    "Or with the numpy-specific slicing syntax:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "e0595193-7096-4ffb-9476-2d59025639cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a[0, 1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1de64828-86b7-4aa4-a617-908956c24381",
   "metadata": {},
   "source": [
    "# Arrays contain items of a single type: this is one major difference with respect to lists."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b44dd951-5246-4122-9b09-06e86d583832",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array([[1, 2, 3], [\"a\", \"b\", \"c\"]])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24a3187b-a658-464f-b4c9-7ac7f527732f",
   "metadata": {},
   "source": [
    "You can create three-dimensional arrays:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "321609e2-7d4e-4335-ae23-7f390c75e1c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix = np.array([\n",
    "    [[1, 2, 3], [4, 5, 6]], \n",
    "    [[7, 8, 9], [10, 11, 12]],\n",
    "    [[13, 14, 15], [16, 17, 18]]\n",
    "])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "943b3e36-f764-4410-b326-97ef61664ff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix[0, 1, 2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a8e9b14-c543-47be-93c0-3dd81b6ef70c",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "source": [
    "## Operations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5e74fdb-62f5-45ee-9e45-d1bc2f65dd92",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])\n",
    "matrix * 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4b01abe-d03f-49a7-ba8c-8b7d835082ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "second_matrix = np.array([[5, 4, 3], [7, 6, 5], [9, 8, 7]])\n",
    "second_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb12b22a-84cd-4ae6-a390-daf1f26693ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "second_matrix - matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8667aa7d-a3b0-46ee-9549-3e18f6c8da7f",
   "metadata": {},
   "source": [
    "All arithmetic operators + - * / operate *element by element*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e79364d5-3b0d-46ef-945d-f252145e51ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix = np.array([[1, 1, 1], [1, 1, 1], [1, 1, 1]])\n",
    "matrix * matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20374acd-57d8-4d6e-82f8-f39487ab4cda",
   "metadata": {},
   "source": [
    "Matrix product is computed with the `@` operator:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bca469c-207a-4aef-9a20-da02c9c20fe1",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix @ matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2ea2d89-63ef-49e4-8648-61fbe71bf8d1",
   "metadata": {},
   "source": [
    "Other common array operations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dedf3742-ffe3-480d-bcce-df07fcb4f435",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [],
   "source": [
    "matrix = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])\n",
    "# dimension or shape\n",
    "print(matrix.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99f95edd-1e88-497a-b606-3b31b74e0d17",
   "metadata": {},
   "source": [
    "Note that `.shape` is an *attribute* of `matrix` and not a function or method: a common error is to call it like `matrix.shape()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c22264c-b697-4de6-87bf-33075af9803c",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.diagonal()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec66698a-cf61-47d6-a010-acc0d1338119",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.flatten()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9880138e-5dd2-4bf9-a445-82f448eb8cb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix.transpose()  # equivalent: matrix.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01dae182-46a2-47da-9f0d-2a1292437356",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(matrix)  # equivalent: matrix.max()\n",
    "# same for min(), sum(), mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ba35730-a86d-4c9e-b3db-d4c1b0999450",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.sqrt(matrix)  # element by element"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0032898e-35cb-46fc-85f9-a2670be72f17",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.square(matrix)  # element by element"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5319013d-7397-43f9-af5b-6c3bc7319d16",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.log2(matrix)  # element by element"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72a7c323-0711-445a-986a-6286432fafb7",
   "metadata": {},
   "source": [
    "## Ranges of values\n",
    "\n",
    "`np.arange()` is the Numpy equivalent to `range()` and returns a numpy array:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56ef953f-30e6-4b22-a8fd-8c03af7379da",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.arange(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2b39588-9abf-4171-8c75-354cb2a7cb29",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.arange(10, 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b820b8ae-64e2-431f-925c-e7136657f8e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.arange(10, 20, 2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d6dfdd9-3fff-4c72-ab1e-e9fc05d52edc",
   "metadata": {},
   "source": [
    "### Shaping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d27e321c-efe5-4804-9987-04fad33609d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1 2 3]\n",
      " [4 5 6]\n",
      " [7 8 9]]\n",
      "(3, 3)\n"
     ]
    }
   ],
   "source": [
    "A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])\n",
    "\n",
    "print(A)\n",
    "print(A.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1bfb5f83-b9b1-41d6-ad1b-665f67becf0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1, 2, 3, 4, 5, 6, 7, 8, 9])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "A.reshape(9, 1) # A.reshape(9,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d949666e-49b8-47ab-8e7d-5e5bdd7efbb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1 2 3]\n",
      " [4 5 6]]\n"
     ]
    }
   ],
   "source": [
    "arr = np.array([[1, 2, 3], [4, 5, 6]])\n",
    "print(arr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a5f0b86a-a57d-4df5-b859-ea46a7e21ec6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 2],\n",
       "       [3, 4],\n",
       "       [5, 6]])"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr.reshape(3, -1) # -1 MEANS AS MANY AS COLs NEEDED"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a76e2cb9-aea5-4e7c-a4ee-db87dd0b91ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[1, 2, 3, 4, 5, 6]])"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr.reshape(1, -1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e67e174-600e-4059-b984-693e5a27c9bb",
   "metadata": {},
   "source": [
    "## Randomness\n",
    "\n",
    "The `np.random` module allows you to perform random sampling."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0043f8b6-675b-40a0-af29-aecea145a49e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 1.74945474, -0.286073  , -0.48456513],\n",
       "       [-2.65331856, -0.00828463, -0.31963136],\n",
       "       [-0.53662936,  0.31540267,  0.42105072]])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# set a seed before generating random data\n",
    "np.random.seed(11)\n",
    "# draw samples from a normal distribution with 0 mean and 1 sd\n",
    "# and put the output in a 3x3 array\n",
    "np.random.normal(size=(3, 3))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5be752e0-66c0-4c50-9602-329f9afbaecd",
   "metadata": {},
   "source": [
    "## Resources\n",
    "\n",
    "- **[Numpy cheat sheet](https://images.datacamp.com/image/upload/v1676302459/Marketing/Blog/Numpy_Cheat_Sheet.pdf)** (also in the `resources` folder)\n",
    "- [Numpy tutorials and books](https://numpy.org/learn/)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e18c5d-c4e2-47cf-82d9-98a6301d61e0",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Scipy\n",
    "\n",
    "- Scientific Python\n",
    "- Built on top of NumPy\n",
    "- Provides more complex mathematic, statistical, and scientific data analysis functions\n",
    "  - Still, NumPy contains some linear algebra functions and Fourier transforms, even though these more properly belong in SciPy\n",
    "\n",
    "More info and full documentation: https://scipy.org\n",
    "\n",
    "Some particularly useful `scipy` sub-packages include:\n",
    "\n",
    "- `scipy.stats`\n",
    "  - randomness\n",
    "  - statistical functions and tests\n",
    "- `scipy.integrate`\n",
    "  - numerical integration\n",
    " \n",
    "Other useful sub-packages include `scipy.linalg` for linear algebra and `scipy.sparse` for sparse matrix problems (e.g. single-cell RNA-seq).\n",
    "\n",
    "We'll briefly touch here how to do simple statistical testing with `scipy`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7109083-34cb-4e27-86d9-4ff4dcf5e52c",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "## Statistical testing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30bf5b7e-7553-478a-a16f-a06216f71d85",
   "metadata": {},
   "source": [
    "`stats` contains functions for statistical hypothesis testing. For example, let's conduct a *paired t-test* to compare two sets of related measurements, such as the same biological parameter measured before and after a treatment in the same subjects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "c3063120-f8f2-4cf7-904b-2b808e900476",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats\n",
    "\n",
    "# set a seed before generating random data\n",
    "np.random.seed(12345)\n",
    "# assume these are two sets of biological measurements\n",
    "data_before = np.random.normal(loc=50, scale=10, size=100)  # before treatment\n",
    "# we manually shift data_before to simulate the effect of a treatment\n",
    "data_after = data_before + np.random.normal(loc=5, scale=5, size=100)  # after treatment\n",
    "\n",
    "# paired t-test\n",
    "t_statistic, p_value = stats.ttest_rel(data_before, data_after)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a20c5e23-43c6-4bd6-be2b-3d15cc157e7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.529722640391192e-16"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6da6fee-a8c0-4b89-bccf-13b264e3624f",
   "metadata": {},
   "source": [
    "The p-value is way less than the usual significance threshold of 0.05, so we reject the null hypothesis that there is no difference between the two sets of measurements.\n",
    "\n",
    "The *nonparametric version* of the paired t-test is the Wilcoxon signed-rank test: let's conduct this kind of test on the same data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c369cd0-376d-4eff-9699-e5fe5c0e3090",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Wilcoxon signed-rank test\n",
    "statistic, p_value = stats.wilcoxon(data_before, data_after)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f24f8c0-8bc7-45db-9b46-6d5f134010df",
   "metadata": {},
   "source": [
    "Again, this p-value indicates a statistically significant difference in this example data.\n",
    "\n",
    "Wilcoxon signed-rank is especially used instead of the paired t-test when the data are not normally distributed, thus they do not meet the assumptions of a t-test."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0f45c6a-2535-4420-9f22-9d99e7d5d12d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set a seed before generating random data\n",
    "np.random.seed(999)\n",
    "# non-normally distributed data from a log-normal distribution\n",
    "data_before_non_normal = np.random.lognormal(mean=1.5, sigma=0.4, size=30)\n",
    "data_after_non_normal = data_before_non_normal * np.random.lognormal(mean=0.1, sigma=0.4, size=30)\n",
    "\n",
    "# Performing the Wilcoxon signed-rank test\n",
    "statistic, p_value = stats.wilcoxon(data_before_non_normal, data_after_non_normal)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f43c85cc-ce70-47ad-9eaa-eea3fe9fe47c",
   "metadata": {},
   "source": [
    "If, instead, you want to assess statistical significance of *independent samples* (i.e., one group of patients vs. another group of different patients), you can use a t-test (`stats.ttest_ind`) or a Mann-Whitney U test (`stats.mannwhitneyu`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dc3fab2-f17f-422a-9dc0-d23547a21700",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "np.random.seed(999)\n",
    "\n",
    "# two independent non-normal samples from log-normal dists\n",
    "group_a = np.random.lognormal(mean=1.5, sigma=0.4, size=30)\n",
    "group_b = np.random.lognormal(mean=1.2, sigma=0.5, size=30)\n",
    "\n",
    "# perform the Mann–Whitney U test\n",
    "statistic, p_value = stats.mannwhitneyu(group_a, group_b, alternative='two-sided')\n",
    "p_value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ef67572-e6dd-4505-88f3-4163fac4956f",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "source": [
    "## Resources\n",
    "\n",
    "- **[Scipy cheat sheet](https://images.datacamp.com/image/upload/v1676303474/Marketing/Blog/SciPy_Cheat_Sheet.pdf)** (also in the `resources` folder)\n",
    "- [Scipy documentation](https://docs.scipy.org/doc/scipy/)\n",
    "- If you work a lot with statistical models and statistical tests, check out the [statsmodels](https://www.statsmodels.org/stable/index.html) library!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c7069aa-8b59-43ec-9a9e-17a787d4884d",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Pandas\n",
    "\n",
    "- \"The\" Python library for **data preprocessing and analysis**\n",
    "- Built **on top of Numpy**\n",
    "- Extremely versatile for manipulating datasets, mostly tabular data\n",
    "- Think of Pandas as the **evolution of spreadsheets**, with more capabilities for coding, and queries on relational data such as joins and group-by\n",
    "- Bonus: can be used for high quality **plots**\n",
    "- Most important structure: the **Data Frame** (R users: yes, that one!)\n",
    "- Trivia: stands for **Pan**el **Da**ta **S**ystem\n",
    "\n",
    "More info and full documentation: https://pandas.pydata.org"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ee5be65-105d-4ce1-a9d5-55da830fbb0d",
   "metadata": {},
   "source": [
    "## Resources\n",
    "We won't cover everything Pandas can do, so keep this **[cheat sheet](https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf)** handy! (it's also in the `resources` folder)\n",
    "\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f74d40e9-9e76-4bf5-8a26-f2ec45e14a69",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# canonical import\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "35efbec4-1873-45f6-8f5c-e6d0e4fbc827",
   "metadata": {},
   "outputs": [],
   "source": [
    "# we'll also need this to improve dealing with files, folders, and paths\n",
    "from pathlib import Path\n",
    "DATADIR = Path(\"data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d2329f6-104c-4a61-b270-b0e9707c55c6",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Series\n",
    "\n",
    "A Pandas Series is a 1D container, similar to the Python list. With respect to lists, however, a Series can only hold items of the same `dtype`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "847fd4f6-e270-4483-9679-ee5aa054d221",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GeneA    0.2\n",
       "GeneB    1.4\n",
       "GeneC    4.0\n",
       "GeneD    5.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a series using lists\n",
    "# note: index defaults to 0, 1, 2, ... if not given.\n",
    "genes = pd.Series([0.2, 1.4, 4, 5], index=[\"GeneA\", \"GeneB\", \"GeneC\", \"GeneD\"])\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c3a29a1b-c312-4f43-94ce-b76820900c52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.series.Series'>\n"
     ]
    }
   ],
   "source": [
    "print(type(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a83d2265-5614-4ba6-9f90-ff33adbba60e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4,)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the number of rows and columns\n",
    "# (shape is an attribute)\n",
    "genes.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "99b2c766-f4e5-4f87-9a85-c23a219ff593",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['GeneA', 'GeneB', 'GeneC', 'GeneD'], dtype='object')"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the index values (another attribute)\n",
    "genes.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c95ec844-2ab3-491b-99a3-b022f99cc10b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dtype('float64')"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the type\n",
    "genes.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6c4258ab-10b6-4bbf-8371-250bb08dbc61",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.series.Series'>\n",
      "Index: 4 entries, GeneA to GeneD\n",
      "Series name: None\n",
      "Non-Null Count  Dtype  \n",
      "--------------  -----  \n",
      "4 non-null      float64\n",
      "dtypes: float64(1)\n",
      "memory usage: 236.0+ bytes\n"
     ]
    }
   ],
   "source": [
    "# get more info\n",
    "genes.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d81c35e9-41f2-4883-9fb0-9614a4cd1d8b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GeneA    0.2\n",
      "GeneB    1.4\n",
      "GeneC    4.0\n",
      "GeneD    5.0\n",
      "GeneE    NaN\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "# create a series from a dictionary\n",
    "genes2 = pd.Series(\n",
    "    {\"GeneA\": 0.2, \"GeneB\": 1.4, \"GeneC\": 4.0, \"GeneD\": 5, \"GeneE\": np.nan}\n",
    ")\n",
    "print(genes2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c8e032e-d9df-4319-893a-3b9644383f99",
   "metadata": {},
   "source": [
    "### Missing values\n",
    "\n",
    "Checking for missing values is one of the main steps in your data preprocessing workflow. If you find missing data, you basically have three options:\n",
    "\n",
    "- keep it (easiest; depends on whether your downstream analysis methods can deal with NANs)\n",
    "- remove it (easy; potential loss of data; limits trained models for future data)\n",
    "- replace it (hardest, somewhat arbitrary; potential to save a lot of data for model training; potential to lead to false conclusions)\n",
    "\n",
    "Keep in mind that no approach applies to all circumstances!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55f2ad5b-f726-4c83-a47c-ac5bc09ddc9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for NA values\n",
    "print(genes2.isnull())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fb34b74-96de-46b7-95b7-1bed4d004293",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop NA values\n",
    "genes2.dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f31aeab7-c2b7-42c9-9065-c548f72802bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd9bd41-f98c-42a3-bc3a-9ef9461d5679",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop NA values inplace\n",
    "genes3 = genes2.copy()  # creates a safe \"deep copy\"\n",
    "genes3.dropna(inplace=True)\n",
    "genes3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10172ed1-3d02-4d9f-b0ef-d07dc64a1514",
   "metadata": {},
   "source": [
    "**Note:** many Pandas functions and methods that alter dataframes support the parameter `inplace`, which by default is always `False`. In this case, the function returns `None` and the input dataframe is overwritten. In general, *you should avoid using `inplace=True`*. Moreover, this parameter could be deprecated in a future Pandas version (see [here](https://github.com/pandas-dev/pandas/issues/16529))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25d0708f-ba71-42de-9f6e-fffd7f2de481",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace NA values with a custom value\n",
    "genes2.fillna(value=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ec19564-fd69-4da4-87ed-f4155c62a35d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace NA values with a custom value\n",
    "genes2.fillna(value=genes2.median())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30662e2d-f576-4017-a12f-16e7fa78f244",
   "metadata": {},
   "outputs": [],
   "source": [
    "# replace NA values propagating the last valid observation to next valid\n",
    "genes2.ffill()  # use this instead of genes2.fillna(method=\"ffill\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3973ee9-efac-4727-a06d-4dee52219dc4",
   "metadata": {},
   "source": [
    "So what to do with missing values? Here are some rules of thumb.\n",
    "\n",
    "- You should **drop values** when a lot of data is missing;\n",
    "- You should **fill with the same value** if you know that NaN is just a placeholder (e.g., for 0);\n",
    "- You should **fill with interpolated or estimated value** if there is a reasonable assumption to do that!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b42ad4d-8ab3-4ec5-a8cb-1f7ef7a6baa2",
   "metadata": {},
   "source": [
    "### Selecting/Filtering Values in a series\n",
    "\n",
    "1. What is the value of GeneC? There are multiple options to access GeneC."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ea7aa34-b1bd-43b3-842e-24af8c22a431",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dot notation\n",
    "genes.GeneC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd24ee4d-ec2f-472a-92ea-4402e9c8d42d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# by row name\n",
    "genes[\"GeneC\"] # space safe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9621712-5036-4188-aa91-087178fed3a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes.loc[\"GeneC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bdc31cd4-27b1-4d4d-b76d-1db433e73fa1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# by row index\n",
    "genes.iloc[2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a53e769a-b316-400f-aea1-d7593041f054",
   "metadata": {},
   "source": [
    "2. What is the value of GeneC and GeneD?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "112f03f3-87a2-4ddf-8a94-44eab42819c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes[[\"GeneC\", \"GeneD\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "788115ac-d064-407a-be95-17c763bc073f",
   "metadata": {},
   "source": [
    "3. What genes are expressed with a value of at least 3?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e6e8050-d3ec-41cf-80e7-24653aecf58a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes[genes > 3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "96fd400d-99b8-4a05-8188-1cca4aa0aa5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GeneA    0.2\n",
       "GeneC    4.0\n",
       "GeneD    5.0\n",
       "dtype: float64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# working with multiple conditions\n",
    "genes[(genes > 3) | (genes < 1)] # | OR, & AND.\n",
    "# round breakets are important!!!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2580fb87-4672-485b-8ce0-88accb8c3647",
   "metadata": {},
   "source": [
    "- In Pandas, the Boolean negation operator is the tilde `~`.\n",
    "- Use the `.isin()` method to select data whose value \"is in\" a list of values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c88dd475-4da3-473b-98a6-98cfe34c8ceb",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes.isin([5.0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5e4b05a-6d60-4101-a0fd-d97e1627e751",
   "metadata": {},
   "source": [
    "4. What is the mean expression of the whole data?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "562949f1-3e5f-4eb5-8235-864c04364ea3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.65"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86a93c9f-0029-4d7c-88ab-ff79fd76bbc8",
   "metadata": {},
   "source": [
    "5. What is the largest value?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "619dcb9d-8eea-4a9c-901f-15b50fce2928",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5.0"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.max()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38247914-7f65-4a0a-97fe-b71f5e61d005",
   "metadata": {},
   "source": [
    "6. What gene has the largest value?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f9f53727-e9fe-43a9-9b05-68c90d3b5607",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GeneD'"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes[genes == genes.max()].index[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ba007c9-78de-40b7-9b8b-1f2d86b00fd0",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# alternative\n",
    "genes.idxmax()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35775f99-3cf7-4c7b-bb04-5f4fc3d38b61",
   "metadata": {},
   "source": [
    "### Sorting values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "fb81fcb2-c637-4e17-af43-3bdd8dbbce3c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GeneD    5.0\n",
       "GeneC    4.0\n",
       "GeneB    1.4\n",
       "GeneA    0.2\n",
       "dtype: float64"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sort from highest to lowest\n",
    "# note: default is increasing order! check ?genes.sort_values\n",
    "genes.sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f7ab4f1d-0e6a-46d9-9859-9ca1e954516b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on method sort_values in module pandas.core.series:\n",
      "\n",
      "sort_values(*, axis: 'Axis' = 0, ascending: 'bool | Sequence[bool]' = True, inplace: 'bool' = False, kind: 'SortKind' = 'quicksort', na_position: 'NaPosition' = 'last', ignore_index: 'bool' = False, key: 'ValueKeyFunc | None' = None) -> 'Series | None' method of pandas.core.series.Series instance\n",
      "    Sort by the values.\n",
      "\n",
      "    Sort a Series in ascending or descending order by some\n",
      "    criterion.\n",
      "\n",
      "    Parameters\n",
      "    ----------\n",
      "    axis : {0 or 'index'}\n",
      "        Unused. Parameter needed for compatibility with DataFrame.\n",
      "    ascending : bool or list of bools, default True\n",
      "        If True, sort values in ascending order, otherwise descending.\n",
      "    inplace : bool, default False\n",
      "        If True, perform operation in-place.\n",
      "    kind : {'quicksort', 'mergesort', 'heapsort', 'stable'}, default 'quicksort'\n",
      "        Choice of sorting algorithm. See also :func:`numpy.sort` for more\n",
      "        information. 'mergesort' and 'stable' are the only stable  algorithms.\n",
      "    na_position : {'first' or 'last'}, default 'last'\n",
      "        Argument 'first' puts NaNs at the beginning, 'last' puts NaNs at\n",
      "        the end.\n",
      "    ignore_index : bool, default False\n",
      "        If True, the resulting axis will be labeled 0, 1, …, n - 1.\n",
      "    key : callable, optional\n",
      "        If not None, apply the key function to the series values\n",
      "        before sorting. This is similar to the `key` argument in the\n",
      "        builtin :meth:`sorted` function, with the notable difference that\n",
      "        this `key` function should be *vectorized*. It should expect a\n",
      "        ``Series`` and return an array-like.\n",
      "\n",
      "    Returns\n",
      "    -------\n",
      "    Series or None\n",
      "        Series ordered by values or None if ``inplace=True``.\n",
      "\n",
      "    See Also\n",
      "    --------\n",
      "    Series.sort_index : Sort by the Series indices.\n",
      "    DataFrame.sort_values : Sort DataFrame by the values along either axis.\n",
      "    DataFrame.sort_index : Sort DataFrame by indices.\n",
      "\n",
      "    Examples\n",
      "    --------\n",
      "    >>> s = pd.Series([np.nan, 1, 3, 10, 5])\n",
      "    >>> s\n",
      "    0     NaN\n",
      "    1     1.0\n",
      "    2     3.0\n",
      "    3     10.0\n",
      "    4     5.0\n",
      "    dtype: float64\n",
      "\n",
      "    Sort values ascending order (default behaviour)\n",
      "\n",
      "    >>> s.sort_values(ascending=True)\n",
      "    1     1.0\n",
      "    2     3.0\n",
      "    4     5.0\n",
      "    3    10.0\n",
      "    0     NaN\n",
      "    dtype: float64\n",
      "\n",
      "    Sort values descending order\n",
      "\n",
      "    >>> s.sort_values(ascending=False)\n",
      "    3    10.0\n",
      "    4     5.0\n",
      "    2     3.0\n",
      "    1     1.0\n",
      "    0     NaN\n",
      "    dtype: float64\n",
      "\n",
      "    Sort values putting NAs first\n",
      "\n",
      "    >>> s.sort_values(na_position='first')\n",
      "    0     NaN\n",
      "    1     1.0\n",
      "    2     3.0\n",
      "    4     5.0\n",
      "    3    10.0\n",
      "    dtype: float64\n",
      "\n",
      "    Sort a series of strings\n",
      "\n",
      "    >>> s = pd.Series(['z', 'b', 'd', 'a', 'c'])\n",
      "    >>> s\n",
      "    0    z\n",
      "    1    b\n",
      "    2    d\n",
      "    3    a\n",
      "    4    c\n",
      "    dtype: object\n",
      "\n",
      "    >>> s.sort_values()\n",
      "    3    a\n",
      "    1    b\n",
      "    4    c\n",
      "    2    d\n",
      "    0    z\n",
      "    dtype: object\n",
      "\n",
      "    Sort using a key function. Your `key` function will be\n",
      "    given the ``Series`` of values and should return an array-like.\n",
      "\n",
      "    >>> s = pd.Series(['a', 'B', 'c', 'D', 'e'])\n",
      "    >>> s.sort_values()\n",
      "    1    B\n",
      "    3    D\n",
      "    0    a\n",
      "    2    c\n",
      "    4    e\n",
      "    dtype: object\n",
      "    >>> s.sort_values(key=lambda x: x.str.lower())\n",
      "    0    a\n",
      "    1    B\n",
      "    2    c\n",
      "    3    D\n",
      "    4    e\n",
      "    dtype: object\n",
      "\n",
      "    NumPy ufuncs work well here. For example, we can\n",
      "    sort by the ``sin`` of the value\n",
      "\n",
      "    >>> s = pd.Series([-4, -2, 0, 2, 4])\n",
      "    >>> s.sort_values(key=np.sin)\n",
      "    1   -2\n",
      "    4    4\n",
      "    2    0\n",
      "    0   -4\n",
      "    3    2\n",
      "    dtype: int64\n",
      "\n",
      "    More complicated user-defined functions can be used,\n",
      "    as long as they expect a Series and return an array-like\n",
      "\n",
      "    >>> s.sort_values(key=lambda x: (np.tan(x.cumsum())))\n",
      "    0   -4\n",
      "    3    2\n",
      "    4    4\n",
      "    1   -2\n",
      "    2    0\n",
      "    dtype: int64\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(genes.sort_values) # access documentation of a function"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4b665fa-e086-4c58-942b-6d014712d33b",
   "metadata": {},
   "source": [
    "### Replacing values\n",
    "\n",
    "If you have a categorical (factor) Series, it may be convenient to rename its levels. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "799a6018-21c7-496a-b968-23160ec2e45e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    M\n",
       "1    F\n",
       "2    M\n",
       "3    F\n",
       "4    M\n",
       "5    F\n",
       "Name: sex, dtype: object"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dummy pandas series with a categorical variable (levels: \"M\", \"F\")\n",
    "sex = pd.Series([\"M\", \"F\", \"M\", \"F\", \"M\", \"F\"], name=\"sex\")\n",
    "sex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "97b88d03-b4d6-46d3-a27b-4511266856bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      Male\n",
       "1    Female\n",
       "2      Male\n",
       "3    Female\n",
       "4      Male\n",
       "5    Female\n",
       "Name: sex, dtype: object"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# replace \"M\" and \"F\" with \"Male\" and \"Female\"\n",
    "sex.replace(to_replace=[\"M\", \"F\"], value=[\"Male\", \"Female\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a7cd089b-8006-4dde-9af6-e988f072ca80",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      Male\n",
       "1    Female\n",
       "2      Male\n",
       "3    Female\n",
       "4      Male\n",
       "5    Female\n",
       "Name: sex, dtype: object"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sex.map({\"M\": \"Male\", \"F\": \"Female\"})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "80491ccc-83d2-45dd-b8cd-0afdcffc98ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_37657/730481490.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  sex.replace(to_replace=[\"M\", \"F\"], value=[0, 1])\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0    0\n",
       "1    1\n",
       "2    0\n",
       "3    1\n",
       "4    0\n",
       "5    1\n",
       "Name: sex, dtype: int64"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sex.replace(to_replace=[\"M\", \"F\"], value=[0, 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b0775312-fe34-4cae-91a1-5e596e9db600",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    0\n",
       "1    1\n",
       "2    0\n",
       "3    1\n",
       "4    0\n",
       "5    1\n",
       "Name: sex, dtype: int64"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sex.map({\"M\": 0, \"F\": 1})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "f7a12495-993b-426b-911b-a9da476880af",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    False\n",
       "1     True\n",
       "2    False\n",
       "3     True\n",
       "4    False\n",
       "5     True\n",
       "Name: sex, dtype: bool"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sex.isin(['F']) # pass a list to isin()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4983c8ba-adda-4070-9f1d-988ae05a47dc",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "97084d04-5977-437b-9bf2-76d2f1d499c7",
   "metadata": {},
   "source": [
    "## DataFrames\n",
    "\n",
    "A Pandas DataFrame can be thought of as a collection (or a dictionary) of Series objects. The keys in this dictionary are the column names and the values are the Series.\n",
    "\n",
    "To create a DataFrame from scratch, there are multiple possibilities: we decide to create a dictionary first, then convert it to a DataFrame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "eadabe43-4e99-4cf5-8817-9ca2baafddd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = {\n",
    "    \"EnsemblID\": [\n",
    "        \"ENSG00000223972\",\n",
    "        \"ENSG00000227232\",\n",
    "        \"ENSG00000243485\",\n",
    "        \"ENSG00000237613\",\n",
    "        \"ENSG00000268020\",\n",
    "        \"ENSG00000186092\",\n",
    "    ],\n",
    "    \"Gene\": [\"DDX11L1\", \"WASH7P\", \"MIR1302-11\", \"FAM138A\", \"OR4G4P\", \"OR4F5\"],\n",
    "    \"GTEX-1117F\": [0.1082, 21.4, 0.1602, 0.05045, 0, 0],\n",
    "    \"GTEX-111CU\": [0.1158, 11.03, 0.06433, 0, 0, 0],\n",
    "    \"GTEX-111FC\": [0.02104, 16.75, 0.04674, 0.02945, 0, 0],\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "2a07d5a5-5413-4b45-9baf-3fd7dc7bd06a",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EnsemblID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.11580</td>\n",
       "      <td>0.02104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>11.03000</td>\n",
       "      <td>16.75000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.06433</td>\n",
       "      <td>0.04674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.02945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         EnsemblID        Gene  GTEX-1117F  GTEX-111CU  GTEX-111FC\n",
       "0  ENSG00000223972     DDX11L1     0.10820     0.11580     0.02104\n",
       "1  ENSG00000227232      WASH7P    21.40000    11.03000    16.75000\n",
       "2  ENSG00000243485  MIR1302-11     0.16020     0.06433     0.04674\n",
       "3  ENSG00000237613     FAM138A     0.05045     0.00000     0.02945\n",
       "4  ENSG00000268020      OR4G4P     0.00000     0.00000     0.00000\n",
       "5  ENSG00000186092       OR4F5     0.00000     0.00000     0.00000"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create dataframe\n",
    "pd.DataFrame(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "id": "bed9888d-de27-4ab8-8bd2-acb1c9970b10",
   "metadata": {
    "slideshow": {
     "slide_type": "slide"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>EnsemblID</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>ENSG00000223972</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.11580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>ENSG00000227232</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>ENSG00000243485</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>ENSG00000237613</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>ENSG00000268020</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>ENSG00000186092</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Gene        EnsemblID  GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "0     DDX11L1  ENSG00000223972     0.10820     0.02104     0.11580\n",
       "1      WASH7P  ENSG00000227232    21.40000    16.75000    11.03000\n",
       "2  MIR1302-11  ENSG00000243485     0.16020     0.04674     0.06433\n",
       "3     FAM138A  ENSG00000237613     0.05045     0.02945     0.00000\n",
       "4      OR4G4P  ENSG00000268020     0.00000     0.00000     0.00000\n",
       "5       OR4F5  ENSG00000186092     0.00000     0.00000     0.00000"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create and re-order column names\n",
    "df_gene = pd.DataFrame(\n",
    "    data, columns=[\"Gene\", \"EnsemblID\", \"GTEX-1117F\", \"GTEX-111FC\", \"GTEX-111CU\"]\n",
    ")\n",
    "df_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "3b7402c6-4d2a-4600-8703-18e908aa70cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    }
   ],
   "source": [
    "print(type(df_gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "0d423d4b-7c45-4ef0-a10b-95ffac0aa289",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, 5)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the number of rows and columns\n",
    "# (shape is an attribute of df)\n",
    "df_gene.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "676673db-0646-4e3e-b86d-f9ec2e77dc9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Gene', 'EnsemblID', 'GTEX-1117F', 'GTEX-111FC', 'GTEX-111CU'], dtype='object')"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the column names (another attribute)\n",
    "df_gene.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "d017b449-1887-4789-b0eb-fe5282696f72",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RangeIndex(start=0, stop=6, step=1)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the index (another attribute)\n",
    "df_gene.index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25e89bed-b90c-4f1a-9b8c-9a88a46f7332",
   "metadata": {},
   "source": [
    "Note that column names and index have their own types `Index` and `RangeIndex` - it is not a Python list (even though it looks like a list)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "68a5edcd-6ec7-4327-8e96-1cc684dcb8b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.11580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                                      \n",
       "ENSG00000223972     DDX11L1     0.10820     0.02104     0.11580\n",
       "ENSG00000227232      WASH7P    21.40000    16.75000    11.03000\n",
       "ENSG00000243485  MIR1302-11     0.16020     0.04674     0.06433\n",
       "ENSG00000237613     FAM138A     0.05045     0.02945     0.00000\n",
       "ENSG00000268020      OR4G4P     0.00000     0.00000     0.00000\n",
       "ENSG00000186092       OR4F5     0.00000     0.00000     0.00000"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a new index (original dataframe is not modified)\n",
    "df_gene.set_index(\"EnsemblID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "ad412152-2bce-4a61-95a8-c73b1dbf5c03",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.11580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                                      \n",
       "ENSG00000223972     DDX11L1     0.10820     0.02104     0.11580\n",
       "ENSG00000227232      WASH7P    21.40000    16.75000    11.03000\n",
       "ENSG00000243485  MIR1302-11     0.16020     0.04674     0.06433\n",
       "ENSG00000237613     FAM138A     0.05045     0.02945     0.00000\n",
       "ENSG00000268020      OR4G4P     0.00000     0.00000     0.00000\n",
       "ENSG00000186092       OR4F5     0.00000     0.00000     0.00000"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a new index in-place (mind the above caveat about the use of inplace)\n",
    "df_gene.set_index(\"EnsemblID\", inplace=True)\n",
    "df_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "16ab9bff-9fea-437a-b271-bb5369d4e053",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene           object\n",
       "GTEX-1117F    float64\n",
       "GTEX-111FC    float64\n",
       "GTEX-111CU    float64\n",
       "dtype: object"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the type of each column\n",
    "df_gene.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3972386b-d2fc-427f-b428-c2e18970c7c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "Index: 6 entries, ENSG00000223972 to ENSG00000186092\n",
      "Data columns (total 4 columns):\n",
      " #   Column      Non-Null Count  Dtype  \n",
      "---  ------      --------------  -----  \n",
      " 0   Gene        6 non-null      object \n",
      " 1   GTEX-1117F  6 non-null      float64\n",
      " 2   GTEX-111FC  6 non-null      float64\n",
      " 3   GTEX-111CU  6 non-null      float64\n",
      "dtypes: float64(3), object(1)\n",
      "memory usage: 240.0+ bytes\n"
     ]
    }
   ],
   "source": [
    "# get more info\n",
    "df_gene.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "7ee08682-6d46-4562-89b0-71dc34e9c47f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>EnsemblID</th>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <th>ENSG00000186092</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <td>0.1082</td>\n",
       "      <td>21.4</td>\n",
       "      <td>0.1602</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <td>0.02104</td>\n",
       "      <td>16.75</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111CU</th>\n",
       "      <td>0.1158</td>\n",
       "      <td>11.03</td>\n",
       "      <td>0.06433</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "EnsemblID  ENSG00000223972 ENSG00000227232 ENSG00000243485 ENSG00000237613  \\\n",
       "Gene               DDX11L1          WASH7P      MIR1302-11         FAM138A   \n",
       "GTEX-1117F          0.1082            21.4          0.1602         0.05045   \n",
       "GTEX-111FC         0.02104           16.75         0.04674         0.02945   \n",
       "GTEX-111CU          0.1158           11.03         0.06433             0.0   \n",
       "\n",
       "EnsemblID  ENSG00000268020 ENSG00000186092  \n",
       "Gene                OR4G4P           OR4F5  \n",
       "GTEX-1117F             0.0             0.0  \n",
       "GTEX-111FC             0.0             0.0  \n",
       "GTEX-111CU             0.0             0.0  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transpose v1\n",
    "df_gene.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "9eb52aa2-56a0-4732-aa32-119fc96f0d1e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>EnsemblID</th>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <th>ENSG00000186092</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Gene</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <td>0.1082</td>\n",
       "      <td>21.4</td>\n",
       "      <td>0.1602</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <td>0.02104</td>\n",
       "      <td>16.75</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-111CU</th>\n",
       "      <td>0.1158</td>\n",
       "      <td>11.03</td>\n",
       "      <td>0.06433</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "EnsemblID  ENSG00000223972 ENSG00000227232 ENSG00000243485 ENSG00000237613  \\\n",
       "Gene               DDX11L1          WASH7P      MIR1302-11         FAM138A   \n",
       "GTEX-1117F          0.1082            21.4          0.1602         0.05045   \n",
       "GTEX-111FC         0.02104           16.75         0.04674         0.02945   \n",
       "GTEX-111CU          0.1158           11.03         0.06433             0.0   \n",
       "\n",
       "EnsemblID  ENSG00000268020 ENSG00000186092  \n",
       "Gene                OR4G4P           OR4F5  \n",
       "GTEX-1117F             0.0             0.0  \n",
       "GTEX-111FC             0.0             0.0  \n",
       "GTEX-111CU             0.0             0.0  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# transpose v2\n",
    "df_gene.transpose()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74ae113d-86e6-482e-8ecc-0357aebadcd8",
   "metadata": {},
   "source": [
    "Quick summary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "510a1f8f-f020-4cf5-9033-88f707d7cc86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>6.000000</td>\n",
       "      <td>6.000000</td>\n",
       "      <td>6.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>3.619808</td>\n",
       "      <td>2.807872</td>\n",
       "      <td>1.868355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>8.710704</td>\n",
       "      <td>6.830244</td>\n",
       "      <td>4.488517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.012613</td>\n",
       "      <td>0.005260</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.079325</td>\n",
       "      <td>0.025245</td>\n",
       "      <td>0.032165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.147200</td>\n",
       "      <td>0.042417</td>\n",
       "      <td>0.102932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>21.400000</td>\n",
       "      <td>16.750000</td>\n",
       "      <td>11.030000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "count    6.000000    6.000000    6.000000\n",
       "mean     3.619808    2.807872    1.868355\n",
       "std      8.710704    6.830244    4.488517\n",
       "min      0.000000    0.000000    0.000000\n",
       "25%      0.012613    0.005260    0.000000\n",
       "50%      0.079325    0.025245    0.032165\n",
       "75%      0.147200    0.042417    0.102932\n",
       "max     21.400000   16.750000   11.030000"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "be79ae00-0fc6-421d-9566-7c08397b6188",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>EnsemblID</th>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <th>ENSG00000186092</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "EnsemblID ENSG00000223972 ENSG00000227232 ENSG00000243485 ENSG00000237613  \\\n",
       "count                   4               4               4               4   \n",
       "unique                  4               4               4               4   \n",
       "top               DDX11L1          WASH7P      MIR1302-11         FAM138A   \n",
       "freq                    1               1               1               1   \n",
       "\n",
       "EnsemblID  ENSG00000268020  ENSG00000186092  \n",
       "count                  4.0              4.0  \n",
       "unique                 2.0              2.0  \n",
       "top                    0.0              0.0  \n",
       "freq                   3.0              3.0  "
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.T.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24aeab58-33ee-4baf-bc5d-3e489eac4edb",
   "metadata": {},
   "source": [
    "To convert a dataframe to a numpy array, just use the `.values` attribute:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "6beb7595-ea7f-4fed-a85d-b74992118935",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([['DDX11L1', 0.1082, 0.02104, 0.1158],\n",
       "       ['WASH7P', 21.4, 16.75, 11.03],\n",
       "       ['MIR1302-11', 0.1602, 0.04674, 0.06433],\n",
       "       ['FAM138A', 0.05045, 0.02945, 0.0],\n",
       "       ['OR4G4P', 0.0, 0.0, 0.0],\n",
       "       ['OR4F5', 0.0, 0.0, 0.0]], dtype=object)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9745201a-b443-478f-a944-bca3af83cfc8",
   "metadata": {},
   "source": [
    "### Select columns: by name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "12c1d72b-72bc-4553-be29-8870eff5eee1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EnsemblID\n",
       "ENSG00000223972       DDX11L1\n",
       "ENSG00000227232        WASH7P\n",
       "ENSG00000243485    MIR1302-11\n",
       "ENSG00000237613       FAM138A\n",
       "ENSG00000268020        OR4G4P\n",
       "ENSG00000186092         OR4F5\n",
       "Name: Gene, dtype: object"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# single column (returns a Series)\n",
    "df_gene[\"Gene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ea5d910b-8abe-4f99-a917-df44cd34ead4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F\n",
       "EnsemblID                              \n",
       "ENSG00000223972     DDX11L1     0.10820\n",
       "ENSG00000227232      WASH7P    21.40000\n",
       "ENSG00000243485  MIR1302-11     0.16020\n",
       "ENSG00000237613     FAM138A     0.05045\n",
       "ENSG00000268020      OR4G4P     0.00000\n",
       "ENSG00000186092       OR4F5     0.00000"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# multicolumn (returns a DataFrame)\n",
    "df_gene[[\"Gene\", \"GTEX-1117F\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "486b6d52-de58-4cd0-9a66-d40dfdde9a8b",
   "metadata": {},
   "source": [
    "### Select columns: by condition\n",
    "\n",
    "Often you'll have to select (or filter out) columns based on a pattern: e.g., all those columns starting with, ending with, or containing something.\n",
    "\n",
    "For this, you can manipulate column names using methods from the built-in `str` Python module: for example, `str.startswith()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "09177e5e-b26e-43f2-b6cc-3cd1fec5439c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([False,  True,  True,  True])"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select all columns starting with GTEX\n",
    "condition = df_gene.columns.str.startswith(\"GTEX\") # startswith() works through str()\n",
    "condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "c35caa26-cdeb-4539-a676-baaad9a2dc5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.11580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                          \n",
       "ENSG00000223972     0.10820     0.02104     0.11580\n",
       "ENSG00000227232    21.40000    16.75000    11.03000\n",
       "ENSG00000243485     0.16020     0.04674     0.06433\n",
       "ENSG00000237613     0.05045     0.02945     0.00000\n",
       "ENSG00000268020     0.00000     0.00000     0.00000\n",
       "ENSG00000186092     0.00000     0.00000     0.00000"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene[df_gene.columns[condition]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "16e15d98-cd89-413e-adde-679c2a855e2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.11580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                          \n",
       "ENSG00000223972     0.10820     0.02104     0.11580\n",
       "ENSG00000227232    21.40000    16.75000    11.03000\n",
       "ENSG00000243485     0.16020     0.04674     0.06433\n",
       "ENSG00000237613     0.05045     0.02945     0.00000\n",
       "ENSG00000268020     0.00000     0.00000     0.00000\n",
       "ENSG00000186092     0.00000     0.00000     0.00000"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene[df_gene.columns[condition]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54e85c06-78bd-4b68-9dd2-3f494fe6b8be",
   "metadata": {},
   "source": [
    "### Select rows: by name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87c20140-70d6-491e-98dc-7420ee732a0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the 1st row\n",
    "df_gene.loc[\"ENSG00000223972\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "7966ffa3-2818-417d-8b7a-4846fc125049",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  Gene  GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                                 \n",
       "ENSG00000186092  OR4F5         0.0         0.0         0.0"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get the last row\n",
    "# df_gene.iloc[-1]\n",
    "df_gene.tail(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9edf05d-b2e8-4f4f-93cd-d3368c3b3460",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset multiple rows by using a list\n",
    "df_gene.loc[[\"ENSG00000223972\", \"ENSG00000243485\"]]  # output: dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9ab6a07-8a74-4088-b6c1-7be650427e82",
   "metadata": {},
   "source": [
    "### Select rows: by condition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "2502baa9-60a9-4a7f-8e7d-3410cd93b81b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.4</td>\n",
       "      <td>16.75</td>\n",
       "      <td>11.03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Gene  GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                                  \n",
       "ENSG00000227232  WASH7P        21.4       16.75       11.03"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cond = df_gene[\"GTEX-1117F\"] >= 5.\n",
    "df_gene.loc[cond]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41ff606b-0299-44fc-b865-719c3ec25857",
   "metadata": {},
   "source": [
    "### Select rows: by index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22ee305a-974c-4b4f-89f0-fcd85360c359",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene.iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8848e1c1-f30f-443c-9aed-4878ddce09b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene.iloc[-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14746c4a-f33e-4625-a4dc-1a70cd17a926",
   "metadata": {},
   "source": [
    "### Select rows & columns\n",
    "\n",
    "You can use `.loc[]` and `.iloc[]` to select columns too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "id": "77d3605d-7881-4b8a-bbdc-fb980cef73da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.11580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>21.40000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GTEX-1117F  GTEX-111CU\n",
       "0     0.10820     0.11580\n",
       "1    21.40000    11.03000\n",
       "2     0.16020     0.06433\n",
       "3     0.05045     0.00000\n",
       "4     0.00000     0.00000\n",
       "5     0.00000     0.00000"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.loc[:, [\"GTEX-1117F\", \"GTEX-111CU\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "979aa828-6c6c-4a11-be55-054dfc292a71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GTEX-1117F  GTEX-111FC\n",
       "EnsemblID                              \n",
       "ENSG00000223972     0.10820     0.02104\n",
       "ENSG00000227232    21.40000    16.75000\n",
       "ENSG00000243485     0.16020     0.04674\n",
       "ENSG00000237613     0.05045     0.02945\n",
       "ENSG00000268020     0.00000     0.00000\n",
       "ENSG00000186092     0.00000     0.00000"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# same as above but using iloc\n",
    "df_gene.iloc[:, [1, -1]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af0a24fb-4ca8-4372-a361-8556d3783a93",
   "metadata": {},
   "source": [
    "You can select contiguous ranges as well:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "e1ddee57-c0d0-47d8-a5fa-dcbe093e497e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.11580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                          \n",
       "ENSG00000223972     0.10820     0.02104     0.11580\n",
       "ENSG00000227232    21.40000    16.75000    11.03000\n",
       "ENSG00000243485     0.16020     0.04674     0.06433\n",
       "ENSG00000237613     0.05045     0.02945     0.00000\n",
       "ENSG00000268020     0.00000     0.00000     0.00000\n",
       "ENSG00000186092     0.00000     0.00000     0.00000"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.loc[:, \"GTEX-1117F\":\"GTEX-111CU\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f5b22a-709a-4f41-8d7a-860c646711fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene.iloc[:, range(1, 4)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "9938082d-9974-4c11-a1a1-a81707dd946d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.05045"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.loc[\"ENSG00000237613\", \"GTEX-1117F\"] # value of a simgle position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "e9d5a77c-13a0-4489-b410-cd8379e92668",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.05045"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.iloc[3, 1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2855c73-336e-4995-80f0-afb0814558ad",
   "metadata": {},
   "source": [
    "### Multiple conditions\n",
    "\n",
    "Multiple conditions can be combined using the same rules that apply to Series."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "413a6476-984d-447e-a8a2-9d40c749303f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.4000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.1602</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                                      \n",
       "ENSG00000227232      WASH7P     21.4000    16.75000    11.03000\n",
       "ENSG00000243485  MIR1302-11      0.1602     0.04674     0.06433"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cond = (df_gene[\"GTEX-1117F\"] > 5.) | (df_gene[\"GTEX-111FC\"] > 0.03)\n",
    "df_gene.loc[cond]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b208218-4f60-4e76-b0f1-2042d3808616",
   "metadata": {},
   "source": [
    "### Sort rows\n",
    "\n",
    "You can apply the `.sort_values()` method on a dataframe to sort by a column, optionally selecting the sort order with the boolean `ascending` parameter."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "c649535d-61a9-4bb7-a32b-14b179fad5b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.11580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F  GTEX-111FC  GTEX-111CU\n",
       "EnsemblID                                                      \n",
       "ENSG00000227232      WASH7P    21.40000    16.75000    11.03000\n",
       "ENSG00000243485  MIR1302-11     0.16020     0.04674     0.06433\n",
       "ENSG00000223972     DDX11L1     0.10820     0.02104     0.11580\n",
       "ENSG00000237613     FAM138A     0.05045     0.02945     0.00000\n",
       "ENSG00000268020      OR4G4P     0.00000     0.00000     0.00000\n",
       "ENSG00000186092       OR4F5     0.00000     0.00000     0.00000"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# sort rows (genes) by decreasing expression in sample GTEX-1117F\n",
    "df_gene.sort_values(by=\"GTEX-1117F\", ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b510cd3-217b-4fbf-8d2b-b613fca09418",
   "metadata": {},
   "source": [
    "A column is a Series, so as we did before we can apply `.sort_values()` to any column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "bec73f80-4c29-484d-ba6e-6dfe5519fa18",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EnsemblID\n",
       "ENSG00000227232    21.40000\n",
       "ENSG00000243485     0.16020\n",
       "ENSG00000223972     0.10820\n",
       "ENSG00000237613     0.05045\n",
       "ENSG00000268020     0.00000\n",
       "ENSG00000186092     0.00000\n",
       "Name: GTEX-1117F, dtype: float64"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene[\"GTEX-1117F\"].sort_values(ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b1b846b-6ab8-403c-9eec-9abe13863e4b",
   "metadata": {},
   "source": [
    "### Add/remove/modify columns\n",
    "\n",
    "You can add and remove columns as part of your initial data cleaning phase. Modifying columns is usually part of the feature engineering, where you create new, more effective features (columns) for downstream analysis (e.g. machine learning)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "6de3b4b3-806b-40f4-b551-1c159a84059a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>GTEX-111CU</th>\n",
       "      <th>dummy_col</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.11580</td>\n",
       "      <td>10.496714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>11.03000</td>\n",
       "      <td>9.861736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.06433</td>\n",
       "      <td>10.647689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>11.523030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>9.765847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>9.765863</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F  GTEX-111FC  GTEX-111CU  dummy_col\n",
       "EnsemblID                                                                 \n",
       "ENSG00000223972     DDX11L1     0.10820     0.02104     0.11580  10.496714\n",
       "ENSG00000227232      WASH7P    21.40000    16.75000    11.03000   9.861736\n",
       "ENSG00000243485  MIR1302-11     0.16020     0.04674     0.06433  10.647689\n",
       "ENSG00000237613     FAM138A     0.05045     0.02945     0.00000  11.523030\n",
       "ENSG00000268020      OR4G4P     0.00000     0.00000     0.00000   9.765847\n",
       "ENSG00000186092       OR4F5     0.00000     0.00000     0.00000   9.765863"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.random.seed(42)\n",
    "df_gene[\"dummy_col\"] = np.random.normal(loc=10, size=df_gene.shape[0])\n",
    "df_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "96273d38-26df-48e8-b0fb-4df81d2c5330",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F  GTEX-111FC\n",
       "EnsemblID                                          \n",
       "ENSG00000223972     DDX11L1     0.10820     0.02104\n",
       "ENSG00000227232      WASH7P    21.40000    16.75000\n",
       "ENSG00000243485  MIR1302-11     0.16020     0.04674\n",
       "ENSG00000237613     FAM138A     0.05045     0.02945\n",
       "ENSG00000268020      OR4G4P     0.00000     0.00000\n",
       "ENSG00000186092       OR4F5     0.00000     0.00000"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop columns: method 1\n",
    "df_gene.drop([\"dummy_col\", \"GTEX-111CU\"], axis=1) # axis=1 means operating on the columns, axis=0 means operating on the rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "b3fdfe2a-b658-48ca-ab36-5483f6d70509",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F  GTEX-111FC\n",
       "EnsemblID                                          \n",
       "ENSG00000223972     DDX11L1     0.10820     0.02104\n",
       "ENSG00000227232      WASH7P    21.40000    16.75000\n",
       "ENSG00000243485  MIR1302-11     0.16020     0.04674\n",
       "ENSG00000237613     FAM138A     0.05045     0.02945\n",
       "ENSG00000268020      OR4G4P     0.00000     0.00000\n",
       "ENSG00000186092       OR4F5     0.00000     0.00000"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop columns: method 2\n",
    "df_gene = df_gene.drop(columns=[\"dummy_col\", \"GTEX-111CU\"])\n",
    "df_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e91440d8-b26e-4d1f-83f1-14c236a958d7",
   "metadata": {},
   "source": [
    "We can assign and modify columns with the `.assign()` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "1fe34d2d-0345-4419-82c7-7990e0ecf794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GTEX-1117F</th>\n",
       "      <th>GTEX-111FC</th>\n",
       "      <th>delta</th>\n",
       "      <th>logdelta</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "      <td>0.08716</td>\n",
       "      <td>0.120564</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "      <td>4.65000</td>\n",
       "      <td>2.498251</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "      <td>0.11346</td>\n",
       "      <td>0.155050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "      <td>0.02100</td>\n",
       "      <td>0.029983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene  GTEX-1117F  GTEX-111FC    delta  logdelta\n",
       "EnsemblID                                                             \n",
       "ENSG00000223972     DDX11L1     0.10820     0.02104  0.08716  0.120564\n",
       "ENSG00000227232      WASH7P    21.40000    16.75000  4.65000  2.498251\n",
       "ENSG00000243485  MIR1302-11     0.16020     0.04674  0.11346  0.155050\n",
       "ENSG00000237613     FAM138A     0.05045     0.02945  0.02100  0.029983\n",
       "ENSG00000268020      OR4G4P     0.00000     0.00000  0.00000  0.000000\n",
       "ENSG00000186092       OR4F5     0.00000     0.00000  0.00000  0.000000"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.assign(\n",
    "    delta=df_gene[\"GTEX-1117F\"] - df_gene[\"GTEX-111FC\"],\n",
    "    logdelta=np.log2(1 + df_gene[\"GTEX-1117F\"] - df_gene[\"GTEX-111FC\"]),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99ae9dc8-02d9-4aa6-addf-3d0a99a738fa",
   "metadata": {},
   "source": [
    "Columns are easily renamed with the `.rename()` method, accepting a dictionary in the form `{'old_name': 'new_name'}`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "ae1f307b-0f35-4991-bc38-e403f0283040",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Sample1</th>\n",
       "      <th>Sample2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnsemblID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>0.10820</td>\n",
       "      <td>0.02104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>21.40000</td>\n",
       "      <td>16.75000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485</th>\n",
       "      <td>MIR1302-11</td>\n",
       "      <td>0.16020</td>\n",
       "      <td>0.04674</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613</th>\n",
       "      <td>FAM138A</td>\n",
       "      <td>0.05045</td>\n",
       "      <td>0.02945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020</th>\n",
       "      <td>OR4G4P</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186092</th>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       Gene   Sample1   Sample2\n",
       "EnsemblID                                      \n",
       "ENSG00000223972     DDX11L1   0.10820   0.02104\n",
       "ENSG00000227232      WASH7P  21.40000  16.75000\n",
       "ENSG00000243485  MIR1302-11   0.16020   0.04674\n",
       "ENSG00000237613     FAM138A   0.05045   0.02945\n",
       "ENSG00000268020      OR4G4P   0.00000   0.00000\n",
       "ENSG00000186092       OR4F5   0.00000   0.00000"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_renamed = df_gene.rename(columns={'GTEX-1117F': 'Sample1', 'GTEX-111FC': 'Sample2'})\n",
    "df_renamed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2342977-0209-4748-80ee-889795cd2937",
   "metadata": {},
   "source": [
    "### Reading from files\n",
    "\n",
    "Pandas has convenient `read_<format>` methods to read data in different formats (CSV, TSV, Excel, JSON, pickle, etc.).\n",
    "\n",
    "Let's load a small toy dataset about how mouse weight responded to a particular treatment. This data contains 4 columns: \n",
    "\n",
    "- `Mouse`, the mouse label/number\n",
    "- `Treated`, whether or not it was treated\n",
    "- `Sex`\n",
    "- `Weight`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "bce771e0-8f70-44b2-aa8e-0ffd4c83f47f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse</th>\n",
       "      <th>Treated</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mouse  Treated Sex  Weight\n",
       "0      1     True   M      22\n",
       "1      2     True   F      25\n",
       "2      3     True   M      27\n",
       "3      4     True   F      23\n",
       "4      5     True   M      23"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv(DATADIR / \"mouse_weight_data.csv\")\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "f6f38f95-03c9-41a3-93d6-fc77152b867b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Treated</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Weight</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mouse</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Treated Sex  Weight\n",
       "Mouse                     \n",
       "1         True   M      22\n",
       "2         True   F      25\n",
       "3         True   M      27\n",
       "4         True   F      23\n",
       "5         True   M      23"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.set_index(\"Mouse\") #df = pd.read_csv(DATADIR / \"mouse_weight_data.csv\",index_col=0)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "11f9b207-d6bb-45b5-9e80-f210cf6cfae5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mouse</th>\n",
       "      <th>Treated</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>F</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>M</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Mouse  Treated Sex  Weight\n",
       "0      1     True   M      22\n",
       "1      2     True   F      25\n",
       "2      3     True   M      27\n",
       "3      4     True   F      23\n",
       "4      5     True   M      23"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.reset_index().head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb088af8-e8d3-47c2-b6cf-8808237eb09d",
   "metadata": {},
   "source": [
    "You can set the correct index straight from `pd.read_csv()`, by setting the argument `index_col` to the column name that you want to use as index!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9be116d9-eb72-47e1-953d-ba05ab9581e8",
   "metadata": {},
   "source": [
    "### Exercise: Statistical testing revisited\n",
    "\n",
    "1. (Re)load the `mouse_weight_data.csv` from the `data` folder using the column \"Mouse\" as index\n",
    "2. Extract the weights of the treated and the untreated mice (hint: the column Treated is boolean: use it to select groups; in Pandas conditions, use `~` to negate)\n",
    "3. Calculate the mean weight per group and print them\n",
    "4. Test for statistical significance: there are two group involved, so you'll need a two-sample t-test (`ttest_ind()` from scipy `stats`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "5b478daf-4c04-414e-942e-e6ea62e807f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25.066666666666666\n",
      "26.8\n",
      "Treated mean weight: 25g\n",
      "Untreated mean weight: 27g\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "0.03166586638057747"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_df = pd.read_csv(DATADIR / \"mouse_weight_data.csv\",index_col=0)\n",
    "condition = mouse_df[\"Treated\"]\n",
    "\n",
    "#treated_weights = mouse_df[\"Weight\"].loc[condition]\n",
    "treated_weights = mouse_df[mouse_df[\"Treated\"]][\"Weight\"]\n",
    "untreated_weights = mouse_df[~mouse_df[\"Treated\"]][\"Weight\"]\n",
    "\n",
    "treated_mean = treated_weights.mean()\n",
    "untreated_mean = untreated_weights.mean()\n",
    "\n",
    "# print\n",
    "print(treated_mean)\n",
    "print(untreated_mean)\n",
    "\n",
    "print(f\"Treated mean weight: {treated_mean:0.0f}g\\nUntreated mean weight: {untreated_mean:0.0f}g\")\n",
    "\n",
    "# t test\n",
    "t_statistic, p_value = stats.ttest_ind(treated_weights, untreated_weights) # can use a \"_\" to catch the t_statistic value and disgard. Like: \"_,p_value=...\"\n",
    "\n",
    "p_value"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a0d25d5e-1c53-45f6-ab0d-f45c9d02a1b1",
   "metadata": {},
   "source": [
    "mouse_df = pd.read_csv(DATADIR / \"mouse_weight_data.csv\",index_col=0)\n",
    "\n",
    "treated_weights = mouse_df[ Treated== ][ <column> ]\n",
    "untreated_weights = ___\n",
    "\n",
    "treated_mean = ___\n",
    "untreated_mean = ___\n",
    "\n",
    "... # print the means\n",
    "\n",
    "... # t-test"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a639cbb-c5d2-4292-9f40-4920f1ae0980",
   "metadata": {},
   "source": [
    "___\n",
    "\n",
    "## Grouped operations\n",
    "\n",
    "Pandas has the `.groupby()` method that allows you to compute grouped (or aggregated) calculations. For example, in our Mouse dataset:\n",
    "\n",
    "- what is the average weight by sex?\n",
    "- what is the average weight by sex, stratified by treatment?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b309dafc-9f8e-4b56-98e6-9d689f32ed4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "mouse_df.groupby(\"Sex\") # group by a col"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78b31287-5ba9-4a4c-a28e-6343624750f7",
   "metadata": {},
   "source": [
    "`.groupby()` *prepares* the data for a downstream aggregation method, such as such as `.sum()`, `.mean()`, etc."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "2166c7bc-29fd-4a20-a2f3-3cfd2ff401fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sex\n",
       "F    26.000000\n",
       "M    25.866667\n",
       "Name: Weight, dtype: float64"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_df.groupby(\"Sex\")[\"Weight\"].mean()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a695fd7-46f8-40d2-82b2-02024febeda3",
   "metadata": {},
   "source": [
    "If you prefer, you can assign the grouped dataframe to its own variable."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ead9ec4-ac31-4fff-bd44-a880b1b77eb0",
   "metadata": {},
   "source": [
    "It is easy to group by more than one variable: for example, let's compute the average `Weight` broken down by `Sex` and `Treated`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "f937282c-e4df-476b-b1b8-890bc5bec612",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Sex  Treated\n",
       "F    False      27.000000\n",
       "     True       24.857143\n",
       "M    False      26.571429\n",
       "     True       25.250000\n",
       "Name: Weight, dtype: float64"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multi_var_df = mouse_df.groupby([\"Sex\", \"Treated\"])[\"Weight\"].mean()\n",
    "multi_var_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "178a0667-9c9b-4973-9093-1c0d39643b1d",
   "metadata": {},
   "source": [
    "Notice the hierarchical structure of the row indexes. If you prefer, you can \"flatten\" it out:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "276dfa9b-fd4f-4a7e-90b0-4cc02c0c103a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sex</th>\n",
       "      <th>Treated</th>\n",
       "      <th>Weight</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>F</td>\n",
       "      <td>False</td>\n",
       "      <td>27.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>F</td>\n",
       "      <td>True</td>\n",
       "      <td>24.857143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M</td>\n",
       "      <td>False</td>\n",
       "      <td>26.571429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M</td>\n",
       "      <td>True</td>\n",
       "      <td>25.250000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Sex  Treated     Weight\n",
       "0   F    False  27.000000\n",
       "1   F     True  24.857143\n",
       "2   M    False  26.571429\n",
       "3   M     True  25.250000"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "flat_df = multi_var_df.reset_index() # reset_index() flatens the output into a more usable dataframe\n",
    "flat_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e9a2e18-f0a2-4172-9bcb-9b45dd614a52",
   "metadata": {},
   "source": [
    "Another `groupby()` method worth mentioning is `.agg()`, which lets you run a bunch of different functions on your DataFrame simultaneously. For example, we can generate a simple statistical summary of the dataset as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "4411d64f-3b02-4530-965d-9dd961e65734",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>mean</th>\n",
       "      <th>median</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Treated</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>23</td>\n",
       "      <td>30</td>\n",
       "      <td>26.800000</td>\n",
       "      <td>27.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>22</td>\n",
       "      <td>27</td>\n",
       "      <td>25.066667</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         min  max       mean  median\n",
       "Treated                             \n",
       "False     23   30  26.800000    27.0\n",
       "True      22   27  25.066667    25.0"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_df.groupby(\"Treated\").Weight.agg([\"min\", \"max\", \"mean\", \"median\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09e0a74a-c30d-456b-af32-c68c2408921b",
   "metadata": {},
   "source": [
    "## Exercise\n",
    "\n",
    "The file \"neuroblastoma.tsv\", in the `data` folder, contains a few clinical parameters and the expression of 3 genes (MYCN, ALK, and TP53) for 20 neuroblastoma patients, randomly selected from a larger set.\n",
    "\n",
    "Read the file with `pd.read_csv()`, setting the appropriate separator through the argument `sep` (e.g., for semicolon use `';'`, for tab use `'\\t'`, etc.), and save it to `nb_df`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "17d9cd25-8613-4b5a-b40f-30e37a964eaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "neuro_data = pd.read_csv(DATADIR / \"neuroblastoma.tsv\",sep='\\t',index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1be4c71c-55dc-4e94-ad10-6396aed32e99",
   "metadata": {},
   "source": [
    "1. Create a DataFrame `avg_surv` containing the average survival time (`os_years`) broken down by the neuroblastoma INSS staging `inss_stage`. What stage is associated with the worst prognosis, on average?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "649b2e1a-0086-40df-92b7-82013a96a648",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "inss_stage\n",
       "4     2.737851\n",
       "2     4.093771\n",
       "4S    6.498289\n",
       "1     7.927447\n",
       "3     8.376728\n",
       "Name: os_years, dtype: float64"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avg_surv = neuro_data.groupby(\"inss_stage\")[\"os_years\"].mean().sort_values()\n",
    "avg_surv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "763409ce-3160-45e6-8a2c-f8215bc54ab4",
   "metadata": {},
   "source": [
    "2. What are the minimum and maximum survival times for each `age_group`? (age at diagnosis) Create a DataFrame whose index is the age group category from the dataset and whose values are the min and max values thereof."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "78a57b8e-8dc7-4672-b308-6bf6fcbacbda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>age_group</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>&lt; 18 months</th>\n",
       "      <td>1.127995</td>\n",
       "      <td>14.409309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>&gt; 18 months</th>\n",
       "      <td>0.457221</td>\n",
       "      <td>10.529774</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  min        max\n",
       "age_group                       \n",
       "< 18 months  1.127995  14.409309\n",
       "> 18 months  0.457221  10.529774"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neuro_data.groupby(\"age_group\")[\"os_years\"].agg([\"min\", \"max\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81f6ef59-984a-4938-a855-aa35f6f66651",
   "metadata": {},
   "source": [
    "3. Create a `Series` whose index is the `high_risk` status and whose values are the average expression value of TP53 for each value of `high_risk`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "71990556-e23a-43eb-8088-fa01e36030e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>high_risk</th>\n",
       "      <th>TP53</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>N</td>\n",
       "      <td>13.735385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Y</td>\n",
       "      <td>14.717143</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  high_risk       TP53\n",
       "0         N  13.735385\n",
       "1         Y  14.717143"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "neuro_data.groupby(\"high_risk\")[\"TP53\"].mean().reset_index()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "1675505e-a815-42ac-8755-94e2f88c908b",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Combining data\n",
    "\n",
    "Combining data is another part of your typical data analysis workflow. For example, you have sample IDs and clinical data in one file, and gene expression value for those samples in another file, and you want to combine those two files in the most robust way.\n",
    "\n",
    "Pandas has a `.merge()` method to perform this kind of operation.\n",
    "\n",
    "The syntax is `left.merge(right)`, meaning that a `left` dataframe will be merged with the `right` dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "9b07e42c-7c3b-449d-8a6c-0d8839790103",
   "metadata": {},
   "outputs": [],
   "source": [
    "sites = pd.read_csv(DATADIR / \"survey_site.csv\")\n",
    "visited = pd.read_csv(DATADIR / \"survey_visited.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "98e1faac-d6e5-455a-afb5-2b93cd084381",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>lat</th>\n",
       "      <th>long</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DR-1</td>\n",
       "      <td>-49.85</td>\n",
       "      <td>-128.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DR-3</td>\n",
       "      <td>-47.15</td>\n",
       "      <td>-126.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSK-4</td>\n",
       "      <td>-48.87</td>\n",
       "      <td>-123.40</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    site    lat    long\n",
       "0   DR-1 -49.85 -128.57\n",
       "1   DR-3 -47.15 -126.72\n",
       "2  MSK-4 -48.87 -123.40"
      ]
     },
     "execution_count": 139,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "2a2aaeb1-d0f2-4cd5-a6a7-4d3aec0c088a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ident</th>\n",
       "      <th>site</th>\n",
       "      <th>dated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>837</td>\n",
       "      <td>MSK-4</td>\n",
       "      <td>1932-01-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>734</td>\n",
       "      <td>DR-3</td>\n",
       "      <td>1939-01-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>619</td>\n",
       "      <td>DR-1</td>\n",
       "      <td>1927-02-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ident   site       dated\n",
       "0    837  MSK-4  1932-01-14\n",
       "1    734   DR-3  1939-01-07\n",
       "2    619   DR-1  1927-02-08"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "visited"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54d3db85-a8fc-424c-bd88-549dde95e6a3",
   "metadata": {},
   "source": [
    "Let's merge `sites` (\"left\") and `visited` (\"right\"), using the common column `site`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f783cd91-4571-4016-9c11-224241c3a49a",
   "metadata": {},
   "source": [
    "### One-to-one merge\n",
    "\n",
    "This kind of merge works when there are no duplicate values in the joining columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "1b1b30f8-843f-4f2c-bee0-77692d7ac0d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>site</th>\n",
       "      <th>lat</th>\n",
       "      <th>long</th>\n",
       "      <th>ident</th>\n",
       "      <th>dated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DR-1</td>\n",
       "      <td>-49.85</td>\n",
       "      <td>-128.57</td>\n",
       "      <td>619</td>\n",
       "      <td>1927-02-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DR-3</td>\n",
       "      <td>-47.15</td>\n",
       "      <td>-126.72</td>\n",
       "      <td>734</td>\n",
       "      <td>1939-01-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSK-4</td>\n",
       "      <td>-48.87</td>\n",
       "      <td>-123.40</td>\n",
       "      <td>837</td>\n",
       "      <td>1932-01-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    site    lat    long  ident       dated\n",
       "0   DR-1 -49.85 -128.57    619  1927-02-08\n",
       "1   DR-3 -47.15 -126.72    734  1939-01-07\n",
       "2  MSK-4 -48.87 -123.40    837  1932-01-14"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sites.merge(visited, left_on=\"site\", right_on=\"site\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97a75c16-a7a1-4e47-b7fa-36191c6e6933",
   "metadata": {},
   "source": [
    "See how the resulting dataframe has the first columns from the \"left\" dataframe.\n",
    "\n",
    "The optional argument `how` determines the type of join:\n",
    "\n",
    "- `how=\"inner\"` (default), use intersection of keys from both dataframes\n",
    "- `how=\"outer\"`, use union of keys from both dataframes\n",
    "- `how=\"left\"` / `how=\"right\"`, use only keys from left (right) dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cbd28b8-fced-475a-94f8-e10e680a036e",
   "metadata": {},
   "source": [
    "### Many-to-one\n",
    "\n",
    "In case the joining column contains duplicates, you would obtain a \"many-to-one\" merge, where all the left dataframe info are matched to the right dataframe and replicated as needed."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9604d633-7d7d-4fdd-af31-e79ffafd4da9",
   "metadata": {},
   "source": [
    "## Exercise\n",
    "\n",
    "You have to merge clinical information (stored in the file `nb_clinical.txt`) with gene expression (provided in the file `nb_expr.txt`) into a single dataframe.\n",
    "\n",
    "Start by reading the two files into dataframes, figuring out the file format and the separator.\n",
    "\n",
    "Have a look at the data and check the sizes.\n",
    "\n",
    "Merge the two dataframes by sample ID. Clean up the merged dataframe be dropping redundant columns, if any.\n",
    "\n",
    "Finally, impute missing values to 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6de686ed-81df-4b53-bb34-bf2a18f5ce51",
   "metadata": {},
   "outputs": [],
   "source": [
    "...\n",
    "df.drop_duplicates()\n",
    "df.fillna(0)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e4ad95-992a-4ed4-a20f-7e6cb5b53e30",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Breast Cancer Detection Data Set\n",
    "\n",
    "source: https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29\n",
    "\n",
    "The data comprises features from a digitized image of a fine needle aspirate (FNA) of a breast mass. They describe characteristics of the cell nuclei present in the image in the 3-dimensional space. For now, there is no need to fully understand where the variables come from or their units: the focus here is to explore what we can do with Pandas and understand how easy it is to apply these tools to any kind of data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "4650e5b3-fea5-4771-8814-6a004f99508c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "      <th>fractal_dimension_worst</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>842302</td>\n",
       "      <td>M</td>\n",
       "      <td>17.99</td>\n",
       "      <td>10.38</td>\n",
       "      <td>122.80</td>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.11840</td>\n",
       "      <td>0.27760</td>\n",
       "      <td>0.30010</td>\n",
       "      <td>0.14710</td>\n",
       "      <td>...</td>\n",
       "      <td>17.33</td>\n",
       "      <td>184.60</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>0.16220</td>\n",
       "      <td>0.66560</td>\n",
       "      <td>0.7119</td>\n",
       "      <td>0.2654</td>\n",
       "      <td>0.4601</td>\n",
       "      <td>0.11890</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>842517</td>\n",
       "      <td>M</td>\n",
       "      <td>20.57</td>\n",
       "      <td>17.77</td>\n",
       "      <td>132.90</td>\n",
       "      <td>1326.0</td>\n",
       "      <td>0.08474</td>\n",
       "      <td>0.07864</td>\n",
       "      <td>0.08690</td>\n",
       "      <td>0.07017</td>\n",
       "      <td>...</td>\n",
       "      <td>23.41</td>\n",
       "      <td>158.80</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>0.12380</td>\n",
       "      <td>0.18660</td>\n",
       "      <td>0.2416</td>\n",
       "      <td>0.1860</td>\n",
       "      <td>0.2750</td>\n",
       "      <td>0.08902</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>84300903</td>\n",
       "      <td>M</td>\n",
       "      <td>19.69</td>\n",
       "      <td>21.25</td>\n",
       "      <td>130.00</td>\n",
       "      <td>1203.0</td>\n",
       "      <td>0.10960</td>\n",
       "      <td>0.15990</td>\n",
       "      <td>0.19740</td>\n",
       "      <td>0.12790</td>\n",
       "      <td>...</td>\n",
       "      <td>25.53</td>\n",
       "      <td>152.50</td>\n",
       "      <td>1709.0</td>\n",
       "      <td>0.14440</td>\n",
       "      <td>0.42450</td>\n",
       "      <td>0.4504</td>\n",
       "      <td>0.2430</td>\n",
       "      <td>0.3613</td>\n",
       "      <td>0.08758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>84348301</td>\n",
       "      <td>M</td>\n",
       "      <td>11.42</td>\n",
       "      <td>20.38</td>\n",
       "      <td>77.58</td>\n",
       "      <td>386.1</td>\n",
       "      <td>0.14250</td>\n",
       "      <td>0.28390</td>\n",
       "      <td>0.24140</td>\n",
       "      <td>0.10520</td>\n",
       "      <td>...</td>\n",
       "      <td>26.50</td>\n",
       "      <td>98.87</td>\n",
       "      <td>567.7</td>\n",
       "      <td>0.20980</td>\n",
       "      <td>0.86630</td>\n",
       "      <td>0.6869</td>\n",
       "      <td>0.2575</td>\n",
       "      <td>0.6638</td>\n",
       "      <td>0.17300</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>84358402</td>\n",
       "      <td>M</td>\n",
       "      <td>20.29</td>\n",
       "      <td>14.34</td>\n",
       "      <td>135.10</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>0.10030</td>\n",
       "      <td>0.13280</td>\n",
       "      <td>0.19800</td>\n",
       "      <td>0.10430</td>\n",
       "      <td>...</td>\n",
       "      <td>16.67</td>\n",
       "      <td>152.20</td>\n",
       "      <td>1575.0</td>\n",
       "      <td>0.13740</td>\n",
       "      <td>0.20500</td>\n",
       "      <td>0.4000</td>\n",
       "      <td>0.1625</td>\n",
       "      <td>0.2364</td>\n",
       "      <td>0.07678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>926424</td>\n",
       "      <td>M</td>\n",
       "      <td>21.56</td>\n",
       "      <td>22.39</td>\n",
       "      <td>142.00</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>0.11100</td>\n",
       "      <td>0.11590</td>\n",
       "      <td>0.24390</td>\n",
       "      <td>0.13890</td>\n",
       "      <td>...</td>\n",
       "      <td>26.40</td>\n",
       "      <td>166.10</td>\n",
       "      <td>2027.0</td>\n",
       "      <td>0.14100</td>\n",
       "      <td>0.21130</td>\n",
       "      <td>0.4107</td>\n",
       "      <td>0.2216</td>\n",
       "      <td>0.2060</td>\n",
       "      <td>0.07115</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>926682</td>\n",
       "      <td>M</td>\n",
       "      <td>20.13</td>\n",
       "      <td>28.25</td>\n",
       "      <td>131.20</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>0.09780</td>\n",
       "      <td>0.10340</td>\n",
       "      <td>0.14400</td>\n",
       "      <td>0.09791</td>\n",
       "      <td>...</td>\n",
       "      <td>38.25</td>\n",
       "      <td>155.00</td>\n",
       "      <td>1731.0</td>\n",
       "      <td>0.11660</td>\n",
       "      <td>0.19220</td>\n",
       "      <td>0.3215</td>\n",
       "      <td>0.1628</td>\n",
       "      <td>0.2572</td>\n",
       "      <td>0.06637</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>926954</td>\n",
       "      <td>M</td>\n",
       "      <td>16.60</td>\n",
       "      <td>28.08</td>\n",
       "      <td>108.30</td>\n",
       "      <td>858.1</td>\n",
       "      <td>0.08455</td>\n",
       "      <td>0.10230</td>\n",
       "      <td>0.09251</td>\n",
       "      <td>0.05302</td>\n",
       "      <td>...</td>\n",
       "      <td>34.12</td>\n",
       "      <td>126.70</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.11390</td>\n",
       "      <td>0.30940</td>\n",
       "      <td>0.3403</td>\n",
       "      <td>0.1418</td>\n",
       "      <td>0.2218</td>\n",
       "      <td>0.07820</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>927241</td>\n",
       "      <td>M</td>\n",
       "      <td>20.60</td>\n",
       "      <td>29.33</td>\n",
       "      <td>140.10</td>\n",
       "      <td>1265.0</td>\n",
       "      <td>0.11780</td>\n",
       "      <td>0.27700</td>\n",
       "      <td>0.35140</td>\n",
       "      <td>0.15200</td>\n",
       "      <td>...</td>\n",
       "      <td>39.42</td>\n",
       "      <td>184.60</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>0.86810</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.2650</td>\n",
       "      <td>0.4087</td>\n",
       "      <td>0.12400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>92751</td>\n",
       "      <td>B</td>\n",
       "      <td>7.76</td>\n",
       "      <td>24.54</td>\n",
       "      <td>47.92</td>\n",
       "      <td>181.0</td>\n",
       "      <td>0.05263</td>\n",
       "      <td>0.04362</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>30.37</td>\n",
       "      <td>59.16</td>\n",
       "      <td>268.6</td>\n",
       "      <td>0.08996</td>\n",
       "      <td>0.06444</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2871</td>\n",
       "      <td>0.07039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "0      842302         M        17.99         10.38          122.80     1001.0   \n",
       "1      842517         M        20.57         17.77          132.90     1326.0   \n",
       "2    84300903         M        19.69         21.25          130.00     1203.0   \n",
       "3    84348301         M        11.42         20.38           77.58      386.1   \n",
       "4    84358402         M        20.29         14.34          135.10     1297.0   \n",
       "..        ...       ...          ...           ...             ...        ...   \n",
       "564    926424         M        21.56         22.39          142.00     1479.0   \n",
       "565    926682         M        20.13         28.25          131.20     1261.0   \n",
       "566    926954         M        16.60         28.08          108.30      858.1   \n",
       "567    927241         M        20.60         29.33          140.10     1265.0   \n",
       "568     92751         B         7.76         24.54           47.92      181.0   \n",
       "\n",
       "     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \\\n",
       "0            0.11840           0.27760         0.30010              0.14710   \n",
       "1            0.08474           0.07864         0.08690              0.07017   \n",
       "2            0.10960           0.15990         0.19740              0.12790   \n",
       "3            0.14250           0.28390         0.24140              0.10520   \n",
       "4            0.10030           0.13280         0.19800              0.10430   \n",
       "..               ...               ...             ...                  ...   \n",
       "564          0.11100           0.11590         0.24390              0.13890   \n",
       "565          0.09780           0.10340         0.14400              0.09791   \n",
       "566          0.08455           0.10230         0.09251              0.05302   \n",
       "567          0.11780           0.27700         0.35140              0.15200   \n",
       "568          0.05263           0.04362         0.00000              0.00000   \n",
       "\n",
       "     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \\\n",
       "0    ...          17.33           184.60      2019.0           0.16220   \n",
       "1    ...          23.41           158.80      1956.0           0.12380   \n",
       "2    ...          25.53           152.50      1709.0           0.14440   \n",
       "3    ...          26.50            98.87       567.7           0.20980   \n",
       "4    ...          16.67           152.20      1575.0           0.13740   \n",
       "..   ...            ...              ...         ...               ...   \n",
       "564  ...          26.40           166.10      2027.0           0.14100   \n",
       "565  ...          38.25           155.00      1731.0           0.11660   \n",
       "566  ...          34.12           126.70      1124.0           0.11390   \n",
       "567  ...          39.42           184.60      1821.0           0.16500   \n",
       "568  ...          30.37            59.16       268.6           0.08996   \n",
       "\n",
       "     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \\\n",
       "0              0.66560           0.7119                0.2654          0.4601   \n",
       "1              0.18660           0.2416                0.1860          0.2750   \n",
       "2              0.42450           0.4504                0.2430          0.3613   \n",
       "3              0.86630           0.6869                0.2575          0.6638   \n",
       "4              0.20500           0.4000                0.1625          0.2364   \n",
       "..                 ...              ...                   ...             ...   \n",
       "564            0.21130           0.4107                0.2216          0.2060   \n",
       "565            0.19220           0.3215                0.1628          0.2572   \n",
       "566            0.30940           0.3403                0.1418          0.2218   \n",
       "567            0.86810           0.9387                0.2650          0.4087   \n",
       "568            0.06444           0.0000                0.0000          0.2871   \n",
       "\n",
       "     fractal_dimension_worst  Unnamed: 32  \n",
       "0                    0.11890          NaN  \n",
       "1                    0.08902          NaN  \n",
       "2                    0.08758          NaN  \n",
       "3                    0.17300          NaN  \n",
       "4                    0.07678          NaN  \n",
       "..                       ...          ...  \n",
       "564                  0.07115          NaN  \n",
       "565                  0.06637          NaN  \n",
       "566                  0.07820          NaN  \n",
       "567                  0.12400          NaN  \n",
       "568                  0.07039          NaN  \n",
       "\n",
       "[569 rows x 33 columns]"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# (re)import libraries in case you start from scratch\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "\n",
    "# compose paths and filenames using pathlib, which is operating system-agnostic\n",
    "# then load the data using Pandas\n",
    "DATADIR = Path(\"data\")\n",
    "DATAFILE = DATADIR / \"breast_cancer_diagnostic_data.csv\"\n",
    "bc_data = pd.read_csv(DATAFILE)\n",
    "bc_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6dc94f4-763c-417b-9c48-d2246b454c41",
   "metadata": {},
   "source": [
    "Other parameters can be used inside `read_csv()` to deal with different types of data: for example, different separators, unwanted rows or columns, variable names, and other."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44e43c64-2c58-4234-9cdd-d1562ecf7202",
   "metadata": {},
   "source": [
    "Our data set contains 569 samples (patients), categorical data (diagnosis result) and numerical data (31 attributes concerning the tumor shape and size and the id of each patient). We are now ready to start exploring and manipulating the data as intended. The script below shows a few examples of simple operations that can be applied with just a single line of code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "227bfa6a-afd7-4702-8aa4-4c258f727c90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(569, 33)"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the dimensions\n",
    "bc_data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "cafb91c2-038a-463f-84b2-882bbbfb8248",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id', 'diagnosis', 'radius_mean', 'texture_mean', 'perimeter_mean',\n",
       "       'area_mean', 'smoothness_mean', 'compactness_mean', 'concavity_mean',\n",
       "       'concave points_mean', 'symmetry_mean', 'fractal_dimension_mean',\n",
       "       'radius_se', 'texture_se', 'perimeter_se', 'area_se', 'smoothness_se',\n",
       "       'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se',\n",
       "       'fractal_dimension_se', 'radius_worst', 'texture_worst',\n",
       "       'perimeter_worst', 'area_worst', 'smoothness_worst',\n",
       "       'compactness_worst', 'concavity_worst', 'concave points_worst',\n",
       "       'symmetry_worst', 'fractal_dimension_worst', 'Unnamed: 32'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the header\n",
    "bc_data.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "id": "fbaf65fb-712a-40ca-a2b9-8f157aafbc68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RangeIndex(start=0, stop=569, step=1)"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the index\n",
    "bc_data.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "faa324c5-ba92-4fe1-8553-51fd98f8768f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "      <th>fractal_dimension_worst</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>926424</td>\n",
       "      <td>M</td>\n",
       "      <td>21.56</td>\n",
       "      <td>22.39</td>\n",
       "      <td>142.00</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>0.11100</td>\n",
       "      <td>0.11590</td>\n",
       "      <td>0.24390</td>\n",
       "      <td>0.13890</td>\n",
       "      <td>...</td>\n",
       "      <td>26.40</td>\n",
       "      <td>166.10</td>\n",
       "      <td>2027.0</td>\n",
       "      <td>0.14100</td>\n",
       "      <td>0.21130</td>\n",
       "      <td>0.4107</td>\n",
       "      <td>0.2216</td>\n",
       "      <td>0.2060</td>\n",
       "      <td>0.07115</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>926682</td>\n",
       "      <td>M</td>\n",
       "      <td>20.13</td>\n",
       "      <td>28.25</td>\n",
       "      <td>131.20</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>0.09780</td>\n",
       "      <td>0.10340</td>\n",
       "      <td>0.14400</td>\n",
       "      <td>0.09791</td>\n",
       "      <td>...</td>\n",
       "      <td>38.25</td>\n",
       "      <td>155.00</td>\n",
       "      <td>1731.0</td>\n",
       "      <td>0.11660</td>\n",
       "      <td>0.19220</td>\n",
       "      <td>0.3215</td>\n",
       "      <td>0.1628</td>\n",
       "      <td>0.2572</td>\n",
       "      <td>0.06637</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>926954</td>\n",
       "      <td>M</td>\n",
       "      <td>16.60</td>\n",
       "      <td>28.08</td>\n",
       "      <td>108.30</td>\n",
       "      <td>858.1</td>\n",
       "      <td>0.08455</td>\n",
       "      <td>0.10230</td>\n",
       "      <td>0.09251</td>\n",
       "      <td>0.05302</td>\n",
       "      <td>...</td>\n",
       "      <td>34.12</td>\n",
       "      <td>126.70</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.11390</td>\n",
       "      <td>0.30940</td>\n",
       "      <td>0.3403</td>\n",
       "      <td>0.1418</td>\n",
       "      <td>0.2218</td>\n",
       "      <td>0.07820</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>927241</td>\n",
       "      <td>M</td>\n",
       "      <td>20.60</td>\n",
       "      <td>29.33</td>\n",
       "      <td>140.10</td>\n",
       "      <td>1265.0</td>\n",
       "      <td>0.11780</td>\n",
       "      <td>0.27700</td>\n",
       "      <td>0.35140</td>\n",
       "      <td>0.15200</td>\n",
       "      <td>...</td>\n",
       "      <td>39.42</td>\n",
       "      <td>184.60</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>0.86810</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.2650</td>\n",
       "      <td>0.4087</td>\n",
       "      <td>0.12400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>92751</td>\n",
       "      <td>B</td>\n",
       "      <td>7.76</td>\n",
       "      <td>24.54</td>\n",
       "      <td>47.92</td>\n",
       "      <td>181.0</td>\n",
       "      <td>0.05263</td>\n",
       "      <td>0.04362</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>30.37</td>\n",
       "      <td>59.16</td>\n",
       "      <td>268.6</td>\n",
       "      <td>0.08996</td>\n",
       "      <td>0.06444</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2871</td>\n",
       "      <td>0.07039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "564  926424         M        21.56         22.39          142.00     1479.0   \n",
       "565  926682         M        20.13         28.25          131.20     1261.0   \n",
       "566  926954         M        16.60         28.08          108.30      858.1   \n",
       "567  927241         M        20.60         29.33          140.10     1265.0   \n",
       "568   92751         B         7.76         24.54           47.92      181.0   \n",
       "\n",
       "     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \\\n",
       "564          0.11100           0.11590         0.24390              0.13890   \n",
       "565          0.09780           0.10340         0.14400              0.09791   \n",
       "566          0.08455           0.10230         0.09251              0.05302   \n",
       "567          0.11780           0.27700         0.35140              0.15200   \n",
       "568          0.05263           0.04362         0.00000              0.00000   \n",
       "\n",
       "     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \\\n",
       "564  ...          26.40           166.10      2027.0           0.14100   \n",
       "565  ...          38.25           155.00      1731.0           0.11660   \n",
       "566  ...          34.12           126.70      1124.0           0.11390   \n",
       "567  ...          39.42           184.60      1821.0           0.16500   \n",
       "568  ...          30.37            59.16       268.6           0.08996   \n",
       "\n",
       "     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \\\n",
       "564            0.21130           0.4107                0.2216          0.2060   \n",
       "565            0.19220           0.3215                0.1628          0.2572   \n",
       "566            0.30940           0.3403                0.1418          0.2218   \n",
       "567            0.86810           0.9387                0.2650          0.4087   \n",
       "568            0.06444           0.0000                0.0000          0.2871   \n",
       "\n",
       "     fractal_dimension_worst  Unnamed: 32  \n",
       "564                  0.07115          NaN  \n",
       "565                  0.06637          NaN  \n",
       "566                  0.07820          NaN  \n",
       "567                  0.12400          NaN  \n",
       "568                  0.07039          NaN  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the last n rows\n",
    "bc_data.tail(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "835c3075-0914-4838-9e55-0c33eeff5b28",
   "metadata": {},
   "source": [
    "## Review Exercise\n",
    "\n",
    "Index this dataframe by sample ID, instead of the default indexing. You can do that in two ways: by using `set_index()` on the dataframe, or by re-reading it from file specifying which column should be used as the index. Then, use indexing/slicing methods to perform the operations in the below cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "251312af-c717-4a1b-ba40-7ebf6f30404e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "d6099fdd-93df-4c99-bbc1-fbc7cba5537a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>symmetry_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "      <th>fractal_dimension_worst</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>842302</th>\n",
       "      <td>M</td>\n",
       "      <td>17.99</td>\n",
       "      <td>10.38</td>\n",
       "      <td>122.80</td>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.11840</td>\n",
       "      <td>0.27760</td>\n",
       "      <td>0.30010</td>\n",
       "      <td>0.14710</td>\n",
       "      <td>0.2419</td>\n",
       "      <td>...</td>\n",
       "      <td>17.33</td>\n",
       "      <td>184.60</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>0.16220</td>\n",
       "      <td>0.66560</td>\n",
       "      <td>0.7119</td>\n",
       "      <td>0.2654</td>\n",
       "      <td>0.4601</td>\n",
       "      <td>0.11890</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>842517</th>\n",
       "      <td>M</td>\n",
       "      <td>20.57</td>\n",
       "      <td>17.77</td>\n",
       "      <td>132.90</td>\n",
       "      <td>1326.0</td>\n",
       "      <td>0.08474</td>\n",
       "      <td>0.07864</td>\n",
       "      <td>0.08690</td>\n",
       "      <td>0.07017</td>\n",
       "      <td>0.1812</td>\n",
       "      <td>...</td>\n",
       "      <td>23.41</td>\n",
       "      <td>158.80</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>0.12380</td>\n",
       "      <td>0.18660</td>\n",
       "      <td>0.2416</td>\n",
       "      <td>0.1860</td>\n",
       "      <td>0.2750</td>\n",
       "      <td>0.08902</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84300903</th>\n",
       "      <td>M</td>\n",
       "      <td>19.69</td>\n",
       "      <td>21.25</td>\n",
       "      <td>130.00</td>\n",
       "      <td>1203.0</td>\n",
       "      <td>0.10960</td>\n",
       "      <td>0.15990</td>\n",
       "      <td>0.19740</td>\n",
       "      <td>0.12790</td>\n",
       "      <td>0.2069</td>\n",
       "      <td>...</td>\n",
       "      <td>25.53</td>\n",
       "      <td>152.50</td>\n",
       "      <td>1709.0</td>\n",
       "      <td>0.14440</td>\n",
       "      <td>0.42450</td>\n",
       "      <td>0.4504</td>\n",
       "      <td>0.2430</td>\n",
       "      <td>0.3613</td>\n",
       "      <td>0.08758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84348301</th>\n",
       "      <td>M</td>\n",
       "      <td>11.42</td>\n",
       "      <td>20.38</td>\n",
       "      <td>77.58</td>\n",
       "      <td>386.1</td>\n",
       "      <td>0.14250</td>\n",
       "      <td>0.28390</td>\n",
       "      <td>0.24140</td>\n",
       "      <td>0.10520</td>\n",
       "      <td>0.2597</td>\n",
       "      <td>...</td>\n",
       "      <td>26.50</td>\n",
       "      <td>98.87</td>\n",
       "      <td>567.7</td>\n",
       "      <td>0.20980</td>\n",
       "      <td>0.86630</td>\n",
       "      <td>0.6869</td>\n",
       "      <td>0.2575</td>\n",
       "      <td>0.6638</td>\n",
       "      <td>0.17300</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84358402</th>\n",
       "      <td>M</td>\n",
       "      <td>20.29</td>\n",
       "      <td>14.34</td>\n",
       "      <td>135.10</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>0.10030</td>\n",
       "      <td>0.13280</td>\n",
       "      <td>0.19800</td>\n",
       "      <td>0.10430</td>\n",
       "      <td>0.1809</td>\n",
       "      <td>...</td>\n",
       "      <td>16.67</td>\n",
       "      <td>152.20</td>\n",
       "      <td>1575.0</td>\n",
       "      <td>0.13740</td>\n",
       "      <td>0.20500</td>\n",
       "      <td>0.4000</td>\n",
       "      <td>0.1625</td>\n",
       "      <td>0.2364</td>\n",
       "      <td>0.07678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926424</th>\n",
       "      <td>M</td>\n",
       "      <td>21.56</td>\n",
       "      <td>22.39</td>\n",
       "      <td>142.00</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>0.11100</td>\n",
       "      <td>0.11590</td>\n",
       "      <td>0.24390</td>\n",
       "      <td>0.13890</td>\n",
       "      <td>0.1726</td>\n",
       "      <td>...</td>\n",
       "      <td>26.40</td>\n",
       "      <td>166.10</td>\n",
       "      <td>2027.0</td>\n",
       "      <td>0.14100</td>\n",
       "      <td>0.21130</td>\n",
       "      <td>0.4107</td>\n",
       "      <td>0.2216</td>\n",
       "      <td>0.2060</td>\n",
       "      <td>0.07115</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926682</th>\n",
       "      <td>M</td>\n",
       "      <td>20.13</td>\n",
       "      <td>28.25</td>\n",
       "      <td>131.20</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>0.09780</td>\n",
       "      <td>0.10340</td>\n",
       "      <td>0.14400</td>\n",
       "      <td>0.09791</td>\n",
       "      <td>0.1752</td>\n",
       "      <td>...</td>\n",
       "      <td>38.25</td>\n",
       "      <td>155.00</td>\n",
       "      <td>1731.0</td>\n",
       "      <td>0.11660</td>\n",
       "      <td>0.19220</td>\n",
       "      <td>0.3215</td>\n",
       "      <td>0.1628</td>\n",
       "      <td>0.2572</td>\n",
       "      <td>0.06637</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926954</th>\n",
       "      <td>M</td>\n",
       "      <td>16.60</td>\n",
       "      <td>28.08</td>\n",
       "      <td>108.30</td>\n",
       "      <td>858.1</td>\n",
       "      <td>0.08455</td>\n",
       "      <td>0.10230</td>\n",
       "      <td>0.09251</td>\n",
       "      <td>0.05302</td>\n",
       "      <td>0.1590</td>\n",
       "      <td>...</td>\n",
       "      <td>34.12</td>\n",
       "      <td>126.70</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.11390</td>\n",
       "      <td>0.30940</td>\n",
       "      <td>0.3403</td>\n",
       "      <td>0.1418</td>\n",
       "      <td>0.2218</td>\n",
       "      <td>0.07820</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927241</th>\n",
       "      <td>M</td>\n",
       "      <td>20.60</td>\n",
       "      <td>29.33</td>\n",
       "      <td>140.10</td>\n",
       "      <td>1265.0</td>\n",
       "      <td>0.11780</td>\n",
       "      <td>0.27700</td>\n",
       "      <td>0.35140</td>\n",
       "      <td>0.15200</td>\n",
       "      <td>0.2397</td>\n",
       "      <td>...</td>\n",
       "      <td>39.42</td>\n",
       "      <td>184.60</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>0.86810</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.2650</td>\n",
       "      <td>0.4087</td>\n",
       "      <td>0.12400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92751</th>\n",
       "      <td>B</td>\n",
       "      <td>7.76</td>\n",
       "      <td>24.54</td>\n",
       "      <td>47.92</td>\n",
       "      <td>181.0</td>\n",
       "      <td>0.05263</td>\n",
       "      <td>0.04362</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.1587</td>\n",
       "      <td>...</td>\n",
       "      <td>30.37</td>\n",
       "      <td>59.16</td>\n",
       "      <td>268.6</td>\n",
       "      <td>0.08996</td>\n",
       "      <td>0.06444</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2871</td>\n",
       "      <td>0.07039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "id                                                                         \n",
       "842302           M        17.99         10.38          122.80     1001.0   \n",
       "842517           M        20.57         17.77          132.90     1326.0   \n",
       "84300903         M        19.69         21.25          130.00     1203.0   \n",
       "84348301         M        11.42         20.38           77.58      386.1   \n",
       "84358402         M        20.29         14.34          135.10     1297.0   \n",
       "...            ...          ...           ...             ...        ...   \n",
       "926424           M        21.56         22.39          142.00     1479.0   \n",
       "926682           M        20.13         28.25          131.20     1261.0   \n",
       "926954           M        16.60         28.08          108.30      858.1   \n",
       "927241           M        20.60         29.33          140.10     1265.0   \n",
       "92751            B         7.76         24.54           47.92      181.0   \n",
       "\n",
       "          smoothness_mean  compactness_mean  concavity_mean  \\\n",
       "id                                                            \n",
       "842302            0.11840           0.27760         0.30010   \n",
       "842517            0.08474           0.07864         0.08690   \n",
       "84300903          0.10960           0.15990         0.19740   \n",
       "84348301          0.14250           0.28390         0.24140   \n",
       "84358402          0.10030           0.13280         0.19800   \n",
       "...                   ...               ...             ...   \n",
       "926424            0.11100           0.11590         0.24390   \n",
       "926682            0.09780           0.10340         0.14400   \n",
       "926954            0.08455           0.10230         0.09251   \n",
       "927241            0.11780           0.27700         0.35140   \n",
       "92751             0.05263           0.04362         0.00000   \n",
       "\n",
       "          concave points_mean  symmetry_mean  ...  texture_worst  \\\n",
       "id                                            ...                  \n",
       "842302                0.14710         0.2419  ...          17.33   \n",
       "842517                0.07017         0.1812  ...          23.41   \n",
       "84300903              0.12790         0.2069  ...          25.53   \n",
       "84348301              0.10520         0.2597  ...          26.50   \n",
       "84358402              0.10430         0.1809  ...          16.67   \n",
       "...                       ...            ...  ...            ...   \n",
       "926424                0.13890         0.1726  ...          26.40   \n",
       "926682                0.09791         0.1752  ...          38.25   \n",
       "926954                0.05302         0.1590  ...          34.12   \n",
       "927241                0.15200         0.2397  ...          39.42   \n",
       "92751                 0.00000         0.1587  ...          30.37   \n",
       "\n",
       "          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \\\n",
       "id                                                                           \n",
       "842302             184.60      2019.0           0.16220            0.66560   \n",
       "842517             158.80      1956.0           0.12380            0.18660   \n",
       "84300903           152.50      1709.0           0.14440            0.42450   \n",
       "84348301            98.87       567.7           0.20980            0.86630   \n",
       "84358402           152.20      1575.0           0.13740            0.20500   \n",
       "...                   ...         ...               ...                ...   \n",
       "926424             166.10      2027.0           0.14100            0.21130   \n",
       "926682             155.00      1731.0           0.11660            0.19220   \n",
       "926954             126.70      1124.0           0.11390            0.30940   \n",
       "927241             184.60      1821.0           0.16500            0.86810   \n",
       "92751               59.16       268.6           0.08996            0.06444   \n",
       "\n",
       "          concavity_worst  concave points_worst  symmetry_worst  \\\n",
       "id                                                                \n",
       "842302             0.7119                0.2654          0.4601   \n",
       "842517             0.2416                0.1860          0.2750   \n",
       "84300903           0.4504                0.2430          0.3613   \n",
       "84348301           0.6869                0.2575          0.6638   \n",
       "84358402           0.4000                0.1625          0.2364   \n",
       "...                   ...                   ...             ...   \n",
       "926424             0.4107                0.2216          0.2060   \n",
       "926682             0.3215                0.1628          0.2572   \n",
       "926954             0.3403                0.1418          0.2218   \n",
       "927241             0.9387                0.2650          0.4087   \n",
       "92751              0.0000                0.0000          0.2871   \n",
       "\n",
       "          fractal_dimension_worst  Unnamed: 32  \n",
       "id                                              \n",
       "842302                    0.11890          NaN  \n",
       "842517                    0.08902          NaN  \n",
       "84300903                  0.08758          NaN  \n",
       "84348301                  0.17300          NaN  \n",
       "84358402                  0.07678          NaN  \n",
       "...                           ...          ...  \n",
       "926424                    0.07115          NaN  \n",
       "926682                    0.06637          NaN  \n",
       "926954                    0.07820          NaN  \n",
       "927241                    0.12400          NaN  \n",
       "92751                     0.07039          NaN  \n",
       "\n",
       "[569 rows x 32 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# use your preferred method to index the df by sample ID\n",
    "\n",
    "bc_data.set_index(\"id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "67c1318a-5f7f-4893-8421-7a9c81a2f679",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      1001.0\n",
       "1      1326.0\n",
       "2      1203.0\n",
       "3       386.1\n",
       "4      1297.0\n",
       "        ...  \n",
       "564    1479.0\n",
       "565    1261.0\n",
       "566     858.1\n",
       "567    1265.0\n",
       "568     181.0\n",
       "Name: area_mean, Length: 569, dtype: float64"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the contents of the column \"area_mean\"\n",
    "# (output: list/array containing column values for all rows)\n",
    "\n",
    "bc_data[\"area_mean\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "1363fede-af23-483f-9686-ac82779cd03c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>area_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1001.0</td>\n",
       "      <td>122.80</td>\n",
       "      <td>10.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1326.0</td>\n",
       "      <td>132.90</td>\n",
       "      <td>17.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1203.0</td>\n",
       "      <td>130.00</td>\n",
       "      <td>21.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>386.1</td>\n",
       "      <td>77.58</td>\n",
       "      <td>20.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1297.0</td>\n",
       "      <td>135.10</td>\n",
       "      <td>14.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>1479.0</td>\n",
       "      <td>142.00</td>\n",
       "      <td>22.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>1261.0</td>\n",
       "      <td>131.20</td>\n",
       "      <td>28.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>858.1</td>\n",
       "      <td>108.30</td>\n",
       "      <td>28.08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>1265.0</td>\n",
       "      <td>140.10</td>\n",
       "      <td>29.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>181.0</td>\n",
       "      <td>47.92</td>\n",
       "      <td>24.54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     area_mean  perimeter_mean  texture_mean\n",
       "0       1001.0          122.80         10.38\n",
       "1       1326.0          132.90         17.77\n",
       "2       1203.0          130.00         21.25\n",
       "3        386.1           77.58         20.38\n",
       "4       1297.0          135.10         14.34\n",
       "..         ...             ...           ...\n",
       "564     1479.0          142.00         22.39\n",
       "565     1261.0          131.20         28.25\n",
       "566      858.1          108.30         28.08\n",
       "567     1265.0          140.10         29.33\n",
       "568      181.0           47.92         24.54\n",
       "\n",
       "[569 rows x 3 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a subset of the data with the columns \"area_mean\", \"perimeter_mean\", \"texture_mean\"\n",
    "# (output: a dataframe containing selected columns)\n",
    "\n",
    "bc_data[[\"area_mean\", \"perimeter_mean\", \"texture_mean\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "id": "759b1850-353b-4c89-8d39-13bc7a56a238",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "id                           862548\n",
       "diagnosis                         M\n",
       "radius_mean                   14.42\n",
       "texture_mean                  19.77\n",
       "perimeter_mean                94.48\n",
       "area_mean                     642.5\n",
       "smoothness_mean             0.09752\n",
       "compactness_mean             0.1141\n",
       "concavity_mean              0.09388\n",
       "concave points_mean         0.05839\n",
       "symmetry_mean                0.1879\n",
       "fractal_dimension_mean       0.0639\n",
       "radius_se                    0.2895\n",
       "texture_se                    1.851\n",
       "perimeter_se                  2.376\n",
       "area_se                       26.85\n",
       "smoothness_se              0.008005\n",
       "compactness_se              0.02895\n",
       "concavity_se                0.03321\n",
       "concave points_se           0.01424\n",
       "symmetry_se                 0.01462\n",
       "fractal_dimension_se       0.004452\n",
       "radius_worst                  16.33\n",
       "texture_worst                 30.86\n",
       "perimeter_worst               109.5\n",
       "area_worst                    826.4\n",
       "smoothness_worst             0.1431\n",
       "compactness_worst            0.3026\n",
       "concavity_worst              0.3194\n",
       "concave points_worst         0.1565\n",
       "symmetry_worst               0.2718\n",
       "fractal_dimension_worst     0.09353\n",
       "Unnamed: 32                     NaN\n",
       "Name: 99, dtype: object"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check the information of a given patient by row index, i.e. select row 100\n",
    "\n",
    "bc_data.iloc[99]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "id": "65ca19ae-d5cc-4953-aa5a-da53df5ced16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "      <th>fractal_dimension_worst</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>84501001</td>\n",
       "      <td>M</td>\n",
       "      <td>12.46</td>\n",
       "      <td>24.04</td>\n",
       "      <td>83.97</td>\n",
       "      <td>475.9</td>\n",
       "      <td>0.11860</td>\n",
       "      <td>0.23960</td>\n",
       "      <td>0.22730</td>\n",
       "      <td>0.08543</td>\n",
       "      <td>...</td>\n",
       "      <td>40.68</td>\n",
       "      <td>97.65</td>\n",
       "      <td>711.4</td>\n",
       "      <td>0.1853</td>\n",
       "      <td>1.0580</td>\n",
       "      <td>1.10500</td>\n",
       "      <td>0.22100</td>\n",
       "      <td>0.4366</td>\n",
       "      <td>0.20750</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>845636</td>\n",
       "      <td>M</td>\n",
       "      <td>16.02</td>\n",
       "      <td>23.24</td>\n",
       "      <td>102.70</td>\n",
       "      <td>797.8</td>\n",
       "      <td>0.08206</td>\n",
       "      <td>0.06669</td>\n",
       "      <td>0.03299</td>\n",
       "      <td>0.03323</td>\n",
       "      <td>...</td>\n",
       "      <td>33.88</td>\n",
       "      <td>123.80</td>\n",
       "      <td>1150.0</td>\n",
       "      <td>0.1181</td>\n",
       "      <td>0.1551</td>\n",
       "      <td>0.14590</td>\n",
       "      <td>0.09975</td>\n",
       "      <td>0.2948</td>\n",
       "      <td>0.08452</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>84610002</td>\n",
       "      <td>M</td>\n",
       "      <td>15.78</td>\n",
       "      <td>17.89</td>\n",
       "      <td>103.60</td>\n",
       "      <td>781.0</td>\n",
       "      <td>0.09710</td>\n",
       "      <td>0.12920</td>\n",
       "      <td>0.09954</td>\n",
       "      <td>0.06606</td>\n",
       "      <td>...</td>\n",
       "      <td>27.28</td>\n",
       "      <td>136.50</td>\n",
       "      <td>1299.0</td>\n",
       "      <td>0.1396</td>\n",
       "      <td>0.5609</td>\n",
       "      <td>0.39650</td>\n",
       "      <td>0.18100</td>\n",
       "      <td>0.3792</td>\n",
       "      <td>0.10480</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>846226</td>\n",
       "      <td>M</td>\n",
       "      <td>19.17</td>\n",
       "      <td>24.80</td>\n",
       "      <td>132.40</td>\n",
       "      <td>1123.0</td>\n",
       "      <td>0.09740</td>\n",
       "      <td>0.24580</td>\n",
       "      <td>0.20650</td>\n",
       "      <td>0.11180</td>\n",
       "      <td>...</td>\n",
       "      <td>29.94</td>\n",
       "      <td>151.70</td>\n",
       "      <td>1332.0</td>\n",
       "      <td>0.1037</td>\n",
       "      <td>0.3903</td>\n",
       "      <td>0.36390</td>\n",
       "      <td>0.17670</td>\n",
       "      <td>0.3176</td>\n",
       "      <td>0.10230</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>846381</td>\n",
       "      <td>M</td>\n",
       "      <td>15.85</td>\n",
       "      <td>23.95</td>\n",
       "      <td>103.70</td>\n",
       "      <td>782.7</td>\n",
       "      <td>0.08401</td>\n",
       "      <td>0.10020</td>\n",
       "      <td>0.09938</td>\n",
       "      <td>0.05364</td>\n",
       "      <td>...</td>\n",
       "      <td>27.66</td>\n",
       "      <td>112.00</td>\n",
       "      <td>876.5</td>\n",
       "      <td>0.1131</td>\n",
       "      <td>0.1924</td>\n",
       "      <td>0.23220</td>\n",
       "      <td>0.11190</td>\n",
       "      <td>0.2809</td>\n",
       "      <td>0.06287</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>86135501</td>\n",
       "      <td>M</td>\n",
       "      <td>14.48</td>\n",
       "      <td>21.46</td>\n",
       "      <td>94.25</td>\n",
       "      <td>648.2</td>\n",
       "      <td>0.09444</td>\n",
       "      <td>0.09947</td>\n",
       "      <td>0.12040</td>\n",
       "      <td>0.04938</td>\n",
       "      <td>...</td>\n",
       "      <td>29.25</td>\n",
       "      <td>108.40</td>\n",
       "      <td>808.9</td>\n",
       "      <td>0.1306</td>\n",
       "      <td>0.1976</td>\n",
       "      <td>0.33490</td>\n",
       "      <td>0.12250</td>\n",
       "      <td>0.3020</td>\n",
       "      <td>0.06846</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>86135502</td>\n",
       "      <td>M</td>\n",
       "      <td>19.02</td>\n",
       "      <td>24.59</td>\n",
       "      <td>122.00</td>\n",
       "      <td>1076.0</td>\n",
       "      <td>0.09029</td>\n",
       "      <td>0.12060</td>\n",
       "      <td>0.14680</td>\n",
       "      <td>0.08271</td>\n",
       "      <td>...</td>\n",
       "      <td>30.41</td>\n",
       "      <td>152.90</td>\n",
       "      <td>1623.0</td>\n",
       "      <td>0.1249</td>\n",
       "      <td>0.3206</td>\n",
       "      <td>0.57550</td>\n",
       "      <td>0.19560</td>\n",
       "      <td>0.3956</td>\n",
       "      <td>0.09288</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>861597</td>\n",
       "      <td>B</td>\n",
       "      <td>12.36</td>\n",
       "      <td>21.80</td>\n",
       "      <td>79.78</td>\n",
       "      <td>466.1</td>\n",
       "      <td>0.08772</td>\n",
       "      <td>0.09445</td>\n",
       "      <td>0.06015</td>\n",
       "      <td>0.03745</td>\n",
       "      <td>...</td>\n",
       "      <td>30.50</td>\n",
       "      <td>91.46</td>\n",
       "      <td>574.7</td>\n",
       "      <td>0.1304</td>\n",
       "      <td>0.2463</td>\n",
       "      <td>0.24340</td>\n",
       "      <td>0.12050</td>\n",
       "      <td>0.2972</td>\n",
       "      <td>0.09261</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>861598</td>\n",
       "      <td>B</td>\n",
       "      <td>14.64</td>\n",
       "      <td>15.24</td>\n",
       "      <td>95.77</td>\n",
       "      <td>651.9</td>\n",
       "      <td>0.11320</td>\n",
       "      <td>0.13390</td>\n",
       "      <td>0.09966</td>\n",
       "      <td>0.07064</td>\n",
       "      <td>...</td>\n",
       "      <td>18.24</td>\n",
       "      <td>109.40</td>\n",
       "      <td>803.6</td>\n",
       "      <td>0.1277</td>\n",
       "      <td>0.3089</td>\n",
       "      <td>0.26040</td>\n",
       "      <td>0.13970</td>\n",
       "      <td>0.3151</td>\n",
       "      <td>0.08473</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>861648</td>\n",
       "      <td>B</td>\n",
       "      <td>14.62</td>\n",
       "      <td>24.02</td>\n",
       "      <td>94.57</td>\n",
       "      <td>662.7</td>\n",
       "      <td>0.08974</td>\n",
       "      <td>0.08606</td>\n",
       "      <td>0.03102</td>\n",
       "      <td>0.02957</td>\n",
       "      <td>...</td>\n",
       "      <td>29.11</td>\n",
       "      <td>102.90</td>\n",
       "      <td>803.7</td>\n",
       "      <td>0.1115</td>\n",
       "      <td>0.1766</td>\n",
       "      <td>0.09189</td>\n",
       "      <td>0.06946</td>\n",
       "      <td>0.2522</td>\n",
       "      <td>0.07246</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "9   84501001         M        12.46         24.04           83.97      475.9   \n",
       "10    845636         M        16.02         23.24          102.70      797.8   \n",
       "11  84610002         M        15.78         17.89          103.60      781.0   \n",
       "12    846226         M        19.17         24.80          132.40     1123.0   \n",
       "13    846381         M        15.85         23.95          103.70      782.7   \n",
       "..       ...       ...          ...           ...             ...        ...   \n",
       "86  86135501         M        14.48         21.46           94.25      648.2   \n",
       "87  86135502         M        19.02         24.59          122.00     1076.0   \n",
       "88    861597         B        12.36         21.80           79.78      466.1   \n",
       "89    861598         B        14.64         15.24           95.77      651.9   \n",
       "90    861648         B        14.62         24.02           94.57      662.7   \n",
       "\n",
       "    smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \\\n",
       "9           0.11860           0.23960         0.22730              0.08543   \n",
       "10          0.08206           0.06669         0.03299              0.03323   \n",
       "11          0.09710           0.12920         0.09954              0.06606   \n",
       "12          0.09740           0.24580         0.20650              0.11180   \n",
       "13          0.08401           0.10020         0.09938              0.05364   \n",
       "..              ...               ...             ...                  ...   \n",
       "86          0.09444           0.09947         0.12040              0.04938   \n",
       "87          0.09029           0.12060         0.14680              0.08271   \n",
       "88          0.08772           0.09445         0.06015              0.03745   \n",
       "89          0.11320           0.13390         0.09966              0.07064   \n",
       "90          0.08974           0.08606         0.03102              0.02957   \n",
       "\n",
       "    ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \\\n",
       "9   ...          40.68            97.65       711.4            0.1853   \n",
       "10  ...          33.88           123.80      1150.0            0.1181   \n",
       "11  ...          27.28           136.50      1299.0            0.1396   \n",
       "12  ...          29.94           151.70      1332.0            0.1037   \n",
       "13  ...          27.66           112.00       876.5            0.1131   \n",
       "..  ...            ...              ...         ...               ...   \n",
       "86  ...          29.25           108.40       808.9            0.1306   \n",
       "87  ...          30.41           152.90      1623.0            0.1249   \n",
       "88  ...          30.50            91.46       574.7            0.1304   \n",
       "89  ...          18.24           109.40       803.6            0.1277   \n",
       "90  ...          29.11           102.90       803.7            0.1115   \n",
       "\n",
       "    compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \\\n",
       "9              1.0580          1.10500               0.22100          0.4366   \n",
       "10             0.1551          0.14590               0.09975          0.2948   \n",
       "11             0.5609          0.39650               0.18100          0.3792   \n",
       "12             0.3903          0.36390               0.17670          0.3176   \n",
       "13             0.1924          0.23220               0.11190          0.2809   \n",
       "..                ...              ...                   ...             ...   \n",
       "86             0.1976          0.33490               0.12250          0.3020   \n",
       "87             0.3206          0.57550               0.19560          0.3956   \n",
       "88             0.2463          0.24340               0.12050          0.2972   \n",
       "89             0.3089          0.26040               0.13970          0.3151   \n",
       "90             0.1766          0.09189               0.06946          0.2522   \n",
       "\n",
       "    fractal_dimension_worst  Unnamed: 32  \n",
       "9                   0.20750          NaN  \n",
       "10                  0.08452          NaN  \n",
       "11                  0.10480          NaN  \n",
       "12                  0.10230          NaN  \n",
       "13                  0.06287          NaN  \n",
       "..                      ...          ...  \n",
       "86                  0.06846          NaN  \n",
       "87                  0.09288          NaN  \n",
       "88                  0.09261          NaN  \n",
       "89                  0.08473          NaN  \n",
       "90                  0.07246          NaN  \n",
       "\n",
       "[82 rows x 33 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get a slice of the table by row index\n",
    "# (output: a subtable containing row indexes from 9 (incl.) to 90 (incl.))\n",
    "\n",
    "bc_data.iloc[9:91]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "c741694f-9f3f-4acb-9a6c-431f41092723",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>symmetry_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "      <th>fractal_dimension_worst</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>842302</td>\n",
       "      <td>M</td>\n",
       "      <td>17.99</td>\n",
       "      <td>10.38</td>\n",
       "      <td>122.80</td>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.11840</td>\n",
       "      <td>0.30010</td>\n",
       "      <td>0.14710</td>\n",
       "      <td>0.2419</td>\n",
       "      <td>...</td>\n",
       "      <td>17.33</td>\n",
       "      <td>184.60</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>0.16220</td>\n",
       "      <td>0.66560</td>\n",
       "      <td>0.7119</td>\n",
       "      <td>0.2654</td>\n",
       "      <td>0.4601</td>\n",
       "      <td>0.11890</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>842517</td>\n",
       "      <td>M</td>\n",
       "      <td>20.57</td>\n",
       "      <td>17.77</td>\n",
       "      <td>132.90</td>\n",
       "      <td>1326.0</td>\n",
       "      <td>0.08474</td>\n",
       "      <td>0.08690</td>\n",
       "      <td>0.07017</td>\n",
       "      <td>0.1812</td>\n",
       "      <td>...</td>\n",
       "      <td>23.41</td>\n",
       "      <td>158.80</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>0.12380</td>\n",
       "      <td>0.18660</td>\n",
       "      <td>0.2416</td>\n",
       "      <td>0.1860</td>\n",
       "      <td>0.2750</td>\n",
       "      <td>0.08902</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>84300903</td>\n",
       "      <td>M</td>\n",
       "      <td>19.69</td>\n",
       "      <td>21.25</td>\n",
       "      <td>130.00</td>\n",
       "      <td>1203.0</td>\n",
       "      <td>0.10960</td>\n",
       "      <td>0.19740</td>\n",
       "      <td>0.12790</td>\n",
       "      <td>0.2069</td>\n",
       "      <td>...</td>\n",
       "      <td>25.53</td>\n",
       "      <td>152.50</td>\n",
       "      <td>1709.0</td>\n",
       "      <td>0.14440</td>\n",
       "      <td>0.42450</td>\n",
       "      <td>0.4504</td>\n",
       "      <td>0.2430</td>\n",
       "      <td>0.3613</td>\n",
       "      <td>0.08758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>84348301</td>\n",
       "      <td>M</td>\n",
       "      <td>11.42</td>\n",
       "      <td>20.38</td>\n",
       "      <td>77.58</td>\n",
       "      <td>386.1</td>\n",
       "      <td>0.14250</td>\n",
       "      <td>0.24140</td>\n",
       "      <td>0.10520</td>\n",
       "      <td>0.2597</td>\n",
       "      <td>...</td>\n",
       "      <td>26.50</td>\n",
       "      <td>98.87</td>\n",
       "      <td>567.7</td>\n",
       "      <td>0.20980</td>\n",
       "      <td>0.86630</td>\n",
       "      <td>0.6869</td>\n",
       "      <td>0.2575</td>\n",
       "      <td>0.6638</td>\n",
       "      <td>0.17300</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>84358402</td>\n",
       "      <td>M</td>\n",
       "      <td>20.29</td>\n",
       "      <td>14.34</td>\n",
       "      <td>135.10</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>0.10030</td>\n",
       "      <td>0.19800</td>\n",
       "      <td>0.10430</td>\n",
       "      <td>0.1809</td>\n",
       "      <td>...</td>\n",
       "      <td>16.67</td>\n",
       "      <td>152.20</td>\n",
       "      <td>1575.0</td>\n",
       "      <td>0.13740</td>\n",
       "      <td>0.20500</td>\n",
       "      <td>0.4000</td>\n",
       "      <td>0.1625</td>\n",
       "      <td>0.2364</td>\n",
       "      <td>0.07678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>926424</td>\n",
       "      <td>M</td>\n",
       "      <td>21.56</td>\n",
       "      <td>22.39</td>\n",
       "      <td>142.00</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>0.11100</td>\n",
       "      <td>0.24390</td>\n",
       "      <td>0.13890</td>\n",
       "      <td>0.1726</td>\n",
       "      <td>...</td>\n",
       "      <td>26.40</td>\n",
       "      <td>166.10</td>\n",
       "      <td>2027.0</td>\n",
       "      <td>0.14100</td>\n",
       "      <td>0.21130</td>\n",
       "      <td>0.4107</td>\n",
       "      <td>0.2216</td>\n",
       "      <td>0.2060</td>\n",
       "      <td>0.07115</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>926682</td>\n",
       "      <td>M</td>\n",
       "      <td>20.13</td>\n",
       "      <td>28.25</td>\n",
       "      <td>131.20</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>0.09780</td>\n",
       "      <td>0.14400</td>\n",
       "      <td>0.09791</td>\n",
       "      <td>0.1752</td>\n",
       "      <td>...</td>\n",
       "      <td>38.25</td>\n",
       "      <td>155.00</td>\n",
       "      <td>1731.0</td>\n",
       "      <td>0.11660</td>\n",
       "      <td>0.19220</td>\n",
       "      <td>0.3215</td>\n",
       "      <td>0.1628</td>\n",
       "      <td>0.2572</td>\n",
       "      <td>0.06637</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>926954</td>\n",
       "      <td>M</td>\n",
       "      <td>16.60</td>\n",
       "      <td>28.08</td>\n",
       "      <td>108.30</td>\n",
       "      <td>858.1</td>\n",
       "      <td>0.08455</td>\n",
       "      <td>0.09251</td>\n",
       "      <td>0.05302</td>\n",
       "      <td>0.1590</td>\n",
       "      <td>...</td>\n",
       "      <td>34.12</td>\n",
       "      <td>126.70</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.11390</td>\n",
       "      <td>0.30940</td>\n",
       "      <td>0.3403</td>\n",
       "      <td>0.1418</td>\n",
       "      <td>0.2218</td>\n",
       "      <td>0.07820</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>927241</td>\n",
       "      <td>M</td>\n",
       "      <td>20.60</td>\n",
       "      <td>29.33</td>\n",
       "      <td>140.10</td>\n",
       "      <td>1265.0</td>\n",
       "      <td>0.11780</td>\n",
       "      <td>0.35140</td>\n",
       "      <td>0.15200</td>\n",
       "      <td>0.2397</td>\n",
       "      <td>...</td>\n",
       "      <td>39.42</td>\n",
       "      <td>184.60</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>0.86810</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.2650</td>\n",
       "      <td>0.4087</td>\n",
       "      <td>0.12400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>92751</td>\n",
       "      <td>B</td>\n",
       "      <td>7.76</td>\n",
       "      <td>24.54</td>\n",
       "      <td>47.92</td>\n",
       "      <td>181.0</td>\n",
       "      <td>0.05263</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.1587</td>\n",
       "      <td>...</td>\n",
       "      <td>30.37</td>\n",
       "      <td>59.16</td>\n",
       "      <td>268.6</td>\n",
       "      <td>0.08996</td>\n",
       "      <td>0.06444</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2871</td>\n",
       "      <td>0.07039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "0      842302         M        17.99         10.38          122.80     1001.0   \n",
       "1      842517         M        20.57         17.77          132.90     1326.0   \n",
       "2    84300903         M        19.69         21.25          130.00     1203.0   \n",
       "3    84348301         M        11.42         20.38           77.58      386.1   \n",
       "4    84358402         M        20.29         14.34          135.10     1297.0   \n",
       "..        ...       ...          ...           ...             ...        ...   \n",
       "564    926424         M        21.56         22.39          142.00     1479.0   \n",
       "565    926682         M        20.13         28.25          131.20     1261.0   \n",
       "566    926954         M        16.60         28.08          108.30      858.1   \n",
       "567    927241         M        20.60         29.33          140.10     1265.0   \n",
       "568     92751         B         7.76         24.54           47.92      181.0   \n",
       "\n",
       "     smoothness_mean  concavity_mean  concave points_mean  symmetry_mean  ...  \\\n",
       "0            0.11840         0.30010              0.14710         0.2419  ...   \n",
       "1            0.08474         0.08690              0.07017         0.1812  ...   \n",
       "2            0.10960         0.19740              0.12790         0.2069  ...   \n",
       "3            0.14250         0.24140              0.10520         0.2597  ...   \n",
       "4            0.10030         0.19800              0.10430         0.1809  ...   \n",
       "..               ...             ...                  ...            ...  ...   \n",
       "564          0.11100         0.24390              0.13890         0.1726  ...   \n",
       "565          0.09780         0.14400              0.09791         0.1752  ...   \n",
       "566          0.08455         0.09251              0.05302         0.1590  ...   \n",
       "567          0.11780         0.35140              0.15200         0.2397  ...   \n",
       "568          0.05263         0.00000              0.00000         0.1587  ...   \n",
       "\n",
       "     texture_worst  perimeter_worst  area_worst  smoothness_worst  \\\n",
       "0            17.33           184.60      2019.0           0.16220   \n",
       "1            23.41           158.80      1956.0           0.12380   \n",
       "2            25.53           152.50      1709.0           0.14440   \n",
       "3            26.50            98.87       567.7           0.20980   \n",
       "4            16.67           152.20      1575.0           0.13740   \n",
       "..             ...              ...         ...               ...   \n",
       "564          26.40           166.10      2027.0           0.14100   \n",
       "565          38.25           155.00      1731.0           0.11660   \n",
       "566          34.12           126.70      1124.0           0.11390   \n",
       "567          39.42           184.60      1821.0           0.16500   \n",
       "568          30.37            59.16       268.6           0.08996   \n",
       "\n",
       "     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \\\n",
       "0              0.66560           0.7119                0.2654          0.4601   \n",
       "1              0.18660           0.2416                0.1860          0.2750   \n",
       "2              0.42450           0.4504                0.2430          0.3613   \n",
       "3              0.86630           0.6869                0.2575          0.6638   \n",
       "4              0.20500           0.4000                0.1625          0.2364   \n",
       "..                 ...              ...                   ...             ...   \n",
       "564            0.21130           0.4107                0.2216          0.2060   \n",
       "565            0.19220           0.3215                0.1628          0.2572   \n",
       "566            0.30940           0.3403                0.1418          0.2218   \n",
       "567            0.86810           0.9387                0.2650          0.4087   \n",
       "568            0.06444           0.0000                0.0000          0.2871   \n",
       "\n",
       "     fractal_dimension_worst  Unnamed: 32  \n",
       "0                    0.11890          NaN  \n",
       "1                    0.08902          NaN  \n",
       "2                    0.08758          NaN  \n",
       "3                    0.17300          NaN  \n",
       "4                    0.07678          NaN  \n",
       "..                       ...          ...  \n",
       "564                  0.07115          NaN  \n",
       "565                  0.06637          NaN  \n",
       "566                  0.07820          NaN  \n",
       "567                  0.12400          NaN  \n",
       "568                  0.07039          NaN  \n",
       "\n",
       "[569 rows x 32 columns]"
      ]
     },
     "execution_count": 170,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop columns from the table without overwriting the original data\n",
    "# (output: table without column \"compactness_mean\")\n",
    "drop_data = bc_data.drop(\"compactness_mean\",axis='columns')\n",
    "drop_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "d2deaa8c-974a-40ef-9018-b435bdadee99",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>475.9</td>\n",
       "      <td>0.11860</td>\n",
       "      <td>0.23960</td>\n",
       "      <td>0.22730</td>\n",
       "      <td>0.08543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>797.8</td>\n",
       "      <td>0.08206</td>\n",
       "      <td>0.06669</td>\n",
       "      <td>0.03299</td>\n",
       "      <td>0.03323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>781.0</td>\n",
       "      <td>0.09710</td>\n",
       "      <td>0.12920</td>\n",
       "      <td>0.09954</td>\n",
       "      <td>0.06606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1123.0</td>\n",
       "      <td>0.09740</td>\n",
       "      <td>0.24580</td>\n",
       "      <td>0.20650</td>\n",
       "      <td>0.11180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>782.7</td>\n",
       "      <td>0.08401</td>\n",
       "      <td>0.10020</td>\n",
       "      <td>0.09938</td>\n",
       "      <td>0.05364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>648.2</td>\n",
       "      <td>0.09444</td>\n",
       "      <td>0.09947</td>\n",
       "      <td>0.12040</td>\n",
       "      <td>0.04938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>1076.0</td>\n",
       "      <td>0.09029</td>\n",
       "      <td>0.12060</td>\n",
       "      <td>0.14680</td>\n",
       "      <td>0.08271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>466.1</td>\n",
       "      <td>0.08772</td>\n",
       "      <td>0.09445</td>\n",
       "      <td>0.06015</td>\n",
       "      <td>0.03745</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>651.9</td>\n",
       "      <td>0.11320</td>\n",
       "      <td>0.13390</td>\n",
       "      <td>0.09966</td>\n",
       "      <td>0.07064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>662.7</td>\n",
       "      <td>0.08974</td>\n",
       "      <td>0.08606</td>\n",
       "      <td>0.03102</td>\n",
       "      <td>0.02957</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    area_mean  smoothness_mean  compactness_mean  concavity_mean  \\\n",
       "9       475.9          0.11860           0.23960         0.22730   \n",
       "10      797.8          0.08206           0.06669         0.03299   \n",
       "11      781.0          0.09710           0.12920         0.09954   \n",
       "12     1123.0          0.09740           0.24580         0.20650   \n",
       "13      782.7          0.08401           0.10020         0.09938   \n",
       "..        ...              ...               ...             ...   \n",
       "86      648.2          0.09444           0.09947         0.12040   \n",
       "87     1076.0          0.09029           0.12060         0.14680   \n",
       "88      466.1          0.08772           0.09445         0.06015   \n",
       "89      651.9          0.11320           0.13390         0.09966   \n",
       "90      662.7          0.08974           0.08606         0.03102   \n",
       "\n",
       "    concave points_mean  \n",
       "9               0.08543  \n",
       "10              0.03323  \n",
       "11              0.06606  \n",
       "12              0.11180  \n",
       "13              0.05364  \n",
       "..                  ...  \n",
       "86              0.04938  \n",
       "87              0.08271  \n",
       "88              0.03745  \n",
       "89              0.07064  \n",
       "90              0.02957  \n",
       "\n",
       "[82 rows x 5 columns]"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select slices of lines and columns simultaneously by their index:\n",
    "# - select row indexes 50 to 70 (incl., excl.)\n",
    "# - select column indexes 5 to 10 (incl., excl.)\n",
    "\n",
    "bc_data.iloc[9:91,5:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "64dff750-cb8b-4083-b3ac-86f2c9bceb97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "      <th>fractal_dimension_worst</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        id  diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "0    False      False        False         False           False      False   \n",
       "1    False      False        False         False           False      False   \n",
       "2    False      False        False         False           False      False   \n",
       "3    False      False        False         False           False      False   \n",
       "4    False      False        False         False           False      False   \n",
       "..     ...        ...          ...           ...             ...        ...   \n",
       "564  False      False        False         False           False      False   \n",
       "565  False      False        False         False           False      False   \n",
       "566  False      False        False         False           False      False   \n",
       "567  False      False        False         False           False      False   \n",
       "568  False      False        False         False           False      False   \n",
       "\n",
       "     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \\\n",
       "0              False             False           False                False   \n",
       "1              False             False           False                False   \n",
       "2              False             False           False                False   \n",
       "3              False             False           False                False   \n",
       "4              False             False           False                False   \n",
       "..               ...               ...             ...                  ...   \n",
       "564            False             False           False                False   \n",
       "565            False             False           False                False   \n",
       "566            False             False           False                False   \n",
       "567            False             False           False                False   \n",
       "568            False             False           False                False   \n",
       "\n",
       "     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \\\n",
       "0    ...          False            False       False             False   \n",
       "1    ...          False            False       False             False   \n",
       "2    ...          False            False       False             False   \n",
       "3    ...          False            False       False             False   \n",
       "4    ...          False            False       False             False   \n",
       "..   ...            ...              ...         ...               ...   \n",
       "564  ...          False            False       False             False   \n",
       "565  ...          False            False       False             False   \n",
       "566  ...          False            False       False             False   \n",
       "567  ...          False            False       False             False   \n",
       "568  ...          False            False       False             False   \n",
       "\n",
       "     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \\\n",
       "0                False            False                 False           False   \n",
       "1                False            False                 False           False   \n",
       "2                False            False                 False           False   \n",
       "3                False            False                 False           False   \n",
       "4                False            False                 False           False   \n",
       "..                 ...              ...                   ...             ...   \n",
       "564              False            False                 False           False   \n",
       "565              False            False                 False           False   \n",
       "566              False            False                 False           False   \n",
       "567              False            False                 False           False   \n",
       "568              False            False                 False           False   \n",
       "\n",
       "     fractal_dimension_worst  Unnamed: 32  \n",
       "0                      False         True  \n",
       "1                      False         True  \n",
       "2                      False         True  \n",
       "3                      False         True  \n",
       "4                      False         True  \n",
       "..                       ...          ...  \n",
       "564                    False         True  \n",
       "565                    False         True  \n",
       "566                    False         True  \n",
       "567                    False         True  \n",
       "568                    False         True  \n",
       "\n",
       "[569 rows x 33 columns]"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check for missingness\n",
    "\n",
    "bc_data.isnull()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "id": "6a7e9b43-fed1-4db9-8c69-00b7110ff2db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(569, 33)\n",
      "(569, 32)\n"
     ]
    }
   ],
   "source": [
    "# drop columns containing any number of missing values,\n",
    "# saving the result to a new dataframe\n",
    "print(bc_data.shape)\n",
    "\n",
    "non_null = bc_data.dropna(axis=1, how='all')\n",
    "non_null\n",
    "print(non_null.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "id": "a7430a04-ae4b-44af-837e-c9555dc90fbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>842302</td>\n",
       "      <td>M</td>\n",
       "      <td>17.99</td>\n",
       "      <td>10.38</td>\n",
       "      <td>122.80</td>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.11840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>842517</td>\n",
       "      <td>M</td>\n",
       "      <td>20.57</td>\n",
       "      <td>17.77</td>\n",
       "      <td>132.90</td>\n",
       "      <td>1326.0</td>\n",
       "      <td>0.08474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>84300903</td>\n",
       "      <td>M</td>\n",
       "      <td>19.69</td>\n",
       "      <td>21.25</td>\n",
       "      <td>130.00</td>\n",
       "      <td>1203.0</td>\n",
       "      <td>0.10960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>84348301</td>\n",
       "      <td>M</td>\n",
       "      <td>11.42</td>\n",
       "      <td>20.38</td>\n",
       "      <td>77.58</td>\n",
       "      <td>386.1</td>\n",
       "      <td>0.14250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>84358402</td>\n",
       "      <td>M</td>\n",
       "      <td>20.29</td>\n",
       "      <td>14.34</td>\n",
       "      <td>135.10</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>0.10030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>926424</td>\n",
       "      <td>M</td>\n",
       "      <td>21.56</td>\n",
       "      <td>22.39</td>\n",
       "      <td>142.00</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>0.11100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>926682</td>\n",
       "      <td>M</td>\n",
       "      <td>20.13</td>\n",
       "      <td>28.25</td>\n",
       "      <td>131.20</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>0.09780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>926954</td>\n",
       "      <td>M</td>\n",
       "      <td>16.60</td>\n",
       "      <td>28.08</td>\n",
       "      <td>108.30</td>\n",
       "      <td>858.1</td>\n",
       "      <td>0.08455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>927241</td>\n",
       "      <td>M</td>\n",
       "      <td>20.60</td>\n",
       "      <td>29.33</td>\n",
       "      <td>140.10</td>\n",
       "      <td>1265.0</td>\n",
       "      <td>0.11780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>92751</td>\n",
       "      <td>B</td>\n",
       "      <td>7.76</td>\n",
       "      <td>24.54</td>\n",
       "      <td>47.92</td>\n",
       "      <td>181.0</td>\n",
       "      <td>0.05263</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "0      842302         M        17.99         10.38          122.80     1001.0   \n",
       "1      842517         M        20.57         17.77          132.90     1326.0   \n",
       "2    84300903         M        19.69         21.25          130.00     1203.0   \n",
       "3    84348301         M        11.42         20.38           77.58      386.1   \n",
       "4    84358402         M        20.29         14.34          135.10     1297.0   \n",
       "..        ...       ...          ...           ...             ...        ...   \n",
       "564    926424         M        21.56         22.39          142.00     1479.0   \n",
       "565    926682         M        20.13         28.25          131.20     1261.0   \n",
       "566    926954         M        16.60         28.08          108.30      858.1   \n",
       "567    927241         M        20.60         29.33          140.10     1265.0   \n",
       "568     92751         B         7.76         24.54           47.92      181.0   \n",
       "\n",
       "     smoothness_mean  \n",
       "0            0.11840  \n",
       "1            0.08474  \n",
       "2            0.10960  \n",
       "3            0.14250  \n",
       "4            0.10030  \n",
       "..               ...  \n",
       "564          0.11100  \n",
       "565          0.09780  \n",
       "566          0.08455  \n",
       "567          0.11780  \n",
       "568          0.05263  \n",
       "\n",
       "[569 rows x 7 columns]"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# create a dataframe named \"bc_data_subset\" containing all the rows and only the first 7 columns\n",
    "\n",
    "bc_data_subset = bc_data.iloc[:,0:7]\n",
    "bc_data_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "cc6ff80e-63bc-4669-99cd-4bc773cddef8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius</th>\n",
       "      <th>texture</th>\n",
       "      <th>perimeter</th>\n",
       "      <th>area</th>\n",
       "      <th>smoothness</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>842302</td>\n",
       "      <td>M</td>\n",
       "      <td>17.99</td>\n",
       "      <td>10.38</td>\n",
       "      <td>122.80</td>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.11840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>842517</td>\n",
       "      <td>M</td>\n",
       "      <td>20.57</td>\n",
       "      <td>17.77</td>\n",
       "      <td>132.90</td>\n",
       "      <td>1326.0</td>\n",
       "      <td>0.08474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>84300903</td>\n",
       "      <td>M</td>\n",
       "      <td>19.69</td>\n",
       "      <td>21.25</td>\n",
       "      <td>130.00</td>\n",
       "      <td>1203.0</td>\n",
       "      <td>0.10960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>84348301</td>\n",
       "      <td>M</td>\n",
       "      <td>11.42</td>\n",
       "      <td>20.38</td>\n",
       "      <td>77.58</td>\n",
       "      <td>386.1</td>\n",
       "      <td>0.14250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>84358402</td>\n",
       "      <td>M</td>\n",
       "      <td>20.29</td>\n",
       "      <td>14.34</td>\n",
       "      <td>135.10</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>0.10030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>926424</td>\n",
       "      <td>M</td>\n",
       "      <td>21.56</td>\n",
       "      <td>22.39</td>\n",
       "      <td>142.00</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>0.11100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>926682</td>\n",
       "      <td>M</td>\n",
       "      <td>20.13</td>\n",
       "      <td>28.25</td>\n",
       "      <td>131.20</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>0.09780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>926954</td>\n",
       "      <td>M</td>\n",
       "      <td>16.60</td>\n",
       "      <td>28.08</td>\n",
       "      <td>108.30</td>\n",
       "      <td>858.1</td>\n",
       "      <td>0.08455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>927241</td>\n",
       "      <td>M</td>\n",
       "      <td>20.60</td>\n",
       "      <td>29.33</td>\n",
       "      <td>140.10</td>\n",
       "      <td>1265.0</td>\n",
       "      <td>0.11780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>92751</td>\n",
       "      <td>B</td>\n",
       "      <td>7.76</td>\n",
       "      <td>24.54</td>\n",
       "      <td>47.92</td>\n",
       "      <td>181.0</td>\n",
       "      <td>0.05263</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id diagnosis  radius  texture  perimeter    area  smoothness\n",
       "0      842302         M   17.99    10.38     122.80  1001.0     0.11840\n",
       "1      842517         M   20.57    17.77     132.90  1326.0     0.08474\n",
       "2    84300903         M   19.69    21.25     130.00  1203.0     0.10960\n",
       "3    84348301         M   11.42    20.38      77.58   386.1     0.14250\n",
       "4    84358402         M   20.29    14.34     135.10  1297.0     0.10030\n",
       "..        ...       ...     ...      ...        ...     ...         ...\n",
       "564    926424         M   21.56    22.39     142.00  1479.0     0.11100\n",
       "565    926682         M   20.13    28.25     131.20  1261.0     0.09780\n",
       "566    926954         M   16.60    28.08     108.30   858.1     0.08455\n",
       "567    927241         M   20.60    29.33     140.10  1265.0     0.11780\n",
       "568     92751         B    7.76    24.54      47.92   181.0     0.05263\n",
       "\n",
       "[569 rows x 7 columns]"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rename the columns with better/shorter names,\n",
    "# i.e., without the _mean suffix\n",
    "\n",
    "bc_data_subset.rename(columns={'radius_mean': 'radius', 'texture_mean': 'texture',\n",
    "                               'perimeter_mean':'perimeter', 'area_mean':'area',\n",
    "                              'smoothness_mean':'smoothness'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74ea9efe-9739-4d15-9d6a-ac4bf73096d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cond = df_gene[\"GTEX-1117F\"] >= 5.\n",
    "df_gene.loc[cond]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "id": "dc5ed45a-8bf8-4fbe-a49d-b23f32b6ae1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>symmetry_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "      <th>fractal_dimension_worst</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>842302</th>\n",
       "      <td>M</td>\n",
       "      <td>17.99</td>\n",
       "      <td>10.38</td>\n",
       "      <td>122.80</td>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.11840</td>\n",
       "      <td>0.27760</td>\n",
       "      <td>0.30010</td>\n",
       "      <td>0.14710</td>\n",
       "      <td>0.2419</td>\n",
       "      <td>...</td>\n",
       "      <td>17.33</td>\n",
       "      <td>184.60</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>0.16220</td>\n",
       "      <td>0.66560</td>\n",
       "      <td>0.7119</td>\n",
       "      <td>0.2654</td>\n",
       "      <td>0.4601</td>\n",
       "      <td>0.11890</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>842517</th>\n",
       "      <td>M</td>\n",
       "      <td>20.57</td>\n",
       "      <td>17.77</td>\n",
       "      <td>132.90</td>\n",
       "      <td>1326.0</td>\n",
       "      <td>0.08474</td>\n",
       "      <td>0.07864</td>\n",
       "      <td>0.08690</td>\n",
       "      <td>0.07017</td>\n",
       "      <td>0.1812</td>\n",
       "      <td>...</td>\n",
       "      <td>23.41</td>\n",
       "      <td>158.80</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>0.12380</td>\n",
       "      <td>0.18660</td>\n",
       "      <td>0.2416</td>\n",
       "      <td>0.1860</td>\n",
       "      <td>0.2750</td>\n",
       "      <td>0.08902</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84300903</th>\n",
       "      <td>M</td>\n",
       "      <td>19.69</td>\n",
       "      <td>21.25</td>\n",
       "      <td>130.00</td>\n",
       "      <td>1203.0</td>\n",
       "      <td>0.10960</td>\n",
       "      <td>0.15990</td>\n",
       "      <td>0.19740</td>\n",
       "      <td>0.12790</td>\n",
       "      <td>0.2069</td>\n",
       "      <td>...</td>\n",
       "      <td>25.53</td>\n",
       "      <td>152.50</td>\n",
       "      <td>1709.0</td>\n",
       "      <td>0.14440</td>\n",
       "      <td>0.42450</td>\n",
       "      <td>0.4504</td>\n",
       "      <td>0.2430</td>\n",
       "      <td>0.3613</td>\n",
       "      <td>0.08758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84348301</th>\n",
       "      <td>M</td>\n",
       "      <td>11.42</td>\n",
       "      <td>20.38</td>\n",
       "      <td>77.58</td>\n",
       "      <td>386.1</td>\n",
       "      <td>0.14250</td>\n",
       "      <td>0.28390</td>\n",
       "      <td>0.24140</td>\n",
       "      <td>0.10520</td>\n",
       "      <td>0.2597</td>\n",
       "      <td>...</td>\n",
       "      <td>26.50</td>\n",
       "      <td>98.87</td>\n",
       "      <td>567.7</td>\n",
       "      <td>0.20980</td>\n",
       "      <td>0.86630</td>\n",
       "      <td>0.6869</td>\n",
       "      <td>0.2575</td>\n",
       "      <td>0.6638</td>\n",
       "      <td>0.17300</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84358402</th>\n",
       "      <td>M</td>\n",
       "      <td>20.29</td>\n",
       "      <td>14.34</td>\n",
       "      <td>135.10</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>0.10030</td>\n",
       "      <td>0.13280</td>\n",
       "      <td>0.19800</td>\n",
       "      <td>0.10430</td>\n",
       "      <td>0.1809</td>\n",
       "      <td>...</td>\n",
       "      <td>16.67</td>\n",
       "      <td>152.20</td>\n",
       "      <td>1575.0</td>\n",
       "      <td>0.13740</td>\n",
       "      <td>0.20500</td>\n",
       "      <td>0.4000</td>\n",
       "      <td>0.1625</td>\n",
       "      <td>0.2364</td>\n",
       "      <td>0.07678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926424</th>\n",
       "      <td>M</td>\n",
       "      <td>21.56</td>\n",
       "      <td>22.39</td>\n",
       "      <td>142.00</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>0.11100</td>\n",
       "      <td>0.11590</td>\n",
       "      <td>0.24390</td>\n",
       "      <td>0.13890</td>\n",
       "      <td>0.1726</td>\n",
       "      <td>...</td>\n",
       "      <td>26.40</td>\n",
       "      <td>166.10</td>\n",
       "      <td>2027.0</td>\n",
       "      <td>0.14100</td>\n",
       "      <td>0.21130</td>\n",
       "      <td>0.4107</td>\n",
       "      <td>0.2216</td>\n",
       "      <td>0.2060</td>\n",
       "      <td>0.07115</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926682</th>\n",
       "      <td>M</td>\n",
       "      <td>20.13</td>\n",
       "      <td>28.25</td>\n",
       "      <td>131.20</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>0.09780</td>\n",
       "      <td>0.10340</td>\n",
       "      <td>0.14400</td>\n",
       "      <td>0.09791</td>\n",
       "      <td>0.1752</td>\n",
       "      <td>...</td>\n",
       "      <td>38.25</td>\n",
       "      <td>155.00</td>\n",
       "      <td>1731.0</td>\n",
       "      <td>0.11660</td>\n",
       "      <td>0.19220</td>\n",
       "      <td>0.3215</td>\n",
       "      <td>0.1628</td>\n",
       "      <td>0.2572</td>\n",
       "      <td>0.06637</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926954</th>\n",
       "      <td>M</td>\n",
       "      <td>16.60</td>\n",
       "      <td>28.08</td>\n",
       "      <td>108.30</td>\n",
       "      <td>858.1</td>\n",
       "      <td>0.08455</td>\n",
       "      <td>0.10230</td>\n",
       "      <td>0.09251</td>\n",
       "      <td>0.05302</td>\n",
       "      <td>0.1590</td>\n",
       "      <td>...</td>\n",
       "      <td>34.12</td>\n",
       "      <td>126.70</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.11390</td>\n",
       "      <td>0.30940</td>\n",
       "      <td>0.3403</td>\n",
       "      <td>0.1418</td>\n",
       "      <td>0.2218</td>\n",
       "      <td>0.07820</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>927241</th>\n",
       "      <td>M</td>\n",
       "      <td>20.60</td>\n",
       "      <td>29.33</td>\n",
       "      <td>140.10</td>\n",
       "      <td>1265.0</td>\n",
       "      <td>0.11780</td>\n",
       "      <td>0.27700</td>\n",
       "      <td>0.35140</td>\n",
       "      <td>0.15200</td>\n",
       "      <td>0.2397</td>\n",
       "      <td>...</td>\n",
       "      <td>39.42</td>\n",
       "      <td>184.60</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>0.86810</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.2650</td>\n",
       "      <td>0.4087</td>\n",
       "      <td>0.12400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92751</th>\n",
       "      <td>B</td>\n",
       "      <td>7.76</td>\n",
       "      <td>24.54</td>\n",
       "      <td>47.92</td>\n",
       "      <td>181.0</td>\n",
       "      <td>0.05263</td>\n",
       "      <td>0.04362</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.1587</td>\n",
       "      <td>...</td>\n",
       "      <td>30.37</td>\n",
       "      <td>59.16</td>\n",
       "      <td>268.6</td>\n",
       "      <td>0.08996</td>\n",
       "      <td>0.06444</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2871</td>\n",
       "      <td>0.07039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "id                                                                         \n",
       "842302           M        17.99         10.38          122.80     1001.0   \n",
       "842517           M        20.57         17.77          132.90     1326.0   \n",
       "84300903         M        19.69         21.25          130.00     1203.0   \n",
       "84348301         M        11.42         20.38           77.58      386.1   \n",
       "84358402         M        20.29         14.34          135.10     1297.0   \n",
       "...            ...          ...           ...             ...        ...   \n",
       "926424           M        21.56         22.39          142.00     1479.0   \n",
       "926682           M        20.13         28.25          131.20     1261.0   \n",
       "926954           M        16.60         28.08          108.30      858.1   \n",
       "927241           M        20.60         29.33          140.10     1265.0   \n",
       "92751            B         7.76         24.54           47.92      181.0   \n",
       "\n",
       "          smoothness_mean  compactness_mean  concavity_mean  \\\n",
       "id                                                            \n",
       "842302            0.11840           0.27760         0.30010   \n",
       "842517            0.08474           0.07864         0.08690   \n",
       "84300903          0.10960           0.15990         0.19740   \n",
       "84348301          0.14250           0.28390         0.24140   \n",
       "84358402          0.10030           0.13280         0.19800   \n",
       "...                   ...               ...             ...   \n",
       "926424            0.11100           0.11590         0.24390   \n",
       "926682            0.09780           0.10340         0.14400   \n",
       "926954            0.08455           0.10230         0.09251   \n",
       "927241            0.11780           0.27700         0.35140   \n",
       "92751             0.05263           0.04362         0.00000   \n",
       "\n",
       "          concave points_mean  symmetry_mean  ...  texture_worst  \\\n",
       "id                                            ...                  \n",
       "842302                0.14710         0.2419  ...          17.33   \n",
       "842517                0.07017         0.1812  ...          23.41   \n",
       "84300903              0.12790         0.2069  ...          25.53   \n",
       "84348301              0.10520         0.2597  ...          26.50   \n",
       "84358402              0.10430         0.1809  ...          16.67   \n",
       "...                       ...            ...  ...            ...   \n",
       "926424                0.13890         0.1726  ...          26.40   \n",
       "926682                0.09791         0.1752  ...          38.25   \n",
       "926954                0.05302         0.1590  ...          34.12   \n",
       "927241                0.15200         0.2397  ...          39.42   \n",
       "92751                 0.00000         0.1587  ...          30.37   \n",
       "\n",
       "          perimeter_worst  area_worst  smoothness_worst  compactness_worst  \\\n",
       "id                                                                           \n",
       "842302             184.60      2019.0           0.16220            0.66560   \n",
       "842517             158.80      1956.0           0.12380            0.18660   \n",
       "84300903           152.50      1709.0           0.14440            0.42450   \n",
       "84348301            98.87       567.7           0.20980            0.86630   \n",
       "84358402           152.20      1575.0           0.13740            0.20500   \n",
       "...                   ...         ...               ...                ...   \n",
       "926424             166.10      2027.0           0.14100            0.21130   \n",
       "926682             155.00      1731.0           0.11660            0.19220   \n",
       "926954             126.70      1124.0           0.11390            0.30940   \n",
       "927241             184.60      1821.0           0.16500            0.86810   \n",
       "92751               59.16       268.6           0.08996            0.06444   \n",
       "\n",
       "          concavity_worst  concave points_worst  symmetry_worst  \\\n",
       "id                                                                \n",
       "842302             0.7119                0.2654          0.4601   \n",
       "842517             0.2416                0.1860          0.2750   \n",
       "84300903           0.4504                0.2430          0.3613   \n",
       "84348301           0.6869                0.2575          0.6638   \n",
       "84358402           0.4000                0.1625          0.2364   \n",
       "...                   ...                   ...             ...   \n",
       "926424             0.4107                0.2216          0.2060   \n",
       "926682             0.3215                0.1628          0.2572   \n",
       "926954             0.3403                0.1418          0.2218   \n",
       "927241             0.9387                0.2650          0.4087   \n",
       "92751              0.0000                0.0000          0.2871   \n",
       "\n",
       "          fractal_dimension_worst  Unnamed: 32  \n",
       "id                                              \n",
       "842302                    0.11890          NaN  \n",
       "842517                    0.08902          NaN  \n",
       "84300903                  0.08758          NaN  \n",
       "84348301                  0.17300          NaN  \n",
       "84358402                  0.07678          NaN  \n",
       "...                           ...          ...  \n",
       "926424                    0.07115          NaN  \n",
       "926682                    0.06637          NaN  \n",
       "926954                    0.07820          NaN  \n",
       "927241                    0.12400          NaN  \n",
       "92751                     0.07039          NaN  \n",
       "\n",
       "[569 rows x 32 columns]"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc_data.head()\n",
    "bc_data.set_index(\"id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "id": "77df450d-5402-4844-ba4c-bce248d8251f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        842302\n",
       "1        842517\n",
       "2      84300903\n",
       "4      84358402\n",
       "6        844359\n",
       "         ...   \n",
       "535      919555\n",
       "563      926125\n",
       "564      926424\n",
       "565      926682\n",
       "567      927241\n",
       "Name: id, Length: 92, dtype: int64"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select patients that showed tumors with an area superior to 1000\n",
    "condi = bc_data[\"area_mean\"] >1000\n",
    "subset_bc = bc_data.loc[condi]\n",
    "subset_bc.id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e8fe5b9-c46f-4021-9831-ff260fec86ed",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "id": "964e18a2-7b7a-4def-b141-e9694568eacc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(92, 33)"
      ]
     },
     "execution_count": 199,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# of those tumors with an area superior to 1000, how many had a perimeter superior to 180?\n",
    "subset_bc.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "id": "02b95592-7860-4cad-b13a-f28f5e533e95",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3, 33)"
      ]
     },
     "execution_count": 201,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "large_bc = subset_bc.loc[subset_bc[\"perimeter_mean\"]>180]\n",
    "large_bc.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "id": "e3b98b7a-8358-464b-b5d1-eb320cbdf90d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "diagnosis\n",
       "B    462.790196\n",
       "M    978.376415\n",
       "Name: area_mean, dtype: float64"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# what is the mean tumor area of patients for each diagnosis (Benign/Malignant)?\n",
    "\n",
    "bc_data.groupby(\"diagnosis\")[\"area_mean\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "id": "96e76dcc-a1d4-4df3-a465-f6fdeeac7f05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "      <th>fractal_dimension_worst</th>\n",
       "      <th>Unnamed: 32</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>842302</td>\n",
       "      <td>M</td>\n",
       "      <td>17.99</td>\n",
       "      <td>10.38</td>\n",
       "      <td>122.80</td>\n",
       "      <td>1001.0</td>\n",
       "      <td>0.11840</td>\n",
       "      <td>0.27760</td>\n",
       "      <td>0.30010</td>\n",
       "      <td>0.14710</td>\n",
       "      <td>...</td>\n",
       "      <td>17.33</td>\n",
       "      <td>184.60</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>0.16220</td>\n",
       "      <td>0.66560</td>\n",
       "      <td>0.7119</td>\n",
       "      <td>0.2654</td>\n",
       "      <td>0.4601</td>\n",
       "      <td>0.11890</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>842517</td>\n",
       "      <td>M</td>\n",
       "      <td>20.57</td>\n",
       "      <td>17.77</td>\n",
       "      <td>132.90</td>\n",
       "      <td>1326.0</td>\n",
       "      <td>0.08474</td>\n",
       "      <td>0.07864</td>\n",
       "      <td>0.08690</td>\n",
       "      <td>0.07017</td>\n",
       "      <td>...</td>\n",
       "      <td>23.41</td>\n",
       "      <td>158.80</td>\n",
       "      <td>1956.0</td>\n",
       "      <td>0.12380</td>\n",
       "      <td>0.18660</td>\n",
       "      <td>0.2416</td>\n",
       "      <td>0.1860</td>\n",
       "      <td>0.2750</td>\n",
       "      <td>0.08902</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>84300903</td>\n",
       "      <td>M</td>\n",
       "      <td>19.69</td>\n",
       "      <td>21.25</td>\n",
       "      <td>130.00</td>\n",
       "      <td>1203.0</td>\n",
       "      <td>0.10960</td>\n",
       "      <td>0.15990</td>\n",
       "      <td>0.19740</td>\n",
       "      <td>0.12790</td>\n",
       "      <td>...</td>\n",
       "      <td>25.53</td>\n",
       "      <td>152.50</td>\n",
       "      <td>1709.0</td>\n",
       "      <td>0.14440</td>\n",
       "      <td>0.42450</td>\n",
       "      <td>0.4504</td>\n",
       "      <td>0.2430</td>\n",
       "      <td>0.3613</td>\n",
       "      <td>0.08758</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>84348301</td>\n",
       "      <td>M</td>\n",
       "      <td>11.42</td>\n",
       "      <td>20.38</td>\n",
       "      <td>77.58</td>\n",
       "      <td>386.1</td>\n",
       "      <td>0.14250</td>\n",
       "      <td>0.28390</td>\n",
       "      <td>0.24140</td>\n",
       "      <td>0.10520</td>\n",
       "      <td>...</td>\n",
       "      <td>26.50</td>\n",
       "      <td>98.87</td>\n",
       "      <td>567.7</td>\n",
       "      <td>0.20980</td>\n",
       "      <td>0.86630</td>\n",
       "      <td>0.6869</td>\n",
       "      <td>0.2575</td>\n",
       "      <td>0.6638</td>\n",
       "      <td>0.17300</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>84358402</td>\n",
       "      <td>M</td>\n",
       "      <td>20.29</td>\n",
       "      <td>14.34</td>\n",
       "      <td>135.10</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>0.10030</td>\n",
       "      <td>0.13280</td>\n",
       "      <td>0.19800</td>\n",
       "      <td>0.10430</td>\n",
       "      <td>...</td>\n",
       "      <td>16.67</td>\n",
       "      <td>152.20</td>\n",
       "      <td>1575.0</td>\n",
       "      <td>0.13740</td>\n",
       "      <td>0.20500</td>\n",
       "      <td>0.4000</td>\n",
       "      <td>0.1625</td>\n",
       "      <td>0.2364</td>\n",
       "      <td>0.07678</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>564</th>\n",
       "      <td>926424</td>\n",
       "      <td>M</td>\n",
       "      <td>21.56</td>\n",
       "      <td>22.39</td>\n",
       "      <td>142.00</td>\n",
       "      <td>1479.0</td>\n",
       "      <td>0.11100</td>\n",
       "      <td>0.11590</td>\n",
       "      <td>0.24390</td>\n",
       "      <td>0.13890</td>\n",
       "      <td>...</td>\n",
       "      <td>26.40</td>\n",
       "      <td>166.10</td>\n",
       "      <td>2027.0</td>\n",
       "      <td>0.14100</td>\n",
       "      <td>0.21130</td>\n",
       "      <td>0.4107</td>\n",
       "      <td>0.2216</td>\n",
       "      <td>0.2060</td>\n",
       "      <td>0.07115</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>565</th>\n",
       "      <td>926682</td>\n",
       "      <td>M</td>\n",
       "      <td>20.13</td>\n",
       "      <td>28.25</td>\n",
       "      <td>131.20</td>\n",
       "      <td>1261.0</td>\n",
       "      <td>0.09780</td>\n",
       "      <td>0.10340</td>\n",
       "      <td>0.14400</td>\n",
       "      <td>0.09791</td>\n",
       "      <td>...</td>\n",
       "      <td>38.25</td>\n",
       "      <td>155.00</td>\n",
       "      <td>1731.0</td>\n",
       "      <td>0.11660</td>\n",
       "      <td>0.19220</td>\n",
       "      <td>0.3215</td>\n",
       "      <td>0.1628</td>\n",
       "      <td>0.2572</td>\n",
       "      <td>0.06637</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>566</th>\n",
       "      <td>926954</td>\n",
       "      <td>M</td>\n",
       "      <td>16.60</td>\n",
       "      <td>28.08</td>\n",
       "      <td>108.30</td>\n",
       "      <td>858.1</td>\n",
       "      <td>0.08455</td>\n",
       "      <td>0.10230</td>\n",
       "      <td>0.09251</td>\n",
       "      <td>0.05302</td>\n",
       "      <td>...</td>\n",
       "      <td>34.12</td>\n",
       "      <td>126.70</td>\n",
       "      <td>1124.0</td>\n",
       "      <td>0.11390</td>\n",
       "      <td>0.30940</td>\n",
       "      <td>0.3403</td>\n",
       "      <td>0.1418</td>\n",
       "      <td>0.2218</td>\n",
       "      <td>0.07820</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>567</th>\n",
       "      <td>927241</td>\n",
       "      <td>M</td>\n",
       "      <td>20.60</td>\n",
       "      <td>29.33</td>\n",
       "      <td>140.10</td>\n",
       "      <td>1265.0</td>\n",
       "      <td>0.11780</td>\n",
       "      <td>0.27700</td>\n",
       "      <td>0.35140</td>\n",
       "      <td>0.15200</td>\n",
       "      <td>...</td>\n",
       "      <td>39.42</td>\n",
       "      <td>184.60</td>\n",
       "      <td>1821.0</td>\n",
       "      <td>0.16500</td>\n",
       "      <td>0.86810</td>\n",
       "      <td>0.9387</td>\n",
       "      <td>0.2650</td>\n",
       "      <td>0.4087</td>\n",
       "      <td>0.12400</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>568</th>\n",
       "      <td>92751</td>\n",
       "      <td>B</td>\n",
       "      <td>7.76</td>\n",
       "      <td>24.54</td>\n",
       "      <td>47.92</td>\n",
       "      <td>181.0</td>\n",
       "      <td>0.05263</td>\n",
       "      <td>0.04362</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>30.37</td>\n",
       "      <td>59.16</td>\n",
       "      <td>268.6</td>\n",
       "      <td>0.08996</td>\n",
       "      <td>0.06444</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2871</td>\n",
       "      <td>0.07039</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>569 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "0      842302         M        17.99         10.38          122.80     1001.0   \n",
       "1      842517         M        20.57         17.77          132.90     1326.0   \n",
       "2    84300903         M        19.69         21.25          130.00     1203.0   \n",
       "3    84348301         M        11.42         20.38           77.58      386.1   \n",
       "4    84358402         M        20.29         14.34          135.10     1297.0   \n",
       "..        ...       ...          ...           ...             ...        ...   \n",
       "564    926424         M        21.56         22.39          142.00     1479.0   \n",
       "565    926682         M        20.13         28.25          131.20     1261.0   \n",
       "566    926954         M        16.60         28.08          108.30      858.1   \n",
       "567    927241         M        20.60         29.33          140.10     1265.0   \n",
       "568     92751         B         7.76         24.54           47.92      181.0   \n",
       "\n",
       "     smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \\\n",
       "0            0.11840           0.27760         0.30010              0.14710   \n",
       "1            0.08474           0.07864         0.08690              0.07017   \n",
       "2            0.10960           0.15990         0.19740              0.12790   \n",
       "3            0.14250           0.28390         0.24140              0.10520   \n",
       "4            0.10030           0.13280         0.19800              0.10430   \n",
       "..               ...               ...             ...                  ...   \n",
       "564          0.11100           0.11590         0.24390              0.13890   \n",
       "565          0.09780           0.10340         0.14400              0.09791   \n",
       "566          0.08455           0.10230         0.09251              0.05302   \n",
       "567          0.11780           0.27700         0.35140              0.15200   \n",
       "568          0.05263           0.04362         0.00000              0.00000   \n",
       "\n",
       "     ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \\\n",
       "0    ...          17.33           184.60      2019.0           0.16220   \n",
       "1    ...          23.41           158.80      1956.0           0.12380   \n",
       "2    ...          25.53           152.50      1709.0           0.14440   \n",
       "3    ...          26.50            98.87       567.7           0.20980   \n",
       "4    ...          16.67           152.20      1575.0           0.13740   \n",
       "..   ...            ...              ...         ...               ...   \n",
       "564  ...          26.40           166.10      2027.0           0.14100   \n",
       "565  ...          38.25           155.00      1731.0           0.11660   \n",
       "566  ...          34.12           126.70      1124.0           0.11390   \n",
       "567  ...          39.42           184.60      1821.0           0.16500   \n",
       "568  ...          30.37            59.16       268.6           0.08996   \n",
       "\n",
       "     compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \\\n",
       "0              0.66560           0.7119                0.2654          0.4601   \n",
       "1              0.18660           0.2416                0.1860          0.2750   \n",
       "2              0.42450           0.4504                0.2430          0.3613   \n",
       "3              0.86630           0.6869                0.2575          0.6638   \n",
       "4              0.20500           0.4000                0.1625          0.2364   \n",
       "..                 ...              ...                   ...             ...   \n",
       "564            0.21130           0.4107                0.2216          0.2060   \n",
       "565            0.19220           0.3215                0.1628          0.2572   \n",
       "566            0.30940           0.3403                0.1418          0.2218   \n",
       "567            0.86810           0.9387                0.2650          0.4087   \n",
       "568            0.06444           0.0000                0.0000          0.2871   \n",
       "\n",
       "     fractal_dimension_worst  Unnamed: 32  \n",
       "0                    0.11890          NaN  \n",
       "1                    0.08902          NaN  \n",
       "2                    0.08758          NaN  \n",
       "3                    0.17300          NaN  \n",
       "4                    0.07678          NaN  \n",
       "..                       ...          ...  \n",
       "564                  0.07115          NaN  \n",
       "565                  0.06637          NaN  \n",
       "566                  0.07820          NaN  \n",
       "567                  0.12400          NaN  \n",
       "568                  0.07039          NaN  \n",
       "\n",
       "[569 rows x 33 columns]"
      ]
     },
     "execution_count": 206,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc_data.reset_index().drop(\"index\",axis='columns')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "5bd5e4dd-5636-4211-b972-3dfd41a55a09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "diagnosis\n",
       "B    458.4\n",
       "M    932.0\n",
       "Name: area_mean, dtype: float64"
      ]
     },
     "execution_count": 207,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# and the median?\n",
    "bc_data.groupby(\"diagnosis\")[\"area_mean\"].median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "eb0e0ed0-f470-4294-b946-1994f8864ec7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['id',\n",
       " 'diagnosis',\n",
       " 'radius_mean',\n",
       " 'texture_mean',\n",
       " 'perimeter_mean',\n",
       " 'area_mean',\n",
       " 'smoothness_mean',\n",
       " 'compactness_mean',\n",
       " 'concavity_mean',\n",
       " 'concave points_mean',\n",
       " 'symmetry_mean',\n",
       " 'fractal_dimension_mean',\n",
       " 'radius_se',\n",
       " 'texture_se',\n",
       " 'perimeter_se',\n",
       " 'area_se',\n",
       " 'smoothness_se',\n",
       " 'compactness_se',\n",
       " 'concavity_se',\n",
       " 'concave points_se',\n",
       " 'symmetry_se',\n",
       " 'fractal_dimension_se',\n",
       " 'radius_worst',\n",
       " 'texture_worst',\n",
       " 'perimeter_worst',\n",
       " 'area_worst',\n",
       " 'smoothness_worst',\n",
       " 'compactness_worst',\n",
       " 'concavity_worst',\n",
       " 'concave points_worst',\n",
       " 'symmetry_worst',\n",
       " 'fractal_dimension_worst',\n",
       " 'Unnamed: 32']"
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bc_data.columns.tolist() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "185373d3-8e3d-475a-bdf9-f161b1ffd35d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# what is the median value of all parameters for each level of diagnosis?\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "id": "b9c7e73a-09eb-478d-bef0-aa72e7a13e99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2, 29)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>radius_mean</th>\n",
       "      <th>texture_mean</th>\n",
       "      <th>perimeter_mean</th>\n",
       "      <th>area_mean</th>\n",
       "      <th>smoothness_mean</th>\n",
       "      <th>compactness_mean</th>\n",
       "      <th>concavity_mean</th>\n",
       "      <th>concave points_mean</th>\n",
       "      <th>symmetry_mean</th>\n",
       "      <th>fractal_dimension_mean</th>\n",
       "      <th>...</th>\n",
       "      <th>fractal_dimension_se</th>\n",
       "      <th>radius_worst</th>\n",
       "      <th>texture_worst</th>\n",
       "      <th>perimeter_worst</th>\n",
       "      <th>area_worst</th>\n",
       "      <th>smoothness_worst</th>\n",
       "      <th>compactness_worst</th>\n",
       "      <th>concavity_worst</th>\n",
       "      <th>concave points_worst</th>\n",
       "      <th>symmetry_worst</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>diagnosis</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>12.200</td>\n",
       "      <td>17.39</td>\n",
       "      <td>78.18</td>\n",
       "      <td>458.4</td>\n",
       "      <td>0.09076</td>\n",
       "      <td>0.07529</td>\n",
       "      <td>0.03709</td>\n",
       "      <td>0.02344</td>\n",
       "      <td>0.1714</td>\n",
       "      <td>0.061540</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002808</td>\n",
       "      <td>13.35</td>\n",
       "      <td>22.820</td>\n",
       "      <td>86.92</td>\n",
       "      <td>547.4</td>\n",
       "      <td>0.12540</td>\n",
       "      <td>0.16980</td>\n",
       "      <td>0.1412</td>\n",
       "      <td>0.07431</td>\n",
       "      <td>0.2687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>17.325</td>\n",
       "      <td>21.46</td>\n",
       "      <td>114.20</td>\n",
       "      <td>932.0</td>\n",
       "      <td>0.10220</td>\n",
       "      <td>0.13235</td>\n",
       "      <td>0.15135</td>\n",
       "      <td>0.08628</td>\n",
       "      <td>0.1899</td>\n",
       "      <td>0.061575</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003739</td>\n",
       "      <td>20.59</td>\n",
       "      <td>28.945</td>\n",
       "      <td>138.00</td>\n",
       "      <td>1303.0</td>\n",
       "      <td>0.14345</td>\n",
       "      <td>0.35635</td>\n",
       "      <td>0.4049</td>\n",
       "      <td>0.18200</td>\n",
       "      <td>0.3103</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           radius_mean  texture_mean  perimeter_mean  area_mean  \\\n",
       "diagnosis                                                         \n",
       "B               12.200         17.39           78.18      458.4   \n",
       "M               17.325         21.46          114.20      932.0   \n",
       "\n",
       "           smoothness_mean  compactness_mean  concavity_mean  \\\n",
       "diagnosis                                                      \n",
       "B                  0.09076           0.07529         0.03709   \n",
       "M                  0.10220           0.13235         0.15135   \n",
       "\n",
       "           concave points_mean  symmetry_mean  fractal_dimension_mean  ...  \\\n",
       "diagnosis                                                              ...   \n",
       "B                      0.02344         0.1714                0.061540  ...   \n",
       "M                      0.08628         0.1899                0.061575  ...   \n",
       "\n",
       "           fractal_dimension_se  radius_worst  texture_worst  perimeter_worst  \\\n",
       "diagnosis                                                                       \n",
       "B                      0.002808         13.35         22.820            86.92   \n",
       "M                      0.003739         20.59         28.945           138.00   \n",
       "\n",
       "           area_worst  smoothness_worst  compactness_worst  concavity_worst  \\\n",
       "diagnosis                                                                     \n",
       "B               547.4           0.12540            0.16980           0.1412   \n",
       "M              1303.0           0.14345            0.35635           0.4049   \n",
       "\n",
       "           concave points_worst  symmetry_worst  \n",
       "diagnosis                                        \n",
       "B                       0.07431          0.2687  \n",
       "M                       0.18200          0.3103  \n",
       "\n",
       "[2 rows x 29 columns]"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = bc_data.groupby(\"diagnosis\")[bc_data.columns[2:-2].tolist()].median()\n",
    "print(results.shape)\n",
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0466dcf-529d-4a6b-bb3d-eb685854a001",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Export & import data\n",
    "\n",
    "It is very common to export (save) data while we process them - either at the end of the processing workflow, or as intermediate steps. It is perfectly fine to save intermediate files! Especially while you are setting up and fine-tuning the preprocessing steps.\n",
    "\n",
    "Pandas offers convenient methods to export data in different formats: you apply these methods directly on DataFrames (or Series). They are named after the output format, e.g. `.to_<format>()`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "973bcaf1-c6a2-44ae-a0fc-dd1e30018b34",
   "metadata": {},
   "source": [
    "### Pickle\n",
    "\n",
    "Pickle is Python's serialized format. I would say it is the counterpart of R's RDS format (e.g. `saveRDS()`).\n",
    "\n",
    "Common file extension for Pickles are `.pkl` or `.pickle`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "793100f9-37d1-4c8c-ba16-5f60934996a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save processed dataframe to pickle\n",
    "bc_data_subset.to_pickle(DATADIR / \"bc_new_table.pickle\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a056fb7-0d98-4a0a-96c5-d6d33001cae4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read back to a dataframe\n",
    "bc_pickle = pd.read_pickle(DATADIR / \"bc_new_table.pickle\")\n",
    "bc_pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "202a8b1e-ae81-447c-94fb-373b94f4eabe",
   "metadata": {},
   "source": [
    "### CSV/TSV\n",
    "\n",
    "Comma-separated values, or tab-separated values, are textual file formats. They are the most flexible storage type: any text editor or program can open this kind of files. You can share them with everyone.\n",
    "\n",
    "On the downside, CSVs are usually slower and bigger than other binary formats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f28c067a-d3c2-4e93-908b-5a4b337b0f3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save dataframe to TSV\n",
    "bc_data_subset.to_csv(DATADIR / \"bc_new_table.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f49bd00d-0e9c-4717-a5bc-63e30d92f815",
   "metadata": {},
   "source": [
    "### Excel\n",
    "\n",
    "Before reading and saving Excel files with Pandas, you need to install the `openpyxl` library. Copy-paste the following in the cell code below and run it:\n",
    "\n",
    "```\n",
    "!conda install -y -c conda-forge openpyxl\n",
    "```\n",
    "\n",
    "If you are a Windows user, replace `conda` with `conda.exe`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3d73a09-9016-4124-be8f-30bddd707201",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16f23799-a05f-4c6e-bb4a-b2f3f2a6e55c",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc_data_subset.to_excel(DATADIR / \"bc_new_table.xlsx\", sheet_name=\"BC data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c7091ce-80a2-44ed-9cef-61d237cc6e35",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc_excel = pd.read_excel(DATADIR / \"bc_new_table.xlsx\", sheet_name=\"BC data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "594175aa-1f84-4c01-99fe-fc12a0e83fbb",
   "metadata": {},
   "source": [
    "### Feather\n",
    "\n",
    "[Feather](https://arrow.apache.org/docs/python/feather.html) is another binary format, similar to Pickle. But it has the advantage that it can be read by other languages, like R, and it is faster than CSV. Feather is part of the Apache Arrow project. \n",
    "\n",
    "Again, you'll probably need to install a dependency:\n",
    "\n",
    "```\n",
    "!conda install -y -c conda-forge pyarrow\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae04702e-0a4e-4e54-aec1-0df900e6f8e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e31ba1a3-4895-4c33-8cb0-3ca0fe315d1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc_data_subset.to_feather(DATADIR / \"bc_new_table.feather\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c68d3dd-8f01-4870-886c-604961962b74",
   "metadata": {},
   "outputs": [],
   "source": [
    "bc_feather = pd.read_feather(DATADIR / \"bc_new_table.feather\")\n",
    "bc_feather"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cea08000-6c53-4b48-999e-3a94aefdff5c",
   "metadata": {},
   "source": [
    "### Dictionary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db954405-4b71-4cdf-882e-0a012f562c68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert just the first rows\n",
    "bc_dict = bc_data_subset.head().to_dict()\n",
    "print(bc_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d1c78f9-794e-4daf-ba07-a5623dcc9892",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pprint\n",
    "pprint.pprint(bc_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "197c5fb9-6732-4074-a54f-8ecde7a82d50",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "# Credits\n",
    "\n",
    "Partially abridged from great work by Paulo Caldas https://github.com/paulocaldas, Samraat Pawar (MIT license), Center for Computational Biomedicine (Harvard Medical School), and others."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
